

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>pyCGAVariantFetcher &mdash; pyCGA 0.2.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="pyCGA 0.2.0 documentation" href="index.html"/>
        <link rel="next" title="How to implement new web services parsers" href="new_web_services.html"/>
        <link rel="prev" title="PyCGA" href="pyCGA.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="documentation.html" class="icon icon-home"> pyCGA
          

          
          </a>

          
            
            
              <div class="version">
                0.2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html">Quick-start</a></li>
<li class="toctree-l1"><a class="reference internal" href="ids.html">Id Considerations in OpenCGA</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="web_services.html">Web Services API</a></li>
<li class="toctree-l1"><a class="reference internal" href="queries_from_python.html">How To Use API from python</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="pyCGA.html">PyCGA</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">pyCGAVariantFetcher</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#commandline-options">CommandLine Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="new_web_services.html">How to implement new web services parsers</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="documentation.html">pyCGA</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="documentation.html">Docs</a> &raquo;</li>
      
    <li>pyCGAVariantFetcher</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pycgavariantfetcher">
<h1>pyCGAVariantFetcher<a class="headerlink" href="#pycgavariantfetcher" title="Permalink to this headline">¶</a></h1>
<div class="section" id="commandline-options">
<h2>CommandLine Options<a class="headerlink" href="#commandline-options" title="Permalink to this headline">¶</a></h2>
<p>This program can be used to fetch variant easily from openCGA</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pyCGAVariantFetcher</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">studyID</span> <span class="n">study</span> <span class="n">Id</span> <span class="p">[</span><span class="o">--</span><span class="n">ids</span> <span class="n">ids</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">region</span> <span class="n">region</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">chromosome</span> <span class="n">chromosome</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gene</span> <span class="n">gene</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">type</span> <span class="nb">type</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">reference</span> <span class="n">reference</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">alternate</span> <span class="n">alternate</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">files</span> <span class="n">files</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">maf</span> <span class="n">maf</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mgf</span> <span class="n">mgf</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">missingAlleles</span> <span class="n">missingAlleles</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">missingGenotypes</span> <span class="n">missingGenotypes</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">annotationExists</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">annotationDoesNotExist</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">genotype</span> <span class="n">genotype</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">consequence_type</span> <span class="n">consequence_type</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">xref</span> <span class="n">xref</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">biotype</span> <span class="n">biotype</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">polyphen</span> <span class="n">polyphen</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sift</span> <span class="n">sift</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">conservation</span> <span class="n">conservation</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">alternate_frequency</span> <span class="n">alternate_frequency</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">reference_frequency</span> <span class="n">reference_frequency</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">limit</span> <span class="n">reference_frequency</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">skip</span> <span class="n">reference_frequency</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sort</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">group_by</span> <span class="n">group_by</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">exclude</span> <span class="n">exclude</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">unknownGenotype</span> <span class="n">unknownGenotype</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">returnedSamples</span> <span class="n">returnedSamples</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">returnedFiles</span> <span class="n">returnedFiles</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">batchSize</span> <span class="n">batchSize</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">outputType</span> <span class="n">output</span><span class="p">]</span> <span class="o">--</span><span class="n">outputFile</span> <span class="n">outputFile</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--studyID</span></kbd></td>
<td>Required. The id of the study, this can be query usgin pyCGA</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ids</span></kbd></td>
<td>Optional. Select by variant ids (i,e dbSNP ids). Use comma to separate the ids</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--region</span></kbd></td>
<td>Optional. Select by region (i,e chr:start-end). Use comma to separate the region</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--chromosome</span></kbd></td>
<td>Optional. Select by chromosome. Use comma to separate the region</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--gene</span></kbd></td>
<td>Optional. Select by gene name (It can be ENSMBL Ids or HGNCs symbols. Use comma to separate the gene names</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--type</span></kbd></td>
<td>Optional. Select by type of variant  [SNV, MNV, INDEL, SV, CNV]. This filter should not be used alone due to the huge amount of results could be retrieved</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--reference</span></kbd></td>
<td>Optional. Select by reference base(s)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--alternate</span></kbd></td>
<td>Optional. Select by alternate base(s)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--files</span></kbd></td>
<td>Optional. Select by files, only the variants found in these file will be selected. Please note these are the id files in the DB, use pyCGA to get the id files. Use comma to separate the file ids</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--maf</span></kbd></td>
<td>Optional. Select by minor allele frequency. In this filter only the samples in the db are consider, not any external population. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.05)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--mgf</span></kbd></td>
<td>Optional. Select by minor genotype frequency. In this filter only the samples in the db are consider, not any external population. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.1)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--missingAlleles</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select by number of missing alleles in the whole sample set. In this filter only the samples in the db are consider, not any external population. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.05)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--missingGenotypes</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select by number of missing genotypes in the whole sample set. In this filter only the samples in the db are consider, not any external population. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.05)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--annotationExists=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select only the annotated variants</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--annotationDoesNotExist=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select only the variants without annotation</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--genotype</span></kbd></td>
<td>Optional. Select by sample genotype. Samples names must be specified as they are stored the db. Please, find more information in the documentation, Syntax: {samp_1}:{gt_1}(,{gt_n})*(;{samp_n}:{gt_1}(,{gt_n})*)* (e.g. HG0097:0/0;HG0098:0/1,1/1)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--consequence_type</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select by consequence type. Consequence type SO term list. Use comma to separate the SO terms. Please, find the information of SO terms supported in the documentation. (e.g. SO:0000045,SO:0000046</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--xref</span></kbd></td>
<td>Optional. Select by XRef, this is a field used to map ids from different dbs. Please, find the information of dbs supported in the documentation. </td></tr>
<tr><td class="option-group">
<kbd><span class="option">--biotype</span></kbd></td>
<td>Optional. Select by Biotype. Consequence type SO term list. Use comma to separate the biotypes. Please, find the information of biotype terms supported in the documentation. (e.g. protein_coding,retained_intron. This filter should not be used alone due to the huge amount of results could be retrieved</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--polyphen</span></kbd></td>
<td>Optional. Select by polyphen, polyphen score ranges from [0-1]. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.9). This filter is slow if it is not used along others</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--sift</span></kbd></td>
<td>Optional. Select by sift, sift score ranges from [0-1]. Syntax: [&lt;|&gt;|&lt;=|&gt;=]{number}, (e.g, &gt;=0.9). This filter is slow if it is not used along others</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--conservation</span></kbd></td>
<td>Optional. Select by conservation sources. Please read the documentation to find the sources available. Use comma to separate the conservation sources. Syntax: sourceName[&lt;|&gt;|&lt;=|&gt;=]number (e.g. phastCons&gt;0.5,phylop&lt;0.1). . This filter is slow if it is not used along others</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--alternate_frequency</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select by frequency of the alternate allele in one population. Please read the documentation to find the population available. Use comma to separate the population. Syntax: populationName[&lt;|&gt;|&lt;=|&gt;=]number (e.g. 1000g_CEU&gt;0.5,1000g_AFR&lt;0.1)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--reference_frequency</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Select by frequency of the reference allele in one population. Please read the documentation to find the population available. Use comma to separate the population. Syntax: populationName[&lt;|&gt;|&lt;=|&gt;=]number (e.g. 1000g_CEU&gt;0.5,1000g_AFR&lt;0.1)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--limit</span></kbd></td>
<td>Optional. limit (number of results)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--skip</span></kbd></td>
<td>Optional. skip (number of results)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--sort=False</span></kbd></td>
<td>Optional. Sort the output by chromosome coordinates</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--group_by</span></kbd></td>
<td>Optional. Group the output by </td></tr>
<tr><td class="option-group">
<kbd><span class="option">--exclude</span></kbd></td>
<td>Optional. Exclude some parts of the output</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--unknownGenotype=./.</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Returned genotype for unknown genotypes. </td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--returnedSamples</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Only the specified samples will be returned. The samples names in the db, if you have doubts about this, please read the documentation. Use comma to separate the sample names</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--returnedFiles</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. Only the information from the specified files will be returned.Please note these are the id files in the DB, use pyCGA to get the id files. Use comma to separate the files id</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--batchSize=5000</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Optional. This parameter control the size of the batches of variants per query, This number is proportional to the memory used and inversely proportional to the time. This value can not be greater than 5000.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--outputType=json</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Optional. This parameter control the size of the batches of variants per query, This number is proportional to the memory used and inversely proportional to the time. This value can not be greater than 5000.</p>
<p class="last">Possible choices: json, VCF, AVRO-GA4GH, AVRO-OPENCGA</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--outputFile</span></kbd></td>
<td>Optional. Output file with the result</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Get all variants in chromosome 1 with a maf in the database less than 0.05:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pyCGAVariantFetcher</span> <span class="p">[</span><span class="n">sid</span><span class="p">]</span> <span class="o">--</span><span class="n">host</span> <span class="p">[</span><span class="n">hostname</span><span class="p">]</span> <span class="o">--</span><span class="n">studyID</span> <span class="p">[</span><span class="n">studyID</span><span class="p">]</span> <span class="o">--</span><span class="n">chromosome</span> <span class="mi">1</span> <span class="o">--</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="mf">0.05</span>
</pre></div>
</div>
<p>Get all nonsense variants in a certain gene panel with a phylop conservation score more than 0.2:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>panel=PAX6,B3GALTL,SOX2,MFRP,RAX,BCOR,OTX2,SIX6,BMP4,ALDH1A3,COL4A1,BMPR1A,HCCS,CYP1B1,RAB3GAP1,SHH,VSX2,FOXC1,VAX1,FOXE3,FRAS1,FREM1,XPA,HDAC6,STRA6,FREM2,ERCC3,SMOC1,GRIP1,XPC,RAB18,ERCC2,DDB1,RAB3GAP2,DDB2,ABCB6,ERCC6,ERCC4,ERCC5,PITX2,POLH,GDF3,PITX3,ERCC8,GTF2H5,GDF6,ERCC1,MPLKIP,PRSS56,RARB,TENM3,MAB21L2,TBC1D20
pyCGAVariantFetcher [sid] --host [hostname] --studyID [studyID] --consequenceType SO:1000062 --gene $panel --conservation phylop&gt;0.2
</pre></div>
</div>
<p>Get all variants in a certain region with an alternate frequency in AMR population for 1000 Genomes less than 0.1:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pyCGAVariantFetcher</span> <span class="p">[</span><span class="n">sid</span><span class="p">]</span> <span class="o">--</span><span class="n">host</span> <span class="p">[</span><span class="n">hostname</span><span class="p">]</span> <span class="o">--</span><span class="n">studyID</span> <span class="p">[</span><span class="n">studyID</span><span class="p">]</span> <span class="o">--</span><span class="n">region</span> <span class="mi">1</span><span class="p">:</span><span class="mi">1100000</span><span class="o">-</span><span class="mi">1300000</span> <span class="o">--</span><span class="n">alternate_frequency</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you have doubts of how to get and use the ids in OpenCGA, pleas read <a class="reference internal" href="ids.html#how-to-use-ids"><span class="std std-ref">Id Considerations in OpenCGA</span></a></p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Please notice this program is fetching variants from a DB which can contain millions of variants.
Although the DB is indexed to obtain the best performance, several general queries are not supported to be
fast. In this way, for example,  if you try to fetch variants filtering only by type of variants (i,e SNVs),
the result will be very slow, but if you query for a type of variant in a specific region will be very fast.</p>
<dl class="last docutils">
<dt>So we recommend, use at least one of these filters in your queries:</dt>
<dd><ul class="first last simple">
<li>ids</li>
<li>chromosome</li>
<li>region</li>
<li>gene</li>
<li>genotype</li>
<li>consequenceType</li>
<li>xref</li>
<li>alternate_frequency</li>
<li>reference_frequency</li>
<li>maf</li>
<li>mgf</li>
</ul>
</dd>
</dl>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="new_web_services.html" class="btn btn-neutral float-right" title="How to implement new web services parsers" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="pyCGA.html" class="btn btn-neutral" title="PyCGA" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Antonio Rueda, Mathew Parker, Razvan Sultana.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.2.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>