{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "------\n",
    "This notebook is intended to provide guidance on performing advanced common administrator actios required for maintaining a OpenCGA installation, from granting permission to other users to updating the metadata for samples or individuals in the database.\n",
    "\n",
    "Before running this notebook, it is highly recommended to check the official OpenCGA documentation about [Sharing and Permissions](https://app.gitbook.com/@opencb/s/opencga/~/drafts/-Magktiifb08PPiPnoYk/manual/data-management/sharing-and-permissions), [Adding Custom Metadata](https://app.gitbook.com/@opencb/s/opencga/~/drafts/-MbG9-KDgXf7z8BZJCXq/manual/data-management/adding-custom-metadata) and [Data Models](https://app.gitbook.com/@opencb/s/opencga/~/drafts/-MbG9-KDgXf7z8BZJCXq/manual/data-models).\n",
    "\n",
    "**[NOTES]** \n",
    "- For guidance on how to loggin and get started with *pyopencga* you can refer to : [pyopencga_first_steps.ipynb](https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/python/notebooks/user-training)<br>\n",
    "\n",
    "- A good first step when start working with OpenCGA is to explore **Catalog** the OpenCGA component which holds information about our user, the projects and studies our user has permission to access and the clinical data from the studies. For guidance you can refer to : [pyopencga_catalog.ipynb](https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/python/notebooks/user-training)<br>\n",
    "\n",
    "- The server methods used by *pyopencga* client are defined in the following swagger URL: https://ws.opencb.org/opencga-prod/webservices/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents:\n",
    "\n",
    "This Notebook is organised in the following sections:\n",
    "\n",
    "* [Permission Management](#Permission-management)\n",
    "* [Assign Samples to Individuals](#Asign-Samples-to-Individuals)\n",
    "* [Define phenotypes or diseases for Individuals](#Define-phenotypes-or-disorders-for-Individuals)\n",
    "* [Define, add, edit and remove variable sets](#Define,-add,-edit-and-remove-variable-sets)\n",
    "* [Use Cases](#Use-Cases)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup the Client and Login into *pyopencga* \n",
    "\n",
    "### Configuration and Credentials\n",
    "\n",
    "Let's assume we already have *pyopencga* installed in our python setup (all the steps described on [pyopencga_first_steps.ipynb](https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/python/notebooks/user-training)).\n",
    "\n",
    "You need to provide **at least** a host server URL in the standard configuration format for OpenCGA as a python dictionary or in a json file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Step 1. Import pyopencga dependecies\n",
    "from pyopencga.opencga_config import ClientConfiguration # import configuration module\n",
    "from pyopencga.opencga_client import OpencgaClient # import client module\n",
    "from pprint import pprint\n",
    "from IPython.display import JSON\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import datetime\n",
    "\n",
    "## Step 2. OpenCGA host\n",
    "host = 'https://ws.opencb.org/opencga-prod'\n",
    "\n",
    "## Step 3. Create the ClientConfiguration dict\n",
    "config_dict = {'rest': {\n",
    "                       'host': host \n",
    "                    }\n",
    "               }\n",
    "\n",
    "## Step 4. Create the ClientConfiguration and OpenCGA client\n",
    "config = ClientConfiguration(config_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with two clients: **`demo-admin`** and **`demo-guest`** \n",
    "For the purpose of the training, we will work with two users: **`demo-admin`** (which belongs to the `@admins` group, and thus it has admin priviledges) and **`demo-guest`** (which is not admin of the study). \n",
    "\n",
    "__[NOTE]:__ Working with both users is required for following up the queries used in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logged succesfuly as demo-admin to https://ws.opencb.org/opencga-prod, your token is: eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJkZW1vLWFkbWluIiwiYXVkIjoiT3BlbkNHQSB1c2VycyIsImlhdCI6MTYyMjcxNzA2OCwiZXhwIjoxNjIyNzIwNjY4fQ.gMDKB5cy23MMuD3P1SRcDYoo4FftBi21NO7TQjlN4Qo well done!\n"
     ]
    }
   ],
   "source": [
    "## Step 5.1 Define admin user credentials\n",
    "admin_user = 'demo-admin'\n",
    "password = 'demo-admin'\n",
    "admin_client = OpencgaClient(config)\n",
    "admin_client.login(admin_user, password)\n",
    "print('Logged succesfuly as {} to {}, your token is: {} well done!'.format(admin_user, host, admin_client.token))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logged succesfuly as demo-guest to https://ws.opencb.org/opencga-prod, your token is: eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJkZW1vLWd1ZXN0IiwiYXVkIjoiT3BlbkNHQSB1c2VycyIsImlhdCI6MTYyMjcxNzA3MCwiZXhwIjoxNjIyNzIwNjcwfQ.AP1xeBZbGTUQ6JdE0pWGksLbmm8w-LzhBPpN0r4LmPw well done!\n"
     ]
    }
   ],
   "source": [
    "## Step 5.2 Define guest user credentials\n",
    "guest_user = 'demo-guest'\n",
    "password = 'demo-guest'\n",
    "guest_client = OpencgaClient(config)\n",
    "guest_client.login(guest_user, password)\n",
    "print('Logged succesfuly as {} to {}, your token is: {} well done!'.format(guest_user, host, guest_client.token))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup OpenCGA Variables\n",
    "\n",
    "Once we have defined a variable with the client configuration and credentials, we can access to all the methods defined for the client. These methods implement calls to query different data models in *OpenCGA*. \n",
    "\n",
    "- For this training we have created a test project and study. As an administrator, we assume that you're familiar with the concept of the `fqn`. If not, please check the documentation [here]().\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the study we will work with\n",
    "study = 'demo@training:admin'\n",
    "\n",
    "## Define the user ids from some NOT admin demo users \n",
    "user1 = 'trainee1'\n",
    "user2 = 'trainee2'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Permission Management\n",
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In OpenCGA all the permissions are established at the study level. For this part of the notebook, it is really handy to take a look at the documentation about [Sharing and Permissions](https://app.gitbook.com/@opencb/s/opencga/manual/data-management/sharing-and-permissions) in OpenCGA.\n",
    "\n",
    "All the studies come intrinsically with two administrative groups: `@admins` and `@members`.\n",
    "- All the users added to a new study belong to the `@members` group. This is useful for keeping track of the users that have access to that specific study. However, `@members` doesn't have any permission defined by default.\n",
    "- Users belonging to the group `@admins`, on the other hand, have administrator priviledges. The admins can add users to an study and grant them permissions. However, an admin can't provide admin priviledges to other users (only the user `owner` of the study is able to do so)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Add/remove users from a Study\n",
    "\n",
    "### - Add users\n",
    "\n",
    "First, we're going to add the users `trainee1` and `trainee2` to the study. Internally, this means that we are adding those users to the `@members` group of the study.\n",
    "The `studies.update_users()` function has a parameter `action`, which by default is set to **ADD**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Groups in Study demo@training:admin\n",
      "----------------------------------------\n",
      "#id\tuserIds\n",
      "@members\ttrainee1,demo,imedina,pfurio,demo-admin,trainee2,llopez,jcoll,wspooner\n"
     ]
    }
   ],
   "source": [
    "## Add users to study\n",
    "update_users = admin_client.studies.update_users(study=study, group='members', data={\"users\": [\"trainee1\", \"trainee2\"]})\n",
    "update_users.print_results(fields='', metadata=False, title='Groups in Study {}'.format(study))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(update_users.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Remove users\n",
    "\n",
    "Inversely, we can remove the users `trainee1` and `trainee2` from the study. Internally, this means that we are removing those users from the `@members` group of the study.\n",
    "\n",
    "In this case, we need to set the parameter `action`, (which by default is **ADD**) to **REMOVE**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove users from members group of study\n",
    "remove_users = admin_client.studies.update_users(study=study, group='members', action='REMOVE', data={\"users\": [\"trainee1\", \"trainee2\"]})\n",
    "remove_users.print_results(fields='', metadata=False, title='Groups in Study {}'.format(study))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(remove_users.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Create a group and asign permissions\n",
    "\n",
    "The defaul group `@members` does not have any permission for the study. This group only defines which users belong to the study.\n",
    "\n",
    "OpenCGA allows to create admin-defined groups of users. Then, one can grant permissions to this group and assign users. The users in the new group will inherit the permissions granted at a group level."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Create a new `training_group`\n",
    "\n",
    "We can create a new group `training_group` and add the user `trainee1` to the group. The `studies.update_groups()` function has a parameter `action`, which by default is set to **ADD**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the python dict with the name of the group and the users\n",
    "group_data = {\n",
    "  \"id\": \"training_group\",\n",
    "  \"users\": [\n",
    "    \"trainee1\"\n",
    "  ]}\n",
    "\n",
    "## Define a new Training group.\n",
    "update_groups = admin_client.studies.update_groups(study=study, action='ADD', data=group_data)\n",
    "update_groups.print_results(fields='', metadata=False, title='Info about groups')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(update_groups.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Grant permissions to the group `training_group`\n",
    "\n",
    "For granting permissions, we are going to use the template `view_only`. The permissions included in `view_only` are: The member (user or group) will be given full READ permissions to all the entities (samples, individuals, files etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info about permissions for new group\n",
      "-----------------------------------------\n",
      "#@training_group\n",
      "DOWNLOAD_FILES,VIEW_COHORTS,VIEW_AGGREGATED_VARIANTS,VIEW_FAMILIES,VIEW_CLINICAL_ANALYSIS,VIEW_SAMPLES,VIEW_FILES,VIEW_FILE_HEADER,VIEW_COHORT_ANNOTATIONS,VIEW_FAMILY_ANNOTATIONS,VIEW_INDIVIDUALS,VIEW_INDIVIDUAL_ANNOTATIONS,VIEW_FILE_CONTENT,VIEW_JOBS,VIEW_SAMPLE_ANNOTATIONS,VIEW_SAMPLE_VARIANTS,VIEW_PANELS\n"
     ]
    }
   ],
   "source": [
    "## Define the python dict.\n",
    "permission_data = {\n",
    "  \"study\": study,\n",
    "  \"template\": \"view_only\"}\n",
    "\n",
    "## Update the permissions\n",
    "update_acl = admin_client.studies.update_acl(study=study, members='@training_group', action='ADD', data=permission_data)\n",
    "update_acl.print_results(metadata=False, title='Info about permissions for new group')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(update_acl.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Remove `training_group`\n",
    "\n",
    "Don't forget to remove the group before running the cells above. Otherwise you will get an error because the group that you're trying to create already exists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the python dict with the name of the group and the users\n",
    "# group_data = {\n",
    "#   \"id\": \"training_group\",\n",
    "#   \"users\": [\n",
    "#     \"trainee1\"\n",
    "#   ]}\n",
    "\n",
    "## Uncomment this to delete the samples\n",
    "# delete_group = admin_client.studies.update_groups(study=study, action='REMOVE', data=group_data)\n",
    "# delete_group.print_results(metadata=False, title='Info about groups')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(delete_samples.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Check users and groups in the Study\n",
    "We can always check the current status of the `@members` group i.e check which users has access to the study, as well that of any other group defined for the study.\n",
    "\n",
    "#### Hands-on exercise: \n",
    "Run the next cell after adding users `trainee1` and `trainee2`. Then remove them from the group `@members` (code cell below) and run it again. \n",
    "- Do you notice any difference in the output? \n",
    "Independently of the group **ALL** the users always belong to the group `@members`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Groups in Study demo@training:admin\n",
      "----------------------------------------\n",
      "#id\tuserIds\n",
      "@members\ttrainee1,demo,imedina,pfurio,demo-admin,trainee2,llopez,jcoll,wspooner\n",
      "@admins\timedina,pfurio,demo-admin,llopez,jcoll,wspooner\n",
      "@training_group\ttrainee1\n"
     ]
    }
   ],
   "source": [
    "## Check Study groups\n",
    "groups = admin_client.studies.groups(study=study)\n",
    "groups.print_results(fields='', metadata=False, title='Groups in Study {}'.format(study))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(groups.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Asign Samples to Individuals\n",
    "-------\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OpenCGA Catalog allows you to create samples entities. Then, you might need to create an individual entity associated with the sample.\n",
    "\n",
    "**[Important Note:]** When VCFs are ingested into OpenCGA, the pipeline automatically recognises the header of the VCF. If the name of the sample is present, it automatically creates the samples contained in the VCFs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Create samples\n",
    "For this section, it is highly recommended to visit our documentation about the `Sample data model` at [**Sample Data Model**](https://app.gitbook.com/@opencb/s/opencga/manual/data-models/sample).\n",
    "\n",
    "First, let's create 3 samples: `sample1`, `sample2-germinal` and `sample2-somatic`. The most straight-forward way to create samples is using a loop, as showed in the next cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sample: sample1 has been created successfully\n",
      "The sample: sample2-germinal has been created successfully\n",
      "The sample: sample2-somatic has been created successfully\n"
     ]
    }
   ],
   "source": [
    "## Create 3 samples taking into account if they're somatic or germinal\n",
    "sample_ids = ['sample1', 'sample2-germinal', 'sample2-somatic']\n",
    "\n",
    "for sample in sample_ids:\n",
    "    if 'somatic' in sample:\n",
    "        somatic=True\n",
    "    else:\n",
    "        somatic=False\n",
    "    sample_data = {\n",
    "      \"id\": sample,\n",
    "      \"description\": \"germinal sample\",\n",
    "      \"processing\": {\n",
    "        \"preparationMethod\": \"Illumina\",\n",
    "        \"extractionMethod\": \"Parafin Embedded\"\n",
    "      },\n",
    "      \"collection\": {\n",
    "        \"tissue\": \"skin\",\n",
    "        \"organ\": \"skin\",\n",
    "        \"method\": \"biopsy\"\n",
    "      },\n",
    "      \"somatic\": somatic\n",
    "    }\n",
    "\n",
    "    admin_client.samples.create(study=study, data=sample_data)\n",
    "    print('The sample: {} has been created successfully'.format(sample))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Check information about the created Samples\n",
    "\n",
    "We can use the `samples.search()` function to look for information about the recently created samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Search Samples\n",
    "sample_ids = ['sample1', 'sample2-germinal', 'sample2-somatic']\n",
    "\n",
    "for sample in sample_ids:\n",
    "    samples = admin_client.samples.search(study=study, id=sample)\n",
    "    samples.print_results(fields='id,collection,somatic', metadata=False, title='Info about sample {}'.format(sample))\n",
    "    print('\\n')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "samples = admin_client.samples.search(study=study, id=sample_ids)\n",
    "# JSON(samples.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Remove Samples:\n",
    "\n",
    "__[NOTE]:__ Be aware that if you delete the samples at this stage, you won't be able to associate the `samples` to the correspondent `individuals` in the next subsection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Uncomment this to delete the samples\n",
    "# delete_samples = admin_client.samples.delete(study=study, samples=['sample1,sample2-germinal,sample2-somatic'])\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(delete_samples.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Create individuals \n",
    "\n",
    "Now let's create a couple of `individuals`. For this section, it is highly recommended to visit our documentation about the individual data model at [**Individual Data Model**](https://app.gitbook.com/@opencb/s/opencga/manual/data-models/individual)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create 2 individuals \n",
    "individual_ids = ['individual1', 'individual2']\n",
    "\n",
    "for individual in individual_ids:\n",
    "    individual_data= {\n",
    "  \"id\": individual,\n",
    "  \"name\": individual,\n",
    "  \"location\": {\n",
    "    \"postalCode\": \"CB1\",\n",
    "  },\n",
    "  \"sex\": \"MALE\",\n",
    "  \"ethnicity\": \"black\",\n",
    "  \"parentalConsanguinity\":'true',\n",
    "  \"population\": {\n",
    "    \"name\": \"spanish\"\n",
    "  },\n",
    "  \"dateOfBirth\": \"19900587\",\n",
    "  \"karyotypicSex\": \"XY\",\n",
    "  \"lifeStatus\": \"ALIVE\"}\n",
    "    admin_client.individuals.create(study=study, data=individual_data)\n",
    "    print('The individual {} has been created succesfully'.format(individual))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Check information about the created individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print information about the individuals\n",
    "individual_ids = ['individual1', 'individual2']\n",
    "for individual in individual_ids:\n",
    "    individuals = admin_client.individuals.search(study=study, id=individual)\n",
    "    individuals.print_results(fields='id,name,sex,ethnicity,lifeStatus,population,dateOfBirth', metadata=False, title='Info about individual {}'.format(individual))\n",
    "    print('\\n')\n",
    "    \n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "individuals = admin_client.individuals.search(study=study, id=individual_ids)\n",
    "# JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Associate samples with individuals\n",
    "\n",
    "Remember that OpenCGA `individual` data model allows to have multiple samples assigned to the same individual. A typical use case can be found in cancer genetic screen, where usually two samples are taken from the same individual: one from the tumour (somatic sample), and one germinal.\n",
    "\n",
    "Bearing that in mind, we can create some individual entities in the database. In this case, let's assume that `sample1` corresponds to `individual1`, whilst `sample2-germinal` and `sample2-somatic` correspond to `individual2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define a dictionary with the samples for each individual\n",
    "individual_ids = ['individual1', 'individual2']\n",
    "sample_dict={'individual1':['sample1'], 'individual2':['sample2-germinal', 'sample2-somatic']}\n",
    "\n",
    "## Asign Samples to individuals\n",
    "for individual in individual_ids:\n",
    "    samples_associated = sample_dict[individual]\n",
    "    admin_client.individuals.update(study=study, individuals=individual, data={'samples':samples_associated})\n",
    "    print('The individual {} has been updated successfully!'.format(individual))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Check Samples associated to the Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print information about the individuals with their correspondent samples\n",
    "individual_ids = ['individual1', 'individual2']\n",
    "for individual in individual_ids:\n",
    "    individuals = admin_client.individuals.search(study=study, id=individual)\n",
    "    individuals.print_results(fields='id,samples.id,samples.somatic,samples.collection.tissue,samples.collection.method', metadata=False, title='Info about individual {} with the samples'.format(individual))\n",
    "    print('\\n')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "individuals = admin_client.individuals.search(study=study, id=individual_ids, include='samples')\n",
    "# JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Delete Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Uncomment this to delete the individuals\n",
    "delete_individuals = admin_client.individuals.delete(study=study, individuals='individual1,individual2')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(delete_individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define phenotypes or disorders for Individuals\n",
    "------\n",
    "\n",
    "OpenCGA defines a set of rich data models, that can be useful for storing any kind of annotation. The individual entity can be associated with phenotypes and diseases ontologies. We support many ontology sources. Individuals that can be easily queried by its associated phenotypes and diseases once they are annotated.\n",
    "\n",
    "For this section, it is highly recommended to visit our documentation about the individual data model at [**Individual Data Model**](https://app.gitbook.com/@opencb/s/opencga/manual/data-models/individual)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Update already existing individuals with phenotypes\n",
    "\n",
    "### - Create new `individuals`\n",
    "Lets create a couple of new individuals for this part of the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create 2 new individuals\n",
    "individual_ids = ['individual_A', 'individual_B']               \n",
    "for individual in individual_ids:\n",
    "    individual_data= {\n",
    "  \"id\": individual,\n",
    "  \"name\": individual,\n",
    "  \"location\": {\n",
    "    \"postalCode\": \"47400A\",\n",
    "  },\n",
    "  \"sex\": \"FEMALE\",\n",
    "  \"ethnicity\": \"black\",\n",
    "  \"parentalConsanguinity\":'false',\n",
    "  \"population\": {\n",
    "    \"name\": \"british\"\n",
    "  },\n",
    "  \"dateOfBirth\": \"19900587\",\n",
    "  \"karyotypicSex\": \"XX\",\n",
    "  \"lifeStatus\": \"ALIVE\"}\n",
    "    admin_client.individuals.create(study=study, data=individual_data)\n",
    "    print('The individual {} has been created succesfully'.format(individual))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Assign Phenotypes\n",
    "Now we can assign them the desired phenotypes by using `individuals.update()`. \n",
    "\n",
    "Remember that one individual might have more than one phenotype (`individual_B` in the next example)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define a dictionary with the phenotyoes for each individual\n",
    "individual_ids = ['individual_A', 'individual_B']               \n",
    "phenotype_dict = {\n",
    "    'individual_A':[{'id':'HP:0001626',\n",
    "                    'name':'phenotype_A',\n",
    "                    'source':'HPO',\n",
    "                    'ageOfOnset':'15',\n",
    "                    'status':'OBSERVED'}], \n",
    "    'individual_B':[{'id':'HP:0000119',\n",
    "                     'name':'phenotype_B',\n",
    "                     'source':'HPO',\n",
    "                     'status':'OBSERVED'}, \n",
    "                    {'id':'HP:0000478',\n",
    "                     'name':'phenotype_C',\n",
    "                     'source':'HPO',\n",
    "                     'status': 'OBSERVED'}]\n",
    "                    }\n",
    "\n",
    "\n",
    "## Assign Phenotypes to individuals\n",
    "for individual in individual_ids:\n",
    "    phenotypes_associated = phenotype_dict[individual]\n",
    "    admin_client.individuals.update(study=study, individuals=individual, data={'phenotypes':phenotypes_associated})\n",
    "    print('The individual {} has been updated successfully!'.format(individual))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Check Phenotypes of the Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info about the phenotypes of individual: individual_A\n",
      "----------------------------------------------------------\n",
      "#id\tphenotypes.id\tphenotypes.name\tphenotypes.source\tphenotypes.status\n",
      "individual_A\tHP:0001626\tphenotype_A\tHPO\tOBSERVED\n",
      "\n",
      "\n",
      "Info about the phenotypes of individual: individual_B\n",
      "----------------------------------------------------------\n",
      "#id\tphenotypes.id\tphenotypes.name\tphenotypes.source\tphenotypes.status\n",
      "individual_B\tHP:0000119,HP:0000478\tphenotype_B,phenotype_C\tHPO\tOBSERVED\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Print information about the individuals with their correspondent samples\n",
    "individual_ids = ['individual_A', 'individual_B']               \n",
    "for individual in individual_ids:\n",
    "    individuals = admin_client.individuals.search(study=study, id=individual)\n",
    "    individuals.print_results(fields='id,phenotypes.id,phenotypes.name,phenotypes.source,phenotypes.status', metadata=False, title='Info about the phenotypes of individual: {}'.format(individual))\n",
    "    print('\\n')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "individuals = admin_client.individuals.search(study=study, id=individual_ids, include='phenotypes')\n",
    "# JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Assign Disorders\n",
    "\n",
    "Similarly to phenotypes, we can also update the disorders for an Individual."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define a dictionary with the Disorders for each Individual\n",
    "individual_ids = ['individual_A', 'individual_B']               \n",
    "disorders_dict = {\n",
    "    'individual_A':[{'id':'OMIM:300600',\n",
    "                    'name':'disorder_A',\n",
    "                    'source':'OMIM',\n",
    "                    'description':'Aland Island Eye Disease'}], \n",
    "    'individual_B':[{'id':'OMIM:203450',\n",
    "                     'name':'disorder_B',\n",
    "                     'source':'OMIM',\n",
    "                     'description':'Alexander Disease'}, \n",
    "                    {'id':'OMIM:104300',\n",
    "                     'name':'disorder_C',\n",
    "                     'source':'OMIM',\n",
    "                     'description':'Alzheimer Disease'}]\n",
    "                    }\n",
    "\n",
    "## Assign Disorders to Individuals\n",
    "for individual in individual_ids:\n",
    "    disorders_associated = disorders_dict[individual]\n",
    "    admin_client.individuals.update(study=study, individuals=individual, data={'disorders':disorders_associated})\n",
    "    print('The individual {} has been updated successfully!'.format(individual))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Check Disorders of the Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print information about the individuals with their correspondent samples\n",
    "individual_ids = ['individual_A', 'individual_B']               \n",
    "for individual in individual_ids:\n",
    "    individuals = admin_client.individuals.search(study=study, id=individual)\n",
    "    individuals.print_results(fields='id,disorders.id,disorders.name,disorders.source,disorders.description', metadata=False, title='Info about the disorders of individual: {}'.format(individual))\n",
    "    print('\\n')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "individuals = admin_client.individuals.search(study=study, id=individual_ids, include='disorders')\n",
    "# JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Delete Individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Uncomment this to delete the individuals\n",
    "delete_individuals = admin_client.individuals.delete(study=study, individuals=individual_ids)\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(delete_individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Create individual with associated phenotypes and disorders\n",
    "\n",
    "If we have phenotype and disorder information about one individual at the time of creating the entity in OpenCGA Catalog, we cal always create the individual with all the associated information directly.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Construct an individual data model\n",
    "individual_data = {\n",
    "  \"id\": \"Individual_with_data\",\n",
    "  \"name\": \"Jonh\",\n",
    "  \"location\": {\n",
    "    \"city\": \"New York\",\n",
    "    \"country\": \"USA\"\n",
    "  },\n",
    "  \"sex\": \"MALE\",\n",
    "  \"ethnicity\": \"american\",\n",
    "  \"dateOfBirth\": \"18800722\",\n",
    "  \"karyotypicSex\": \"XY\",\n",
    "  \"lifeStatus\": \"ALIVE\",\n",
    "  \"phenotypes\": [{'id':'HP:0001626',\n",
    "                  'name':'phenotype_A',\n",
    "                  'source':'HPO',\n",
    "                  'ageOfOnset':'15',\n",
    "                  'status':'OBSERVED'}],\n",
    "  \"disorders\": [{'id':'OMIM:104300',\n",
    "                 'name':'disorder_C',\n",
    "                 'source':'OMIM',\n",
    "                 'description':'Alzheimer Disease',\n",
    "                 'evidences':[{\n",
    "                      'id':'HP:0001626',\n",
    "                      'name':'phenotype_A',\n",
    "                      'source':'HPO',\n",
    "                  'ageOfOnset':'15',\n",
    "                  'status':'OBSERVED'}]}]\n",
    "                }\n",
    "\n",
    "## Create the individual\n",
    "Individual_with_data = admin_client.individuals.create(study=study, data=individual_data)\n",
    "print('The individual has been successfully created')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Hands-on Exercise: Check the individual data model\n",
    "\n",
    "Tun the next cell to display the data from the created individual using the interactive JSON viewer. \n",
    "\n",
    "Now compare the data with the individual data model at [**Individual Data Model**](https://app.gitbook.com/@opencb/s/opencga/manual/data-models/individual). Can you see the parallelism?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Display an interactive JSON viewer of the REST response\n",
    "individuals = admin_client.individuals.search(study=study, id='Individual_with_data')\n",
    "JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Delete the individual\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Uncomment this to delete the individuals\n",
    "delete_individuals = admin_client.individuals.delete(study=study, individuals='Individual_with_data')\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "# JSON(delete_individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define, add, edit and remove variable sets\n",
    "----------\n",
    "One incredibly practical utility of OpenCGA is the possibility of defining variable sets (i.e: groups of variables) for a specific entity or group of entities (Samples, Individuals, Cohorts, etc.)\n",
    "\n",
    "For more information, and before running this part of the notebook, refer to the section of [Adding custom metadata](https://app.gitbook.com/@opencb/s/opencga/manual/data-management/adding-custom-metadata) in our OpenCGA GitBook Documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Define the Variable Set\n",
    "In this case, we will create a Variable set for the *annotable* entity `Sample`. \n",
    "\n",
    "According to our documentation, we consider an entry to be Annotable if the entry can have Annotation Sets. At this stage, only **File, Sample, Individual, Cohort and Family** are Annotable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the variable set\n",
    "sample_variable_set = {\n",
    "  \"id\": \"sample_metadata\",  \n",
    "  \"unique\": \"true\",\n",
    "  \"confidential\": \"false\",\n",
    "  \"description\": \"Sample origin\",\n",
    "  \"entities\": [\n",
    "    \"SAMPLE\"\n",
    "  ],\n",
    "  \"variables\": [\n",
    "    {\n",
    "      \"id\": \"cell_line\",\n",
    "      \"name\": \"Cell line\",\n",
    "      \"category\": \"string\",\n",
    "      \"type\": \"TEXT\",\n",
    "      \"required\": \"false\",\n",
    "      \"multiValue\": \"false\",\n",
    "      \"allowedValues\": [],\n",
    "      \"rank\": 2,\n",
    "      \"dependsOn\": \"\",\n",
    "      \"description\": \"Sample cell line\",\n",
    "      \"attributes\": {}\n",
    "    },\n",
    "      {\n",
    "      \"id\": \"cell_type\",\n",
    "      \"name\": \"Cell type\",\n",
    "      \"category\": \"string\",\n",
    "      \"type\": \"TEXT\",\n",
    "      \"required\": \"false\",\n",
    "      \"multiValue\": \"false\",\n",
    "      \"allowedValues\": [],\n",
    "      \"rank\": 3,\n",
    "      \"dependsOn\": \"\",\n",
    "      \"description\": \"Sample cell type\",\n",
    "      \"attributes\": {}\n",
    "    }]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add the variable set to the Sample data model of the study\n",
    "add_variable_set = admin_client.studies.update_variable_sets(study=study, action='ADD', data=sample_variable_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of Variable Sets in study: demo@training:admin\n",
      "--------------------------------------------------------\n",
      "#id\tname\tunique\tdescription\n",
      "opencga_file_variant_stats\topencga_file_variant_stats\tTrue\tOpenCGA file variant stats\n",
      "opencga_alignment_stats\topencga_alignment_stats\tTrue\tOpenCGA alignment stats\n",
      "opencga_sample_qc\topencga_sample_qc\tTrue\tOpenCGA variable set for sample quality control (QC)\n",
      "opencga_cohort_variant_stats\topencga_cohort_variant_stats\tTrue\tOpenCGA cohort variant stats\n",
      "sample_metadata\tsample_metadata\tTrue\tSample origin\n"
     ]
    }
   ],
   "source": [
    "## Search for the variable set\n",
    "variable_sets = admin_client.studies.variable_sets(study=study)\n",
    "variable_sets.print_results(fields='id,name,unique,description', metadata=False, title='List of Variable Sets in study: {}'.format(study))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## - Remove Variable Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove the variable set to the Sample data model of the study\n",
    "#remove_variable_set = admin_client.studies.update_variable_sets(study=study, action='REMOVE', data=sample_variable_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Annotate the Variable Set: Annotation Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the annotation set\n",
    "sample_annotation_set = [{\n",
    "    \"id\": \"annotation_set_id\",\n",
    "    \"variableSetId\": \"sample_metadata\",\n",
    "    \"annotations\": {\n",
    "    \"cell_line\": \"umbilical cord\",\n",
    "    \"cell_type\": \"multipotent progenitor\",\n",
    "  }\n",
    "}]\n",
    "\n",
    "## Annotate the variable set\n",
    "admin_client.samples.update(samples='sample1', data={'annotationSets':sample_annotation_set})\n",
    "samples = admin_client.samples.search(study=study)\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer of the REST response\n",
    "JSON(samples.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use Cases\n",
    "------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}