<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Zetta Genomics" />

<meta name="date" content="2024-11-28" />

<title>Introduction to opencgaR</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to opencgaR</h1>
<h4 class="author">Zetta Genomics</h4>
<h4 class="date">28 November 2024</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>OpenCGA is an open-source project that aims to provide a <strong>Big
Data storage engine and analysis framework for genomic scale data
analysis</strong> of hundreds of terabytes or even petabytes. For users,
its main features include uploading and downloading files to a
repository, storing their information in a generic way (non-dependant of
the original file-format) and retrieving this information efficiently.
For developers, it will be a platform for supporting the most used
bioinformatics file formats and accelerating the development of
visualization and analysis applications.</p>
<p>The OpencgaR client provides a user-friendly interface to work with
OpenCGA REST Web Services through R and has been implemented following
the Bioconductor guidelines for package development which promote
high-quality, well documented and inter-operable software. The source
code of the R package can be found in <a href="https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/R">GitHub</a>.</p>
<div id="installation-and-configuration" class="section level2">
<h2>Installation and configuration</h2>
<p>The R client requires at least R version 3.4, although most of the
code is fully compatible with earlier versions. The pre-build R package
of the R client can be downloaded from the OpenCGA GitHub Release at <a href="https://github.com/opencb/opencga/releases" class="uri">https://github.com/opencb/opencga/releases</a> and installed
using the <code>install.packages</code> function in R.
<code>install.packages</code> can also install a source package from a
remote <code>.tar.gz</code> file by providing the URL to such file.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## Install opencgaR</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># install.packages(&quot;opencgaR_4.0.0.tar.gz&quot;, repos=NULL, type=&quot;source&quot;)</span></span></code></pre></div>
</div>
</div>
<div id="connection-and-authentication-into-an-opencga-instance" class="section level1">
<h1>Connection and authentication into an OpenCGA instance</h1>
<p>A set of methods have been implemented to deal with the connectivity
and log into the REST host. Connection to the host is done in two steps
using the functions <strong><em>initOpencgaR</em></strong> and
<strong><em>opencgaLogin</em></strong> for defining the connection
details and login, respectively.</p>
<p>The <strong><em>initOpencgaR</em></strong> function accepts either
host and version information or a configuration file (as a
<code>list()</code> or in <a href="http://docs.opencb.org/display/opencga/client-configuration.yml">YAML
or JSON format</a>). The <strong><em>opencgaLogin</em></strong> function
establishes the connection with the host. It requires an opencgaR object
(created using the <code>initOpencgaR</code> function) and the login
details: user and password. User and password can be introduced
interactively through a popup window using
<code>interactive=TRUE</code>, to avoid typing user credentials within
the R script or a config file.</p>
<p>The code below shows different ways to initialise the OpenCGA
connection with the REST server.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">## Initialise connection using a configuration in R list</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>conf <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">version=</span><span class="st">&quot;v2&quot;</span>, <span class="at">rest=</span><span class="fu">list</span>(<span class="at">host=</span><span class="st">&quot;https://test.app.zettagenomics.com/myenv/opencga/&quot;</span>))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">initOpencgaR</span>(<span class="at">opencgaConfig=</span>conf)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">## Initialise connection using a configuration file (in YAML or JSON format)</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># conf &lt;- &quot;/path/to/conf/client-configuration.yml&quot;</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># con &lt;- initOpencgaR(opencgaConfig=conf)</span></span></code></pre></div>
<p>Once the connection has been initialised users can login specifying
their OpenCGA user ID and password.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Log in</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">opencgaLogin</span>(<span class="at">opencga =</span> con, <span class="at">userid =</span> <span class="st">&quot;userID&quot;</span>, <span class="at">passwd =</span> <span class="st">&quot;1234567890&quot;</span>, </span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>                    <span class="at">organization =</span> <span class="st">&quot;myorg&quot;</span>, <span class="at">autoRenew =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, </span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>                    <span class="at">showToken =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
</div>
<div id="retrieving-basic-information-about-projects-users-and-studies" class="section level1">
<h1>Retrieving basic information about projects, users and studies</h1>
<p>Retrieving the list of projects accessible to the user:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>projects <span class="ot">&lt;-</span> <span class="fu">projectClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;search&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">getResponseResults</span>(projects)[[<span class="dv">1</span>]][,<span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;name&quot;</span>, <span class="st">&quot;description&quot;</span>)]</span></code></pre></div>
<p>List the studies accessible in a project</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>projects <span class="ot">&lt;-</span> <span class="fu">projectClient</span>(<span class="at">OpencgaR=</span>con, <span class="at">project=</span><span class="st">&quot;population&quot;</span>, <span class="at">endpointName=</span><span class="st">&quot;studies&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">getResponseResults</span>(projects)[[<span class="dv">1</span>]][,<span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;name&quot;</span>, <span class="st">&quot;description&quot;</span>)]</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># study &lt;- &quot;test@population:1000G&quot;</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>study <span class="ot">&lt;-</span> <span class="st">&quot;test@germline:platinum&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># Study name</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>study_result <span class="ot">&lt;-</span> <span class="fu">studyClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">studies =</span> study, <span class="at">endpointName =</span> <span class="st">&quot;info&quot;</span>)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>study_name <span class="ot">&lt;-</span> <span class="fu">getResponseResults</span>(study_result)[[<span class="dv">1</span>]][,<span class="st">&quot;name&quot;</span>]</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># Number of samples in this study</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>count_samples <span class="ot">&lt;-</span> <span class="fu">sampleClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;search&quot;</span>, </span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>                              <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">count=</span><span class="cn">TRUE</span>, <span class="at">limit=</span><span class="dv">0</span>))</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>num_samples <span class="ot">&lt;-</span> <span class="fu">getResponseNumMatches</span>(count_samples)</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co"># Number of individuals</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>count_individuals <span class="ot">&lt;-</span> <span class="fu">individualClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;search&quot;</span>, </span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>                                      <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">count=</span><span class="cn">TRUE</span>, <span class="at">limit=</span><span class="dv">0</span>))</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>num_individuals <span class="ot">&lt;-</span> <span class="fu">getResponseNumMatches</span>(count_individuals)</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co"># Number of clinical cases</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>count_cases <span class="ot">&lt;-</span> <span class="fu">clinicalClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;search&quot;</span>, </span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>                              <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">count=</span><span class="cn">TRUE</span>, <span class="at">limit=</span><span class="dv">0</span>))</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>num_cases <span class="ot">&lt;-</span> <span class="fu">getResponseNumMatches</span>(count_cases)</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co"># Number of variants</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>count_variants <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">count=</span><span class="cn">TRUE</span>, <span class="at">limit=</span><span class="dv">0</span>))</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>num_variants <span class="ot">&lt;-</span> <span class="fu">getResponseNumMatches</span>(count_variants)</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>to_print <span class="ot">&lt;-</span> <span class="st">&quot;Study name: {study_name}</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="st">            - Number of samples: {num_samples}</span></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="st">            - Number of individuals: {num_individuals}</span></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="st">            - Number of cases: {num_cases}</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a><span class="st">            - Number of variants: {num_variants}&quot;</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a><span class="fu">glue</span>(to_print)</span></code></pre></div>
</div>
<div id="extracting-information-about-variants-and-biomarkers-of-interest" class="section level1">
<h1>Extracting information about variants and biomarkers of
interest</h1>
<div id="fetch-the-samples-containing-a-specific-variant" class="section level2">
<h2>Fetch the samples containing a specific variant</h2>
<p>As an example we are going to retrieve a random variant from the
database</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Retrieve the first variant</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>variant_example <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                 <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">limit=</span><span class="dv">1</span>, <span class="at">type=</span><span class="st">&quot;SNV&quot;</span>))</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>variant_id <span class="ot">&lt;-</span> <span class="fu">getResponseResults</span>(variant_example)[[<span class="dv">1</span>]][,<span class="st">&quot;id&quot;</span>]</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Variant example: {variant_id}&quot;</span>)</span></code></pre></div>
</div>
<div id="fetch-the-samples-containing-a-specific-variant-1" class="section level2">
<h2>Fetch the samples containing a specific variant</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Retrieve the samples that have this variant</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;querySample&quot;</span>, </span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">variant=</span>variant_id))</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of samples with this variant: </span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="st">     {getResponseAttributes(variant_result)$numTotalSamples}&quot;</span>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>data_keys <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]][, <span class="st">&quot;studies&quot;</span>][[<span class="dv">1</span>]][, <span class="st">&quot;sampleDataKeys&quot;</span>])</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]][, <span class="st">&quot;studies&quot;</span>][[<span class="dv">1</span>]][, <span class="st">&quot;samples&quot;</span>][[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> </span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>        <span class="fu">select</span>(<span class="sc">-</span>fileIndex) <span class="sc">%&gt;%</span> <span class="fu">unnest_wider</span>(data, <span class="at">names_sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="fu">colnames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;samples&quot;</span>, data_keys)</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>df</span></code></pre></div>
</div>
<div id="fetch-samples-by-genotype" class="section level2">
<h2>Fetch samples by genotype</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Search for homozygous alternate samples</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>genotype <span class="ot">&lt;-</span> <span class="st">&quot;1/1&quot;</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;querySample&quot;</span>, </span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">variant=</span>variant_id, </span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>                                            <span class="at">genotype=</span>genotype))</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of samples with genotype {genotype} in this variant: </span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="st">     {getResponseAttributes(variant_result)$numTotalSamples}&quot;</span>)</span></code></pre></div>
</div>
<div id="retrieve-variants-by-region" class="section level2">
<h2>Retrieve variants by region</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Search for homozygous alternate samples</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>region <span class="ot">&lt;-</span> <span class="st">&quot;1:62273600-62273700&quot;</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>region <span class="ot">&lt;-</span> <span class="st">&quot;11:119345500-119345600&quot;</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">region=</span>region, <span class="at">limit=</span><span class="dv">100</span>))</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants in region {region}: {getResponseNumResults(variant_result)}&quot;</span>)</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
</div>
<div id="retrieve-variants-by-feature" class="section level2">
<h2>Retrieve variants by feature</h2>
<p>Variants can be filtered by gene using the parameters gene or
xref:</p>
<ul>
<li><code>gene</code>: only accepts gene IDs</li>
<li><code>xref</code>: accepts different IDs including gene, transcript,
dbSNP, … Remember you can pass different IDs using comma as
separator.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## Filter by gene</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="st">&quot;HCN3,MTOR&quot;</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">gene=</span>genes, <span class="at">limit=</span><span class="dv">10</span>))</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">## Filter by xref</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#snp &lt;- &quot;rs147394986&quot;</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>hgvs <span class="ot">&lt;-</span> <span class="st">&quot;NP_002747.2:p.Arg67Trp&quot;</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">xref=</span>hgvs, <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
</div>
<div id="retrieve-variants-by-consequence-type" class="section level2">
<h2>Retrieve variants by consequence type</h2>
<p>OpenCGA provides a rich variant annotation that includes Ensembl
consequence types (<a href="https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html" class="uri">https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html</a>).
You can filter by consequence type by using parameter <code>ct</code>.
You can provide a list of consequence type names separated by comma.
Also, an alias called <code>lof</code> filter by a combination of
loss-of-function terms.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="do">## Filter by missense variants and stop_gained</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>ct <span class="ot">&lt;-</span> <span class="st">&quot;missense_variant,stop_gained&quot;</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">ct=</span>ct, <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of missense and stop gained variants: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="do">## Filter by LoF (alias created containing 9 different CTs)</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">ct=</span><span class="st">&quot;lof&quot;</span>, <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of LoF variants: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
</div>
<div id="filter-variants-by-population-frequencies" class="section level2">
<h2>Filter variants by population frequencies</h2>
<p>OpenCGA allows to filter variants by population frequencies,
including:</p>
<ul>
<li>Minor Allele frequency (MAF) with the parameter
<code>populationFrequencyMaf</code></li>
<li>Alternate Allele frequency (ALT) with the parameter
<code>populationFrequencyAlt</code></li>
<li>Reference Allele frequency with the parameter
<code>populationFrequencyRef</code></li>
</ul>
<p>The population frequency studies indexed in OpenCGA include different
sources such as <strong>genomAD</strong> or <strong>1000
Genomes</strong>.</p>
<p>The syntax for the query parameter is:
<code>{study}:{population}:{cohort}[&lt;|&gt;|&lt;=|&gt;=]{proportion}</code>.
Note that you can specify several populations separated by comma (OR) or
by semi-colon (AND), e.g. for all variants less than 1% in the two
studies we should use
<code>1kG_phase3:ALL&lt;0.01;GNOMAD_GENOMES:ALL&lt;0.01</code></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">## Filter by population alternate frequency</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>population_frequency_alt <span class="ot">&lt;-</span> <span class="st">&#39;1000G:ALL&lt;0.001&#39;</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, </span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>                                            <span class="at">populationFrequencyAlt=</span>population_frequency_alt, </span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co"># getResponseResults(variant_result)[[1]] %&gt;% select(-names, -studies, -annotation)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="do">## Filter by two population alternate frequency</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="do">## Remember to use commas for OR and semi-colon for AND</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>population_frequency_alt <span class="ot">&lt;-</span> <span class="st">&#39;1000G:ALL&lt;0.001;GNOMAD_GENOMES:ALL&lt;0.001&#39;</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, </span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>                                            <span class="at">populationFrequencyAlt=</span>population_frequency_alt, </span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="do">## Filter by population alternate frequency using a range of values</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>population_frequency_alt <span class="ot">&lt;-</span> <span class="st">&#39;1000G:ALL&gt;0.001;1000G:ALL&lt;0.005&#39;</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, </span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>                                            <span class="at">populationFrequencyAlt=</span>population_frequency_alt, </span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="co"># getResponseResults(variant_result)[[1]] %&gt;% select(-names, -studies, -annotation)</span></span></code></pre></div>
</div>
<div id="complex-variant-queries" class="section level2">
<h2>Complex variant queries</h2>
<p>OpenCGA implements a very advanced variant query engine that allows
to combine many filters to build very complex and useful queries. In
this section you will find some examples.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="do">## Filter by Consequence Type, Clinical Significance and population frequency</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>ct <span class="ot">&lt;-</span> <span class="st">&#39;lof,missense_variant&#39;</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>clinicalSignificance <span class="ot">&lt;-</span> <span class="st">&#39;likely_pathogenic,pathogenic&#39;</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>populationFrequencyAlt <span class="ot">&lt;-</span> <span class="st">&#39;1000G:ALL&lt;0.01;GNOMAD_GENOMES:ALL&lt;0.01&#39;</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">limit=</span><span class="dv">5</span>, <span class="at">ct=</span>ct, </span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>                                            <span class="at">clinicalSignificance=</span>clinicalSignificance,</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>                                            <span class="at">populationFrequencyAlt=</span>population_frequency_alt))</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>names, <span class="sc">-</span>studies, <span class="sc">-</span>annotation)</span></code></pre></div>
</div>
<div id="filter-variants-using-cohort-information" class="section level2">
<h2>Filter variants using cohort information</h2>
<p>OpenCGA allows users to define cohorts of samples and calculate and
index the allele and genotype frequencies among other stats. By default,
a cohort called “ALL” containing all samples is defined and the variant
stats are calculated. You can filter by the internal cohort stats using
the parameter <code>cohortStatsAlt</code> and pass the study and the
cohort you would like to filter by. Format:
<code>cohortStatsAlt={study}:{cohort}[&lt;|&gt;|&lt;=|&gt;=]{proportion}</code>
Note that any cohorts created will be shared among the studies belonging
to the same project.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="do">## Filter by cohort ALL frequency</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>cohort_stats <span class="ot">&lt;-</span> <span class="fu">paste0</span>(study, <span class="st">&#39;:ALL&lt;0.001&#39;</span>)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">cohortStatsAlt=</span>cohort_stats, </span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a><span class="co"># getResponseResults(variant_result)[[1]] %&gt;% select(-names, -studies, -annotation)</span></span></code></pre></div>
<p>The following examples are just examples of how queries should be
made, they are not intended to be executed since the custom cohorts have
not been created.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="do">## Filter by custom cohorts&#39; frequency</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>cohort_stats <span class="ot">=</span> study <span class="sc">+</span> <span class="st">&#39;:MY_COHORT_A&lt;0.001&#39;</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">cohortStatsAlt=</span>cohort_stats, </span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="do">## Filter by more than one cohort frequency and consequence type</span></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>cohort_stats <span class="ot">=</span> study <span class="sc">+</span> <span class="st">&#39;:MY_COHORT_A&lt;0.001&#39;</span> <span class="sc">+</span> <span class="st">&#39;;&#39;</span> <span class="sc">+</span> study <span class="sc">+</span> <span class="st">&#39;:MY_COHORT_B&lt;0.001&#39;</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">cohortStatsAlt=</span>cohort_stats, </span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a>                                            <span class="at">limit=</span><span class="dv">5</span>))</span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Number of variants matching the filter: {getResponseNumMatches(variant_result)}&quot;</span>)</span></code></pre></div>
</div>
<div id="retrieve-cohort-frequencies-for-a-specific-variant" class="section level2">
<h2>Retrieve cohort frequencies for a specific variant</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>variant_result <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;query&quot;</span>, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                                <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">id=</span>variant_id))</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">glue</span>(<span class="st">&quot;Variant example: {variant_id}&quot;</span>)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="fu">getResponseResults</span>(variant_result)[[<span class="dv">1</span>]]<span class="sc">$</span>studies[[<span class="dv">1</span>]]<span class="sc">$</span>stats[[<span class="dv">1</span>]][,<span class="fu">c</span>(<span class="st">&#39;cohortId&#39;</span>, <span class="st">&#39;alleleCount&#39;</span>,</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>                                                              <span class="st">&#39;altAlleleCount&#39;</span>, <span class="st">&#39;altAlleleFreq&#39;</span>)]</span></code></pre></div>
</div>
</div>
<div id="aggregated-stats" class="section level1">
<h1>Aggregated stats</h1>
<div id="number-of-lof-variants-per-gene" class="section level2">
<h2>Number of LoF variants per gene</h2>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="do">## Aggregate by number of LoF variants per gene</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>variants_agg <span class="ot">&lt;-</span> <span class="fu">variantClient</span>(<span class="at">OpencgaR =</span> con, <span class="at">endpointName =</span> <span class="st">&quot;aggregationStats&quot;</span>, </span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>                              <span class="at">params=</span><span class="fu">list</span>(<span class="at">study=</span>study, <span class="at">ct=</span><span class="st">&#39;lof&#39;</span>, <span class="at">fields=</span><span class="st">&quot;genes&quot;</span>))</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">getResponseResults</span>(variants_agg)[[<span class="dv">1</span>]]<span class="sc">$</span>buckets[[<span class="dv">1</span>]]</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="fu">barplot</span>(<span class="at">height =</span> df<span class="sc">$</span>count, <span class="at">names.arg =</span> df<span class="sc">$</span>value, <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.names =</span> <span class="fl">0.6</span>)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
