{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pyOpenCGA Basic User Usage [UNDER CONSTRUCTION]\n",
    "\n",
    "------\n",
    "\n",
    "\n",
    "**[NOTE]** The server methods used by pyopencga client are defined in the following swagger URL:\n",
    "- http://bioinfodev.hpc.cam.ac.uk/opencga-test/webservices\n",
    "\n",
    "**[NOTE]** Current implemented methods are registered at the following spreadsheet:\n",
    "- https://docs.google.com/spreadsheets/d/1QpU9yl3UTneqwRqFX_WAqCiCfZBk5eU-4E3K-WVvuoc/edit?usp=sharing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading pyOpenCGA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current_dir: ..../notebooks\n",
      "base_modules_dir: ..../pyOpenCGA\n"
     ]
    }
   ],
   "source": [
    "# Initialize PYTHONPATH for pyopencga\n",
    "import sys\n",
    "import os\n",
    "\n",
    "cwd = os.getcwd()\n",
    "print(\"current_dir: ....\"+cwd[-10:])\n",
    "\n",
    "base_modules_dir = os.path.dirname(cwd)\n",
    "print(\"base_modules_dir: ....\"+base_modules_dir[-10:])\n",
    "\n",
    "sys.path.append(base_modules_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyopencga.opencga_config import ConfigClient\n",
    "from pyopencga.opencga_client import OpenCGAClient\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting credentials for LogIn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Credentials** \n",
    "\n",
    "Plese add the credentials for opencga login into a file in json format and read them from there.\n",
    "\n",
    "i.e:\n",
    "file: \\__user_config.json\n",
    "flie_content: {\"user\":\"xxx\",\"pwd\":\"yyy\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User: caf***\n"
     ]
    }
   ],
   "source": [
    "## Reading user config/credentials to connect to server\n",
    "user_config_json = \"./__user_config.json\"\n",
    "with open(user_config_json,\"r\") as f:\n",
    "    user_credentials = json.loads(f.read())\n",
    "    \n",
    "print('User: {}***'.format(user_credentials[\"user\"][:3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "user = user_credentials[\"user\"]\n",
    "passwd = user_credentials[\"pwd\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating ConfigClient for server connection configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Config information:\n",
      " {'version': 'v1', 'rest': {'hosts': ['http://bioinfodev.hpc.cam.ac.uk/opencga-test']}}\n"
     ]
    }
   ],
   "source": [
    "## Creating ConfigClient\n",
    "host = 'http://bioinfodev.hpc.cam.ac.uk/opencga-test'\n",
    "cc = ConfigClient()\n",
    "config_dict = cc.get_basic_config_dict(host)\n",
    "print(\"Config information:\\n\",config_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LogIn with user credentials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "oc = OpenCGAClient(configuration=config_dict, \n",
    "                   user=user, \n",
    "                   pwd=passwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Session token:\n",
      "eyJhbGciOi...\n"
     ]
    }
   ],
   "source": [
    "## Getting the session id / token\n",
    "token = oc.session_id\n",
    "print(\"Session token:\\n{}...\".format(token[:10]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "oc = OpenCGAClient(configuration=config_dict, \n",
    "                   session_id=token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'tmp:HG00096.chrom20.small.bam',\n",
       "   'annotationSets': [],\n",
       "   'name': 'HG00096.chrom20.small.bam',\n",
       "   'uuid': 'TM9SrwFoAAMAAR5FsdmSkQ',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAM',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///tmp/HG00096.chrom20.small.bam',\n",
       "   'path': 'tmp/HG00096.chrom20.small.bam',\n",
       "   'release': 1,\n",
       "   'creationDate': '20190114144122',\n",
       "   'modificationDate': '20190306123127',\n",
       "   'description': '',\n",
       "   'status': {'name': 'MISSING', 'date': '20190306123127', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 15397,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'HG00096',\n",
       "     'uuid': 'TM9TegFoAAQAARYSE1KsKQ',\n",
       "     'source': 'HG00096.chrom20.small.bam',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20190114144122',\n",
       "     'status': {'name': 'READY', 'date': '20190114144122', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {'globalStats': '{\"numMapped\":108,\"numUnmapped\":0,\"numPaired\":104,\"numMappedFirst\":53,\"numMappedSecond\":55,\"NM\":49,\"numHardC\":0,\"numSoftC\":35,\"numIn\":0,\"numDel\":0,\"numPad\":0,\"numSkip\":0,\"accMappingQuality\":6226,\"mappingQualityMap\":{\"48\":2,\"37\":10,\"60\":96},\"accInsert\":19175,\"insertMap\":{\"192\":4,\"193\":4,\"195\":4,\"196\":4,\"134\":2,\"198\":4,\"135\":4,\"199\":1,\"200\":4,\"137\":2,\"202\":4,\"138\":2,\"203\":2,\"205\":2,\"210\":2,\"211\":4,\"212\":2,\"218\":2,\"219\":2,\"221\":2,\"158\":2,\"162\":2,\"165\":2,\"167\":4,\"168\":2,\"170\":4,\"172\":2,\"173\":2,\"177\":2,\"178\":2,\"179\":4,\"180\":2,\"183\":2,\"184\":4,\"185\":3,\"186\":2,\"188\":2,\"189\":1,\"190\":3},\"seqStats\":{\"numSeqs\":108,\"numA\":3462,\"numT\":2260,\"numG\":2480,\"numC\":2587,\"numN\":11,\"minSeqLength\":100,\"maxSeqLength\":100,\"accSeqQual\":334978,\"lengthMap\":{\"100\":108},\"infoMap\":{\"0\":{\"numA\":32,\"numT\":24,\"numG\":22,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":1879},\"1\":{\"numA\":38,\"numT\":16,\"numG\":32,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":2995},\"2\":{\"numA\":29,\"numT\":20,\"numG\":40,\"numC\":19,\"numN\":0,\"numQual\":108,\"accQual\":2975},\"3\":{\"numA\":38,\"numT\":24,\"numG\":25,\"numC\":21,\"numN\":0,\"numQual\":108,\"accQual\":3009},\"4\":{\"numA\":34,\"numT\":25,\"numG\":24,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3075},\"5\":{\"numA\":31,\"numT\":19,\"numG\":28,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3098},\"6\":{\"numA\":41,\"numT\":20,\"numG\":23,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3155},\"7\":{\"numA\":29,\"numT\":24,\"numG\":21,\"numC\":34,\"numN\":0,\"numQual\":108,\"accQual\":3096},\"8\":{\"numA\":35,\"numT\":20,\"numG\":20,\"numC\":33,\"numN\":0,\"numQual\":108,\"accQual\":3178},\"9\":{\"numA\":35,\"numT\":20,\"numG\":29,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3198},\"10\":{\"numA\":39,\"numT\":19,\"numG\":24,\"numC\":26,\"numN\":0,\"numQual\":108,\"accQual\":3234},\"11\":{\"numA\":34,\"numT\":19,\"numG\":31,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3214},\"12\":{\"numA\":41,\"numT\":22,\"numG\":20,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3259},\"13\":{\"numA\":37,\"numT\":21,\"numG\":27,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3229},\"14\":{\"numA\":34,\"numT\":31,\"numG\":24,\"numC\":19,\"numN\":0,\"numQual\":108,\"accQual\":3237},\"15\":{\"numA\":31,\"numT\":29,\"numG\":23,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3292},\"16\":{\"numA\":33,\"numT\":21,\"numG\":31,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3292},\"17\":{\"numA\":39,\"numT\":27,\"numG\":24,\"numC\":18,\"numN\":0,\"numQual\":108,\"accQual\":3344},\"18\":{\"numA\":36,\"numT\":24,\"numG\":26,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3335},\"19\":{\"numA\":30,\"numT\":23,\"numG\":26,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3381},\"20\":{\"numA\":32,\"numT\":27,\"numG\":25,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3329},\"21\":{\"numA\":35,\"numT\":18,\"numG\":27,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3458},\"22\":{\"numA\":35,\"numT\":26,\"numG\":24,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3375},\"23\":{\"numA\":32,\"numT\":21,\"numG\":30,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3326},\"24\":{\"numA\":37,\"numT\":27,\"numG\":17,\"numC\":27,\"numN\":0,\"numQual\":108,\"accQual\":3412},\"25\":{\"numA\":46,\"numT\":26,\"numG\":19,\"numC\":17,\"numN\":0,\"numQual\":108,\"accQual\":3493},\"26\":{\"numA\":32,\"numT\":19,\"numG\":28,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3434},\"27\":{\"numA\":45,\"numT\":15,\"numG\":23,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3419},\"28\":{\"numA\":29,\"numT\":24,\"numG\":21,\"numC\":34,\"numN\":0,\"numQual\":108,\"accQual\":3378},\"29\":{\"numA\":38,\"numT\":26,\"numG\":18,\"numC\":26,\"numN\":0,\"numQual\":108,\"accQual\":3467},\"30\":{\"numA\":34,\"numT\":26,\"numG\":21,\"numC\":27,\"numN\":0,\"numQual\":108,\"accQual\":3393},\"31\":{\"numA\":35,\"numT\":23,\"numG\":28,\"numC\":21,\"numN\":1,\"numQual\":108,\"accQual\":3356},\"32\":{\"numA\":28,\"numT\":26,\"numG\":24,\"numC\":29,\"numN\":1,\"numQual\":108,\"accQual\":3367},\"33\":{\"numA\":46,\"numT\":13,\"numG\":24,\"numC\":24,\"numN\":1,\"numQual\":108,\"accQual\":3390},\"34\":{\"numA\":38,\"numT\":24,\"numG\":18,\"numC\":27,\"numN\":1,\"numQual\":108,\"accQual\":3390},\"35\":{\"numA\":40,\"numT\":24,\"numG\":21,\"numC\":22,\"numN\":1,\"numQual\":108,\"accQual\":3387},\"36\":{\"numA\":31,\"numT\":21,\"numG\":23,\"numC\":33,\"numN\":0,\"numQual\":108,\"accQual\":3483},\"37\":{\"numA\":37,\"numT\":21,\"numG\":27,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3402},\"38\":{\"numA\":32,\"numT\":29,\"numG\":29,\"numC\":18,\"numN\":0,\"numQual\":108,\"accQual\":3424},\"39\":{\"numA\":39,\"numT\":16,\"numG\":31,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3458},\"40\":{\"numA\":30,\"numT\":21,\"numG\":29,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3398},\"41\":{\"numA\":35,\"numT\":25,\"numG\":24,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3424},\"42\":{\"numA\":37,\"numT\":26,\"numG\":23,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3452},\"43\":{\"numA\":29,\"numT\":27,\"numG\":21,\"numC\":31,\"numN\":0,\"numQual\":108,\"accQual\":3466},\"44\":{\"numA\":33,\"numT\":29,\"numG\":27,\"numC\":19,\"numN\":0,\"numQual\":108,\"accQual\":3504},\"45\":{\"numA\":34,\"numT\":23,\"numG\":24,\"numC\":27,\"numN\":0,\"numQual\":108,\"accQual\":3441},\"46\":{\"numA\":41,\"numT\":23,\"numG\":16,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3495},\"47\":{\"numA\":40,\"numT\":15,\"numG\":20,\"numC\":32,\"numN\":1,\"numQual\":108,\"accQual\":3455},\"48\":{\"numA\":28,\"numT\":25,\"numG\":26,\"numC\":28,\"numN\":1,\"numQual\":108,\"accQual\":3433},\"49\":{\"numA\":37,\"numT\":26,\"numG\":17,\"numC\":27,\"numN\":1,\"numQual\":108,\"accQual\":3506},\"50\":{\"numA\":37,\"numT\":25,\"numG\":18,\"numC\":27,\"numN\":1,\"numQual\":108,\"accQual\":3481},\"51\":{\"numA\":35,\"numT\":26,\"numG\":25,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3503},\"52\":{\"numA\":34,\"numT\":24,\"numG\":33,\"numC\":17,\"numN\":0,\"numQual\":108,\"accQual\":3486},\"53\":{\"numA\":27,\"numT\":30,\"numG\":30,\"numC\":21,\"numN\":0,\"numQual\":108,\"accQual\":3541},\"54\":{\"numA\":31,\"numT\":19,\"numG\":31,\"numC\":27,\"numN\":0,\"numQual\":108,\"accQual\":3501},\"55\":{\"numA\":26,\"numT\":26,\"numG\":27,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3537},\"56\":{\"numA\":25,\"numT\":25,\"numG\":29,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3475},\"57\":{\"numA\":34,\"numT\":15,\"numG\":29,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3472},\"58\":{\"numA\":32,\"numT\":21,\"numG\":29,\"numC\":26,\"numN\":0,\"numQual\":108,\"accQual\":3507},\"59\":{\"numA\":25,\"numT\":25,\"numG\":28,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3497},\"60\":{\"numA\":41,\"numT\":17,\"numG\":25,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3470},\"61\":{\"numA\":28,\"numT\":18,\"numG\":31,\"numC\":30,\"numN\":1,\"numQual\":108,\"accQual\":3443},\"62\":{\"numA\":35,\"numT\":20,\"numG\":16,\"numC\":36,\"numN\":1,\"numQual\":108,\"accQual\":3460},\"63\":{\"numA\":32,\"numT\":28,\"numG\":23,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3468},\"64\":{\"numA\":32,\"numT\":23,\"numG\":23,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3462},\"65\":{\"numA\":35,\"numT\":21,\"numG\":27,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3519},\"66\":{\"numA\":35,\"numT\":15,\"numG\":32,\"numC\":26,\"numN\":0,\"numQual\":108,\"accQual\":3478},\"67\":{\"numA\":33,\"numT\":21,\"numG\":26,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3491},\"68\":{\"numA\":41,\"numT\":16,\"numG\":23,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3466},\"69\":{\"numA\":38,\"numT\":22,\"numG\":25,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3502},\"70\":{\"numA\":28,\"numT\":27,\"numG\":25,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3530},\"71\":{\"numA\":30,\"numT\":27,\"numG\":29,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3543},\"72\":{\"numA\":32,\"numT\":19,\"numG\":27,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3527},\"73\":{\"numA\":37,\"numT\":20,\"numG\":31,\"numC\":20,\"numN\":0,\"numQual\":108,\"accQual\":3525},\"74\":{\"numA\":25,\"numT\":27,\"numG\":31,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3489},\"75\":{\"numA\":34,\"numT\":23,\"numG\":26,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3508},\"76\":{\"numA\":38,\"numT\":24,\"numG\":24,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3500},\"77\":{\"numA\":39,\"numT\":22,\"numG\":22,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3456},\"78\":{\"numA\":37,\"numT\":19,\"numG\":22,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3421},\"79\":{\"numA\":32,\"numT\":27,\"numG\":21,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3437},\"80\":{\"numA\":38,\"numT\":20,\"numG\":26,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3503},\"81\":{\"numA\":38,\"numT\":20,\"numG\":29,\"numC\":21,\"numN\":0,\"numQual\":108,\"accQual\":3505},\"82\":{\"numA\":37,\"numT\":19,\"numG\":22,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3494},\"83\":{\"numA\":41,\"numT\":16,\"numG\":27,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3446},\"84\":{\"numA\":36,\"numT\":20,\"numG\":24,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3464},\"85\":{\"numA\":37,\"numT\":28,\"numG\":21,\"numC\":22,\"numN\":0,\"numQual\":108,\"accQual\":3449},\"86\":{\"numA\":30,\"numT\":27,\"numG\":21,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3439},\"87\":{\"numA\":39,\"numT\":25,\"numG\":18,\"numC\":26,\"numN\":0,\"numQual\":108,\"accQual\":3434},\"88\":{\"numA\":31,\"numT\":22,\"numG\":27,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":3365},\"89\":{\"numA\":41,\"numT\":22,\"numG\":16,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3355},\"90\":{\"numA\":36,\"numT\":18,\"numG\":24,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3379},\"91\":{\"numA\":35,\"numT\":26,\"numG\":22,\"numC\":25,\"numN\":0,\"numQual\":108,\"accQual\":3297},\"92\":{\"numA\":38,\"numT\":29,\"numG\":18,\"numC\":23,\"numN\":0,\"numQual\":108,\"accQual\":3254},\"93\":{\"numA\":35,\"numT\":21,\"numG\":25,\"numC\":27,\"numN\":0,\"numQual\":108,\"accQual\":3254},\"94\":{\"numA\":34,\"numT\":20,\"numG\":30,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":3208},\"95\":{\"numA\":27,\"numT\":26,\"numG\":25,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3107},\"96\":{\"numA\":32,\"numT\":20,\"numG\":27,\"numC\":29,\"numN\":0,\"numQual\":108,\"accQual\":3124},\"97\":{\"numA\":34,\"numT\":23,\"numG\":21,\"numC\":30,\"numN\":0,\"numQual\":108,\"accQual\":3094},\"98\":{\"numA\":34,\"numT\":30,\"numG\":16,\"numC\":28,\"numN\":0,\"numQual\":108,\"accQual\":2992},\"99\":{\"numA\":40,\"numT\":16,\"numG\":28,\"numC\":24,\"numN\":0,\"numQual\":108,\"accQual\":1930}},\"kmers\":{\"kvalue\":0,\"kmersMap\":{}}},\"nm\":49}'},\n",
       "   'attributes': {'alignmentHeader': {'attributes': {},\n",
       "     'sequenceDiccionary': [{'sequenceName': '1',\n",
       "       'sequenceLength': 249250621,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1b22b98cdeb4a9304cb5d48026a85128',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '2',\n",
       "       'sequenceLength': 243199373,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'a0d9851da00400dec1098a9255ac712e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '3',\n",
       "       'sequenceLength': 198022430,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'fdfd811849cc2fadebc929bb925902e5',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '4',\n",
       "       'sequenceLength': 191154276,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '23dccd106897542ad87d2765d28a19a1',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '5',\n",
       "       'sequenceLength': 180915260,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '0740173db9ffd264d728f32784845cd7',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '6',\n",
       "       'sequenceLength': 171115067,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1d3a93a248d92a729ee764823acbbc6b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '7',\n",
       "       'sequenceLength': 159138663,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '618366e953d6aaad97dbe4777c29375e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '8',\n",
       "       'sequenceLength': 146364022,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '96f514a9929e410c6651697bded59aec',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '9',\n",
       "       'sequenceLength': 141213431,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '3e273117f15e0a400f01055d9f393768',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '10',\n",
       "       'sequenceLength': 135534747,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '988c28e000e84c26d552359af1ea2e1d',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '11',\n",
       "       'sequenceLength': 135006516,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '98c59049a2df285c76ffb1c6db8f8b96',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '12',\n",
       "       'sequenceLength': 133851895,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '51851ac0e1a115847ad36449b0015864',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '13',\n",
       "       'sequenceLength': 115169878,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '283f8d7892baa81b510a015719ca7b0b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '14',\n",
       "       'sequenceLength': 107349540,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '98f3cae32b2a2e9524bc19813927542e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '15',\n",
       "       'sequenceLength': 102531392,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'e5645a794a8238215b2cd77acb95a078',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '16',\n",
       "       'sequenceLength': 90354753,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'fc9b1a7b42b97a864f56b348b06095e6',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '17',\n",
       "       'sequenceLength': 81195210,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '351f64d4f4f9ddd45b35336ad97aa6de',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '18',\n",
       "       'sequenceLength': 78077248,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'b15d4b2d29dde9d3e4f93d1d0f2cbc9c',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '19',\n",
       "       'sequenceLength': 59128983,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1aacd71f30db8e561810913e0b72636d',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '20',\n",
       "       'sequenceLength': 63025520,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '0dec9660ec1efaaf33281c0d5ea2560f',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '21',\n",
       "       'sequenceLength': 48129895,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '2979a6085bfe28e3ad6f552f361ed74d',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': '22',\n",
       "       'sequenceLength': 51304566,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'a718acaa6135fdca8357d5bfe94211dd',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'X',\n",
       "       'sequenceLength': 155270560,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '7e0e2e580297b7764e31dbc80c2540dd',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'Y',\n",
       "       'sequenceLength': 59373566,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1fa3474750af0948bdf97d5a0ee52e51',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'MT',\n",
       "       'sequenceLength': 16569,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'c68f52674c9fb33aef52dcf399755519',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000207.1',\n",
       "       'sequenceLength': 4262,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'f3814841f1939d3ca19072d9e89f3fd7',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000226.1',\n",
       "       'sequenceLength': 15008,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1c1b2cd1fccbc0a99b6a447fa24d1504',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000229.1',\n",
       "       'sequenceLength': 19913,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'd0f40ec87de311d8e715b52e4c7062e1',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000231.1',\n",
       "       'sequenceLength': 27386,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'ba8882ce3a1efa2080e5d29b956568a4',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000210.1',\n",
       "       'sequenceLength': 27682,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '851106a74238044126131ce2a8e5847c',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000239.1',\n",
       "       'sequenceLength': 33824,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '99795f15702caec4fa1c4e15f8a29c07',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000235.1',\n",
       "       'sequenceLength': 34474,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '118a25ca210cfbcdfb6c2ebb249f9680',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000201.1',\n",
       "       'sequenceLength': 36148,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'dfb7e7ec60ffdcb85cb359ea28454ee9',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000247.1',\n",
       "       'sequenceLength': 36422,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '7de00226bb7df1c57276ca6baabafd15',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000245.1',\n",
       "       'sequenceLength': 36651,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '89bc61960f37d94abf0df2d481ada0ec',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000197.1',\n",
       "       'sequenceLength': 37175,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '6f5efdd36643a9b8c8ccad6f2f1edc7b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000203.1',\n",
       "       'sequenceLength': 37498,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '96358c325fe0e70bee73436e8bb14dbd',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000246.1',\n",
       "       'sequenceLength': 38154,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'e4afcd31912af9d9c2546acf1cb23af2',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000249.1',\n",
       "       'sequenceLength': 38502,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1d78abec37c15fe29a275eb08d5af236',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000196.1',\n",
       "       'sequenceLength': 38914,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'd92206d1bb4c3b4019c43c0875c06dc0',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000248.1',\n",
       "       'sequenceLength': 39786,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '5a8e43bec9be36c7b49c84d585107776',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000244.1',\n",
       "       'sequenceLength': 39929,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '0996b4475f353ca98bacb756ac479140',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000238.1',\n",
       "       'sequenceLength': 39939,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '131b1efc3270cc838686b54e7c34b17b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000202.1',\n",
       "       'sequenceLength': 40103,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '06cbf126247d89664a4faebad130fe9c',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000234.1',\n",
       "       'sequenceLength': 40531,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '93f998536b61a56fd0ff47322a911d4b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000232.1',\n",
       "       'sequenceLength': 40652,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '3e06b6741061ad93a8587531307057d8',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000206.1',\n",
       "       'sequenceLength': 41001,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '43f69e423533e948bfae5ce1d45bd3f1',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000240.1',\n",
       "       'sequenceLength': 41933,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '445a86173da9f237d7bcf41c6cb8cc62',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000236.1',\n",
       "       'sequenceLength': 41934,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'fdcd739913efa1fdc64b6c0cd7016779',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000241.1',\n",
       "       'sequenceLength': 42152,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'ef4258cdc5a45c206cea8fc3e1d858cf',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000243.1',\n",
       "       'sequenceLength': 43341,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'cc34279a7e353136741c9fce79bc4396',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000242.1',\n",
       "       'sequenceLength': 43523,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '2f8694fc47576bc81b5fe9e7de0ba49e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000230.1',\n",
       "       'sequenceLength': 43691,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'b4eb71ee878d3706246b7c1dbef69299',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000237.1',\n",
       "       'sequenceLength': 45867,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'e0c82e7751df73f4f6d0ed30cdc853c0',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000233.1',\n",
       "       'sequenceLength': 45941,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '7fed60298a8d62ff808b74b6ce820001',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000204.1',\n",
       "       'sequenceLength': 81310,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'efc49c871536fa8d79cb0a06fa739722',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000198.1',\n",
       "       'sequenceLength': 90085,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '868e7784040da90d900d2d1b667a1383',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000208.1',\n",
       "       'sequenceLength': 92689,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'aa81be49bf3fe63a79bdc6a6f279abf6',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000191.1',\n",
       "       'sequenceLength': 106433,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'd75b436f50a8214ee9c2a51d30b2c2cc',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000227.1',\n",
       "       'sequenceLength': 128374,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'a4aead23f8053f2655e468bcc6ecdceb',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000228.1',\n",
       "       'sequenceLength': 129120,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'c5a17c97e2c1a0b6a9cc5a6b064b714f',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000214.1',\n",
       "       'sequenceLength': 137718,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '46c2032c37f2ed899eb41c0473319a69',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000221.1',\n",
       "       'sequenceLength': 155397,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '3238fb74ea87ae857f9c7508d315babb',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000209.1',\n",
       "       'sequenceLength': 159169,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'f40598e2a5a6b26e84a3775e0d1e2c81',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000218.1',\n",
       "       'sequenceLength': 161147,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '1d708b54644c26c7e01c2dad5426d38c',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000220.1',\n",
       "       'sequenceLength': 161802,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'fc35de963c57bf7648429e6454f1c9db',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000213.1',\n",
       "       'sequenceLength': 164239,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '9d424fdcc98866650b58f004080a992a',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000211.1',\n",
       "       'sequenceLength': 166566,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '7daaa45c66b288847b9b32b964e623d3',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000199.1',\n",
       "       'sequenceLength': 169874,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '569af3b73522fab4b40995ae4944e78e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000217.1',\n",
       "       'sequenceLength': 172149,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '6d243e18dea1945fb7f2517615b8f52e',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000216.1',\n",
       "       'sequenceLength': 172294,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '642a232d91c486ac339263820aef7fe0',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000215.1',\n",
       "       'sequenceLength': 172545,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '5eb3b418480ae67a997957c909375a73',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000205.1',\n",
       "       'sequenceLength': 174588,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'd22441398d99caf673e9afb9a1908ec5',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000219.1',\n",
       "       'sequenceLength': 179198,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'f977edd13bac459cb2ed4a5457dba1b3',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000224.1',\n",
       "       'sequenceLength': 179693,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'd5b2fc04f6b41b212a4198a07f450e20',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000223.1',\n",
       "       'sequenceLength': 180455,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '399dfa03bf32022ab52a846f7ca35b30',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000195.1',\n",
       "       'sequenceLength': 182896,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '5d9ec007868d517e73543b005ba48535',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000212.1',\n",
       "       'sequenceLength': 186858,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '563531689f3dbd691331fd6c5730a88b',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000222.1',\n",
       "       'sequenceLength': 186861,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '6fe9abac455169f50470f5a6b01d0f59',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000200.1',\n",
       "       'sequenceLength': 187035,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '75e4c8d17cd4addf3917d1703cacaf25',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000193.1',\n",
       "       'sequenceLength': 189789,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': 'dbb6e8ece0b5de29da56601613007c2a',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000194.1',\n",
       "       'sequenceLength': 191469,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '6ac8f815bf8e845bb3031b73f812c012',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000225.1',\n",
       "       'sequenceLength': 211173,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '63945c3e6962f28ffd469719a747e73c',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}},\n",
       "      {'sequenceName': 'GL000192.1',\n",
       "       'sequenceLength': 547496,\n",
       "       'attributes': {'AS': 'NCBI37',\n",
       "        'M5': '325ba9e808f669dfeee210fdd7b470ac',\n",
       "        'UR': 'ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz',\n",
       "        'SP': 'Human'}}],\n",
       "     'readGroups': [{'id': 'SRR062634',\n",
       "       'attributes': {'LB': '2845856850',\n",
       "        'PI': '206',\n",
       "        'SM': 'HG00096',\n",
       "        'CN': 'WUGSC',\n",
       "        'PL': 'ILLUMINA',\n",
       "        'DS': 'SRP001294'}},\n",
       "      {'id': 'SRR062635',\n",
       "       'attributes': {'LB': '2845856850',\n",
       "        'PI': '206',\n",
       "        'SM': 'HG00096',\n",
       "        'CN': 'WUGSC',\n",
       "        'PL': 'ILLUMINA',\n",
       "        'DS': 'SRP001294'}},\n",
       "      {'id': 'SRR062641',\n",
       "       'attributes': {'LB': '2845856850',\n",
       "        'PI': '206',\n",
       "        'SM': 'HG00096',\n",
       "        'CN': 'WUGSC',\n",
       "        'PL': 'ILLUMINA',\n",
       "        'DS': 'SRP001294'}}],\n",
       "     'programRecords': [{'id': 'GATK IndelRealigner',\n",
       "       'attributes': {'VN': '1.0.4487',\n",
       "        'CL': 'SNPsFileForDebugging=null targetIntervals=/lustre/scratch105/projects/g1k/ref/broad_recal_data/pilot_data/indel.dbsnp_129_b37-vs-pilot.intervals output=null useOnlyKnownIndels=true maxReadsForConsensuses=120 maxConsensuses=30 entropyThreshold=0.15 indelsFileForDebugging=null noOriginalAlignmentTags=false realignReadsWithBadMates=false maxReadsForRealignment=20000 noPGTag=false LODThresholdForCleaning=0.4 maxReadsInRam=500000 targetIntervalsAreNotSorted=false sortInCoordinateOrderEvenThoughItIsHighlyUnsafe=false statisticsFileForDebugging=null'}},\n",
       "      {'id': 'GATK TableRecalibration',\n",
       "       'attributes': {'VN': '1.0.4702',\n",
       "        'CL': 'default_read_group=RG default_platform=ILLUMINA force_read_group=null force_platform=null window_size_nqs=5 homopolymer_nback=7 exception_if_no_tile=false solid_recal_mode=SET_Q_ZERO solid_nocall_strategy=THROW_EXCEPTION output_bam=null preserve_qscores_less_than=5 smoothing=1 max_quality_score=40 no_pg_tag=false fail_with_no_eof_marker=true skipUQUpdate=false Covariates=[ReadGroupCovariate, QualityScoreCovariate, CycleCovariate, DinucCovariate] pQ=5 maxQ=40 smoothing=1'}},\n",
       "      {'id': 'GATK TableRecalibration.1',\n",
       "       'attributes': {'VN': '1.0.4487',\n",
       "        'CL': 'output_bam=null window_size_nqs=5 force_read_group=null smoothing=1 default_platform=ILLUMINA exception_if_no_tile=false homopolymer_nback=7 no_pg_tag=false skipUQUpdate=false default_read_group=RG max_quality_score=40 fail_with_no_eof_marker=true solid_recal_mode=SET_Q_ZERO solid_nocall_strategy=THROW_EXCEPTION force_platform=null preserve_qscores_less_than=5 Covariates=[ReadGroupCovariate, QualityScoreCovariate, CycleCovariate, DinucCovariate] pQ=5 maxQ=40 smoothing=1'}},\n",
       "      {'id': 'bwa', 'attributes': {'VN': '0.5.5'}}],\n",
       "     'comments': [],\n",
       "     'studyName': 'Platinum',\n",
       "     'origin': 'SAM'}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "   'uuid': '7RrmjQFlAAMAAUQhWCKtIg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143447',\n",
       "   'modificationDate': '20180919113915',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143447', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 887890738,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12877',\n",
       "     'uuid': '7RrnvgFlAAQAAfde27jGkg',\n",
       "     'source': 'platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143448',\n",
       "     'status': {'name': 'READY', 'date': '20180918143448', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180918155042',\n",
       "     'message': 'Start indexing file'},\n",
       "    'jobId': 0,\n",
       "    'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.8337 <= x < -0.5214',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -48.5831',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -48.5831 <= x < -1.8337',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '169.26'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '158.94'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '159.84'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '136.65'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '136.44'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '131.73'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '157.29'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '156.09'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '161.58'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '152.31'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '156.6'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '152.1'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '134.25'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '106.38'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '165.3'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '163.65'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '164.73'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '163.47'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '164.55'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '162.45'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '163.65'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '140.73'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '42108.3'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '82.53'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '59.97'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/NA12877_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12877/Data/Intensities/BaseCalls/Alignment/NA12877_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '8',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12877']}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "   'uuid': '7RrqeAFlAAMAAaQJhgj7Yg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143448',\n",
       "   'modificationDate': '20180920010648',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143448', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 883195909,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12878',\n",
       "     'uuid': '7RrrAgFlAAQAAU3OyOO-qA',\n",
       "     'source': 'platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143448',\n",
       "     'status': {'name': 'READY', 'date': '20180918143448', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010648',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 50768118,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5156262,\n",
       "      'INDEL': 699379,\n",
       "      'NO_VARIATION': 44912475,\n",
       "      'DELETION': 2},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 44214457,\n",
       "     'chromosomeStats': {'22': {'count': 790254, 'density': 0.01540319},\n",
       "      'X': {'count': 2392878, 'density': 0.015411022},\n",
       "      'Y': {'count': 208414, 'density': 0.003510215},\n",
       "      '10': {'count': 2388191, 'density': 0.017620506},\n",
       "      '11': {'count': 2302533, 'density': 0.017054977},\n",
       "      '12': {'count': 2268969, 'density': 0.01695134},\n",
       "      '13': {'count': 1635874, 'density': 0.0142040085},\n",
       "      '14': {'count': 1560798, 'density': 0.014539401},\n",
       "      '15': {'count': 1530631, 'density': 0.014928414},\n",
       "      '16': {'count': 1652318, 'density': 0.018287007},\n",
       "      '17': {'count': 1545120, 'density': 0.019029694},\n",
       "      '18': {'count': 1292896, 'density': 0.016559191},\n",
       "      'MT': {'count': 195, 'density': 0.011767546},\n",
       "      '19': {'count': 1274329, 'density': 0.021551682},\n",
       "      '1': {'count': 4046629, 'density': 0.01623518},\n",
       "      '2': {'count': 4124323, 'density': 0.01695861},\n",
       "      '3': {'count': 3216700, 'density': 0.016244119},\n",
       "      '4': {'count': 3230972, 'density': 0.016902432},\n",
       "      '5': {'count': 2972936, 'density': 0.016432755},\n",
       "      '6': {'count': 2851440, 'density': 0.016663874},\n",
       "      '7': {'count': 2881425, 'density': 0.01810638},\n",
       "      '8': {'count': 2450227, 'density': 0.016740639},\n",
       "      '9': {'count': 2342886, 'density': 0.0165911},\n",
       "      '20': {'count': 1092384, 'density': 0.017332407},\n",
       "      '21': {'count': 714796, 'density': 0.014851393}},\n",
       "     'tiTvRatio': 1.6145304,\n",
       "     'meanQuality': 787.9977,\n",
       "     'stdDevQuality': 797.48773,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.8667 <= x < -0.5971',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -110.8111',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -110.8111 <= x < -1.8667',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '163.71'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '151.44'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '152.01'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '128.88'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '128.73'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '123.54'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '152.97'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '150.36'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '152.97'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '147.24'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '149.37'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '146.76'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '127.47'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '103.2'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '157.56'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '155.01'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '154.92'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '154.65'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '155.97'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '154.8'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '154.44'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '136.53'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '60475.3'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '150.15'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '9.87'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/NA12878_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12878/Data/Intensities/BaseCalls/Alignment/NA12878_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '10',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12878']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12878_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 181514,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782162,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "   'uuid': '7RrsjgFlAAMAATJ0obPliA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143449',\n",
       "   'modificationDate': '20180920010648',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143449', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 889974818,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12879',\n",
       "     'uuid': '7Rrs-wFlAAQAAQMADI03VQ',\n",
       "     'source': 'platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143449',\n",
       "     'status': {'name': 'READY', 'date': '20180918143449', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010648',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 51037071,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5213124,\n",
       "      'INDEL': 679903,\n",
       "      'NO_VARIATION': 45144044},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 44189143,\n",
       "     'chromosomeStats': {'22': {'count': 780946, 'density': 0.015221763},\n",
       "      'X': {'count': 2446707, 'density': 0.0157577},\n",
       "      'Y': {'count': 199749, 'density': 0.0033642747},\n",
       "      '10': {'count': 2410452, 'density': 0.017784752},\n",
       "      '11': {'count': 2322302, 'density': 0.017201407},\n",
       "      '12': {'count': 2279314, 'density': 0.017028626},\n",
       "      '13': {'count': 1656495, 'density': 0.014383057},\n",
       "      '14': {'count': 1564904, 'density': 0.01457765},\n",
       "      '15': {'count': 1539912, 'density': 0.015018932},\n",
       "      '16': {'count': 1669915, 'density': 0.018481761},\n",
       "      '17': {'count': 1565825, 'density': 0.019284697},\n",
       "      '18': {'count': 1296568, 'density': 0.016606221},\n",
       "      'MT': {'count': 204, 'density': 0.012310663},\n",
       "      '19': {'count': 1296130, 'density': 0.021920383},\n",
       "      '1': {'count': 4062378, 'density': 0.016298367},\n",
       "      '2': {'count': 4122170, 'density': 0.016949756},\n",
       "      '3': {'count': 3220144, 'density': 0.01626151},\n",
       "      '4': {'count': 3253464, 'density': 0.017020095},\n",
       "      '5': {'count': 2963604, 'density': 0.016381172},\n",
       "      '6': {'count': 2841813, 'density': 0.016607614},\n",
       "      '7': {'count': 2908147, 'density': 0.018274296},\n",
       "      '8': {'count': 2469339, 'density': 0.016871216},\n",
       "      '9': {'count': 2347783, 'density': 0.016625777},\n",
       "      '20': {'count': 1105930, 'density': 0.017547335},\n",
       "      '21': {'count': 712876, 'density': 0.014811501}},\n",
       "     'tiTvRatio': 1.5819521,\n",
       "     'meanQuality': 784.0122,\n",
       "     'stdDevQuality': 813.6806,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.7936 <= x < -0.6005',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -125.5373',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -125.5373 <= x < -1.7936',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '166.95'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '155.25'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '156.81'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '131.07'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '134.01'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '128.97'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '159.06'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '155.67'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '155.94'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '150.75'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '154.14'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '150.3'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '130.26'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '107.04'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '162'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '159.81'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '157.98'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '158.97'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '161.01'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '159.06'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '158.91'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '137.49'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '51996.6'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '151.05'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '9.57'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/NA12879_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12879/Data/Intensities/BaseCalls/Alignment/NA12879_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '12',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12879']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12879_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 175169,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782117,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "   'uuid': '7RruXAFlAAMAAXOvEUUaxg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143449',\n",
       "   'modificationDate': '20180920010648',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143449', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 899309868,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12880',\n",
       "     'uuid': '7RruywFlAAQAAR0dSJSDqw',\n",
       "     'source': 'platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143449',\n",
       "     'status': {'name': 'READY', 'date': '20180918143449', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010648',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 51750087,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5231334,\n",
       "      'INDEL': 678243,\n",
       "      'NO_VARIATION': 45840509,\n",
       "      'DELETION': 1},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 44785754,\n",
       "     'chromosomeStats': {'22': {'count': 826953, 'density': 0.016118506},\n",
       "      'X': {'count': 2408445, 'density': 0.015511279},\n",
       "      'Y': {'count': 210221, 'density': 0.0035406495},\n",
       "      '10': {'count': 2438354, 'density': 0.017990619},\n",
       "      '11': {'count': 2351581, 'density': 0.017418278},\n",
       "      '12': {'count': 2316393, 'density': 0.017305642},\n",
       "      '13': {'count': 1639910, 'density': 0.014239053},\n",
       "      '14': {'count': 1581645, 'density': 0.014733599},\n",
       "      '15': {'count': 1580696, 'density': 0.015416703},\n",
       "      '16': {'count': 1714522, 'density': 0.01897545},\n",
       "      '17': {'count': 1635213, 'density': 0.020139279},\n",
       "      '18': {'count': 1300208, 'density': 0.016652841},\n",
       "      'MT': {'count': 258, 'density': 0.015569368},\n",
       "      '19': {'count': 1364069, 'density': 0.02306938},\n",
       "      '1': {'count': 4152665, 'density': 0.0166606},\n",
       "      '2': {'count': 4171720, 'density': 0.017153498},\n",
       "      '3': {'count': 3253176, 'density': 0.01642832},\n",
       "      '4': {'count': 3237338, 'density': 0.016935734},\n",
       "      '5': {'count': 2999812, 'density': 0.01658131},\n",
       "      '6': {'count': 2882168, 'density': 0.01684345},\n",
       "      '7': {'count': 2945065, 'density': 0.018506283},\n",
       "      '8': {'count': 2485710, 'density': 0.016983068},\n",
       "      '9': {'count': 2393787, 'density': 0.016951554},\n",
       "      '20': {'count': 1136312, 'density': 0.018029395},\n",
       "      '21': {'count': 723866, 'density': 0.015039842}},\n",
       "     'tiTvRatio': 1.5723941,\n",
       "     'meanQuality': 784.91266,\n",
       "     'stdDevQuality': 813.3507,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.9271 <= x < -0.6207',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -64.9164',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -64.9164 <= x < -1.9271',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '166.89'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '155.85'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '157.02'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '134.22'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '133.08'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '127.83'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '150.81'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '152.52'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '158.46'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '147.81'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '153.03'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '149.55'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '130.14'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '104.19'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '162.87'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '160.95'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '161.16'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '160.02'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '161.94'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '159.21'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '159.84'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '141.06'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '71547.7'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '156.63'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '9.9'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/NA12880_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.10/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12880/Data/Intensities/BaseCalls/Alignment/NA12880_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '14',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12880']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12880_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 183451,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782114,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "   'uuid': '7RrwMQFlAAMAASet-PHSjg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143450',\n",
       "   'modificationDate': '20180920010648',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143450', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 918334187,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12881',\n",
       "     'uuid': '7RrwrAFlAAQAAZV9nvZABw',\n",
       "     'source': 'platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143450',\n",
       "     'status': {'name': 'READY', 'date': '20180918143450', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010648',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 53339756,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5252726,\n",
       "      'INDEL': 675088,\n",
       "      'NO_VARIATION': 47411940,\n",
       "      'DELETION': 2},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 46237899,\n",
       "     'chromosomeStats': {'22': {'count': 830111, 'density': 0.01618006},\n",
       "      'X': {'count': 2532242, 'density': 0.016308578},\n",
       "      'Y': {'count': 207284, 'density': 0.003491183},\n",
       "      '10': {'count': 2509500, 'density': 0.018515546},\n",
       "      '11': {'count': 2428946, 'density': 0.017991325},\n",
       "      '12': {'count': 2386806, 'density': 0.017831694},\n",
       "      '13': {'count': 1713718, 'density': 0.014879915},\n",
       "      '14': {'count': 1659432, 'density': 0.015458213},\n",
       "      '15': {'count': 1620791, 'density': 0.015807753},\n",
       "      '16': {'count': 1742757, 'density': 0.01928794},\n",
       "      '17': {'count': 1659438, 'density': 0.020437635},\n",
       "      '18': {'count': 1352532, 'density': 0.017322998},\n",
       "      'MT': {'count': 220, 'density': 0.013276205},\n",
       "      '19': {'count': 1378381, 'density': 0.023311427},\n",
       "      '1': {'count': 4233251, 'density': 0.016983913},\n",
       "      '2': {'count': 4309169, 'density': 0.017718667},\n",
       "      '3': {'count': 3357864, 'density': 0.016956989},\n",
       "      '4': {'count': 3365867, 'density': 0.017608117},\n",
       "      '5': {'count': 3111829, 'density': 0.017200477},\n",
       "      '6': {'count': 2965606, 'density': 0.017331062},\n",
       "      '7': {'count': 3050188, 'density': 0.019166857},\n",
       "      '8': {'count': 2568468, 'density': 0.017548494},\n",
       "      '9': {'count': 2453548, 'density': 0.01737475},\n",
       "      '20': {'count': 1156822, 'density': 0.018354818},\n",
       "      '21': {'count': 744986, 'density': 0.015478654}},\n",
       "     'tiTvRatio': 1.5674541,\n",
       "     'meanQuality': 750.5528,\n",
       "     'stdDevQuality': 785.29114,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.0751 <= x < -0.6875',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -284.5331',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -284.5331 <= x < -2.0751',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '161.55'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '149.52'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '151.02'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '126.81'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '128.61'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '124.11'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '153.54'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '149.79'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '150.78'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '146.4'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '148.68'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '145.32'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '125.52'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '104.04'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '155.79'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '153.84'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '152.37'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '153.12'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '154.92'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '152.76'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '152.7'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '135.48'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '42991.5'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '147.3'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '9.39'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/NA12881_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12881/Data/Intensities/BaseCalls/Alignment/NA12881_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '16',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12881']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12881_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 184193,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782118,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rrx9QFlAAMAAaNgVvWTeA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143450',\n",
       "   'modificationDate': '20180920010648',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143450', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 879641456,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12882',\n",
       "     'uuid': '7RrySwFlAAQAARpvXnR9gg',\n",
       "     'source': 'platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143450',\n",
       "     'status': {'name': 'READY', 'date': '20180918143450', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010648',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 51234362,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5085097,\n",
       "      'INDEL': 711555,\n",
       "      'NO_VARIATION': 45437710},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 44728865,\n",
       "     'chromosomeStats': {'22': {'count': 718628, 'density': 0.014007096},\n",
       "      'X': {'count': 4558018, 'density': 0.029355327},\n",
       "      'Y': {'count': 661292, 'density': 0.011137818},\n",
       "      '10': {'count': 2291783, 'density': 0.016909191},\n",
       "      '11': {'count': 2202277, 'density': 0.016312376},\n",
       "      '12': {'count': 2170124, 'density': 0.016212875},\n",
       "      '13': {'count': 1583527, 'density': 0.013749489},\n",
       "      '14': {'count': 1484222, 'density': 0.0138260685},\n",
       "      '15': {'count': 1469124, 'density': 0.014328529},\n",
       "      '16': {'count': 1564207, 'density': 0.01731184},\n",
       "      '17': {'count': 1449659, 'density': 0.017853998},\n",
       "      '18': {'count': 1241967, 'density': 0.0159069},\n",
       "      'MT': {'count': 237, 'density': 0.0143020945},\n",
       "      '19': {'count': 1165116, 'density': 0.019704651},\n",
       "      '1': {'count': 3858777, 'density': 0.015481514},\n",
       "      '2': {'count': 3944408, 'density': 0.016218824},\n",
       "      '3': {'count': 3088624, 'density': 0.015597344},\n",
       "      '4': {'count': 3104905, 'density': 0.016242927},\n",
       "      '5': {'count': 2847219, 'density': 0.01573786},\n",
       "      '6': {'count': 2730570, 'density': 0.015957506},\n",
       "      '7': {'count': 2764462, 'density': 0.017371405},\n",
       "      '8': {'count': 2353372, 'density': 0.016078897},\n",
       "      '9': {'count': 2251247, 'density': 0.01594216},\n",
       "      '20': {'count': 1041945, 'density': 0.016532112},\n",
       "      '21': {'count': 688652, 'density': 0.0143081965}},\n",
       "     'tiTvRatio': 1.6298203,\n",
       "     'meanQuality': 836.4162,\n",
       "     'stdDevQuality': 818.36304,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.7838 <= x < -0.58',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -89.274',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -89.274 <= x < -1.7838',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '168.57'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '156.75'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '157.92'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '132.24'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '133.29'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '130.56'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '156.57'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '157.95'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '157.5'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '155.7'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '156.06'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '151.92'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '130.53'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '108.54'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '163.17'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '160.95'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '159.6'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '159.42'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '161.76'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '159.87'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '160.32'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '141.24'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '48215.9'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '79.95'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '57.51'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/NA12882_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12882/Data/Intensities/BaseCalls/Alignment/NA12882_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '18',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12882']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12882_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 177681,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782135,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "   'uuid': '7RrzsgFlAAMAAUGKHKOgfA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143451',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143451', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 925901259,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12883',\n",
       "     'uuid': '7Rr0KAFlAAQAAWj3f0rTCw',\n",
       "     'source': 'platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143451',\n",
       "     'status': {'name': 'READY', 'date': '20180918143451', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 55030461,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5164222,\n",
       "      'INDEL': 706325,\n",
       "      'NO_VARIATION': 49159914},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 48130451,\n",
       "     'chromosomeStats': {'22': {'count': 800414, 'density': 0.015601223},\n",
       "      'X': {'count': 4701296, 'density': 0.03027809},\n",
       "      'Y': {'count': 662048, 'density': 0.011150551},\n",
       "      '10': {'count': 2454649, 'density': 0.018110845},\n",
       "      '11': {'count': 2371707, 'density': 0.017567353},\n",
       "      '12': {'count': 2357769, 'density': 0.01761476},\n",
       "      '13': {'count': 1700567, 'density': 0.014765727},\n",
       "      '14': {'count': 1626895, 'density': 0.015155119},\n",
       "      '15': {'count': 1575609, 'density': 0.015367089},\n",
       "      '16': {'count': 1700249, 'density': 0.018817483},\n",
       "      '17': {'count': 1607124, 'density': 0.019793335},\n",
       "      '18': {'count': 1330740, 'density': 0.01704389},\n",
       "      'MT': {'count': 196, 'density': 0.011827892},\n",
       "      '19': {'count': 1322347, 'density': 0.02236377},\n",
       "      '1': {'count': 4176016, 'density': 0.016754285},\n",
       "      '2': {'count': 4245761, 'density': 0.017457943},\n",
       "      '3': {'count': 3324697, 'density': 0.016789496},\n",
       "      '4': {'count': 3325518, 'density': 0.017397037},\n",
       "      '5': {'count': 3049547, 'density': 0.016856218},\n",
       "      '6': {'count': 2921439, 'density': 0.01707295},\n",
       "      '7': {'count': 2982377, 'density': 0.018740745},\n",
       "      '8': {'count': 2531902, 'density': 0.017298665},\n",
       "      '9': {'count': 2392302, 'density': 0.016941039},\n",
       "      '20': {'count': 1133366, 'density': 0.017982652},\n",
       "      '21': {'count': 735926, 'density': 0.015290413}},\n",
       "     'tiTvRatio': 1.5860314,\n",
       "     'meanQuality': 741.4956,\n",
       "     'stdDevQuality': 768.4001,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.8278 <= x < -0.5959',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -92.9515',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -92.9515 <= x < -1.8278',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '156.24'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '146.07'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '146.49'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '123.12'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '124.74'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '121.29'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '146.19'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '146.46'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '146.58'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '143.73'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '143.85'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '141.6'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '122.43'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '102.96'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '151.32'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '148.89'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '148.83'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '148.2'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '149.67'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '148.89'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '148.5'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '131.49'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '49541.2'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '74.01'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '53.94'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/NA12883_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.10/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12883/Data/Intensities/BaseCalls/Alignment/NA12883_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '20',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12883']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12883_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 183594,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782119,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr1ggFlAAMAAWvxvbzlQQ',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143451',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143451', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 938451577,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12884',\n",
       "     'uuid': '7Rr19QFlAAQAAdckba0OTQ',\n",
       "     'source': 'platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143451',\n",
       "     'status': {'name': 'READY', 'date': '20180918143451', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 56033818,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5162111,\n",
       "      'INDEL': 705312,\n",
       "      'NO_VARIATION': 50166393,\n",
       "      'DELETION': 2},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 48942952,\n",
       "     'chromosomeStats': {'22': {'count': 820844, 'density': 0.015999433},\n",
       "      'X': {'count': 4852543, 'density': 0.031252176},\n",
       "      'Y': {'count': 679933, 'density': 0.011451779},\n",
       "      '10': {'count': 2503595, 'density': 0.018471979},\n",
       "      '11': {'count': 2429325, 'density': 0.017994132},\n",
       "      '12': {'count': 2394439, 'density': 0.01788872},\n",
       "      '13': {'count': 1708279, 'density': 0.0148326885},\n",
       "      '14': {'count': 1640414, 'density': 0.015281053},\n",
       "      '15': {'count': 1615105, 'density': 0.015752297},\n",
       "      '16': {'count': 1738418, 'density': 0.019239917},\n",
       "      '17': {'count': 1671152, 'density': 0.020581905},\n",
       "      '18': {'count': 1346040, 'density': 0.01723985},\n",
       "      'MT': {'count': 205, 'density': 0.012371009},\n",
       "      '19': {'count': 1378628, 'density': 0.023315605},\n",
       "      '1': {'count': 4250650, 'density': 0.017053718},\n",
       "      '2': {'count': 4301394, 'density': 0.017686699},\n",
       "      '3': {'count': 3355299, 'density': 0.016944034},\n",
       "      '4': {'count': 3345575, 'density': 0.017501963},\n",
       "      '5': {'count': 3094172, 'density': 0.01710288},\n",
       "      '6': {'count': 2982766, 'density': 0.017431347},\n",
       "      '7': {'count': 3016336, 'density': 0.018954137},\n",
       "      '8': {'count': 2559047, 'density': 0.017484127},\n",
       "      '9': {'count': 2442415, 'density': 0.017295912},\n",
       "      '20': {'count': 1168301, 'density': 0.018536951},\n",
       "      '21': {'count': 738943, 'density': 0.015353098}},\n",
       "     'tiTvRatio': 1.5787449,\n",
       "     'meanQuality': 736.1315,\n",
       "     'stdDevQuality': 764.63275,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.6986 <= x < -0.5829',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -104.8684',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -104.8684 <= x < -1.6986',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '157.53'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '145.95'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '146.31'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '123.84'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '125.22'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '120.45'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '148.02'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '145.89'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '147.39'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '142.77'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '144.36'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '140.85'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '123.87'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '101.07'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '151.59'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '148.98'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '149.34'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '149.13'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '150.15'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '148.65'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '148.71'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '129.12'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '47751.1'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '74.55'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '54.87'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/NA12884_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12884/Data/Intensities/BaseCalls/Alignment/NA12884_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '22',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12884']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12884_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 196949,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782128,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr3JgFlAAMAAYwDjoYf8Q',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143452',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 914454122,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12885',\n",
       "     'uuid': '7Rr3lAFlAAQAAVk-ZvMytg',\n",
       "     'source': 'platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143452',\n",
       "     'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 53752995,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5179451,\n",
       "      'INDEL': 694299,\n",
       "      'NO_VARIATION': 47879245},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 46979280,\n",
       "     'chromosomeStats': {'22': {'count': 820251, 'density': 0.015987875},\n",
       "      'X': {'count': 2573578, 'density': 0.016574796},\n",
       "      'Y': {'count': 198849, 'density': 0.0033491165},\n",
       "      '10': {'count': 2528342, 'density': 0.018654566},\n",
       "      '11': {'count': 2445768, 'density': 0.018115927},\n",
       "      '12': {'count': 2412973, 'density': 0.018027186},\n",
       "      '13': {'count': 1742586, 'density': 0.01513057},\n",
       "      '14': {'count': 1639722, 'density': 0.015274608},\n",
       "      '15': {'count': 1610444, 'density': 0.015706839},\n",
       "      '16': {'count': 1731582, 'density': 0.01916426},\n",
       "      '17': {'count': 1639342, 'density': 0.020190133},\n",
       "      '18': {'count': 1375692, 'density': 0.017619627},\n",
       "      'MT': {'count': 208, 'density': 0.012552049},\n",
       "      '19': {'count': 1362336, 'density': 0.023040071},\n",
       "      '1': {'count': 4262816, 'density': 0.017102528},\n",
       "      '2': {'count': 4362540, 'density': 0.017938122},\n",
       "      '3': {'count': 3432449, 'density': 0.017333638},\n",
       "      '4': {'count': 3435923, 'density': 0.017974608},\n",
       "      '5': {'count': 3151575, 'density': 0.017420173},\n",
       "      '6': {'count': 3010026, 'density': 0.017590653},\n",
       "      '7': {'count': 3065725, 'density': 0.01926449},\n",
       "      '8': {'count': 2590822, 'density': 0.017701222},\n",
       "      '9': {'count': 2442301, 'density': 0.017295105},\n",
       "      '20': {'count': 1166115, 'density': 0.018502267},\n",
       "      '21': {'count': 751030, 'density': 0.015604231}},\n",
       "     'tiTvRatio': 1.5925604,\n",
       "     'meanQuality': 727.2522,\n",
       "     'stdDevQuality': 747.72394,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.889 <= x < -0.5976',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -17.3489',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -17.3489 <= x < -1.889',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '151.89'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '142.08'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '142.53'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '118.83'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '120.72'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '118.11'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '139.59'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '143.52'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '141.81'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '140.7'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '140.67'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '138.03'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '118.53'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '98.79'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '146.94'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '144.33'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '143.19'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '143.4'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '145.08'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '144.36'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '144.18'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '125.22'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '42336.4'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '138.54'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '8.88'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/NA12885_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12885/Data/Intensities/BaseCalls/Alignment/NA12885_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '24',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12885']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12885_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 187307,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782122,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr4uwFlAAMAAfIMLxPJiQ',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143452',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 873848519,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12886',\n",
       "     'uuid': '7Rr5HAFlAAQAAczlbPuDpA',\n",
       "     'source': 'platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143452',\n",
       "     'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 49850565,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5175946,\n",
       "      'INDEL': 706548,\n",
       "      'NO_VARIATION': 43968071},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 43039160,\n",
       "     'chromosomeStats': {'22': {'count': 734431, 'density': 0.014315119},\n",
       "      'X': {'count': 4404977, 'density': 0.028369686},\n",
       "      'Y': {'count': 640460, 'density': 0.010786955},\n",
       "      '10': {'count': 2229259, 'density': 0.016447878},\n",
       "      '11': {'count': 2151573, 'density': 0.015936809},\n",
       "      '12': {'count': 2099708, 'density': 0.0156868},\n",
       "      '13': {'count': 1510770, 'density': 0.013117753},\n",
       "      '14': {'count': 1465054, 'density': 0.013647512},\n",
       "      '15': {'count': 1447074, 'density': 0.014113473},\n",
       "      '16': {'count': 1553630, 'density': 0.01719478},\n",
       "      '17': {'count': 1443519, 'density': 0.017778376},\n",
       "      '18': {'count': 1193723, 'density': 0.015288999},\n",
       "      'MT': {'count': 201, 'density': 0.012129624},\n",
       "      '19': {'count': 1183706, 'density': 0.020019049},\n",
       "      '1': {'count': 3763987, 'density': 0.015101214},\n",
       "      '2': {'count': 3831741, 'density': 0.015755555},\n",
       "      '3': {'count': 2966193, 'density': 0.014979076},\n",
       "      '4': {'count': 2984823, 'density': 0.015614733},\n",
       "      '5': {'count': 2747159, 'density': 0.015184782},\n",
       "      '6': {'count': 2632595, 'density': 0.0153849395},\n",
       "      '7': {'count': 2697661, 'density': 0.01695164},\n",
       "      '8': {'count': 2276578, 'density': 0.0155542195},\n",
       "      '9': {'count': 2192978, 'density': 0.015529529},\n",
       "      '20': {'count': 1020763, 'density': 0.016196027},\n",
       "      '21': {'count': 678002, 'density': 0.01408692}},\n",
       "     'tiTvRatio': 1.5858048,\n",
       "     'meanQuality': 876.6191,\n",
       "     'stdDevQuality': 888.56165,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.1189 <= x < -0.6322',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -67.4908',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -67.4908 <= x < -2.1189',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '186'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '171.63'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '172.89'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '146.52'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '147.63'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '142.26'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '171.27'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '171.42'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '173.79'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '168.75'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '171.12'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '166.14'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '144.72'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '119.76'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '179.1'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '176.49'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '176.37'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '175.14'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '177.15'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '175.92'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '176.34'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '155.52'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '48363.2'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '88.05'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '64.62'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/NA12886_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12886/Data/Intensities/BaseCalls/Alignment/NA12886_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '26',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12886']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12886_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 183989,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 11782121,\n",
       "     'loadStats': {'stage': True,\n",
       "      'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr6CQFlAAMAAfy69DWx9g',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143452',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 883732819,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12887',\n",
       "     'uuid': '7Rr6SQFlAAQAAd6h7LX5nw',\n",
       "     'source': 'platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143452',\n",
       "     'status': {'name': 'READY', 'date': '20180918143452', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 50472413,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5225689,\n",
       "      'INDEL': 690088,\n",
       "      'NO_VARIATION': 44556634,\n",
       "      'DELETION': 2},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 43698605,\n",
       "     'chromosomeStats': {'22': {'count': 777069, 'density': 0.015146195},\n",
       "      'X': {'count': 2342764, 'density': 0.0150882695},\n",
       "      'Y': {'count': 200570, 'density': 0.0033781026},\n",
       "      '10': {'count': 2375215, 'density': 0.017524768},\n",
       "      '11': {'count': 2292043, 'density': 0.016977277},\n",
       "      '12': {'count': 2262423, 'density': 0.016902436},\n",
       "      '13': {'count': 1630941, 'density': 0.014161177},\n",
       "      '14': {'count': 1557895, 'density': 0.014512359},\n",
       "      '15': {'count': 1534064, 'density': 0.0149618955},\n",
       "      '16': {'count': 1638948, 'density': 0.018139035},\n",
       "      '17': {'count': 1543099, 'density': 0.019004803},\n",
       "      '18': {'count': 1282116, 'density': 0.016421122},\n",
       "      'MT': {'count': 210, 'density': 0.012672742},\n",
       "      '19': {'count': 1270209, 'density': 0.021482002},\n",
       "      '1': {'count': 4023754, 'density': 0.016143406},\n",
       "      '2': {'count': 4096247, 'density': 0.016843164},\n",
       "      '3': {'count': 3205736, 'density': 0.016188752},\n",
       "      '4': {'count': 3207441, 'density': 0.016779331},\n",
       "      '5': {'count': 2946890, 'density': 0.016288785},\n",
       "      '6': {'count': 2836348, 'density': 0.016575675},\n",
       "      '7': {'count': 2879114, 'density': 0.018091857},\n",
       "      '8': {'count': 2435956, 'density': 0.016643135},\n",
       "      '9': {'count': 2332670, 'density': 0.016518755},\n",
       "      '20': {'count': 1090306, 'density': 0.017299436},\n",
       "      '21': {'count': 710385, 'density': 0.014759745}},\n",
       "     'tiTvRatio': 1.5907608,\n",
       "     'meanQuality': 792.11676,\n",
       "     'stdDevQuality': 813.06537,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.2883 <= x < -0.8336',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -165.5806',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -165.5806 <= x < -2.2883',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '168.12'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '156.03'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '156.66'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '132.6'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '132.36'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '128.61'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '155.19'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '155.37'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '157.2'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '152.04'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '154.68'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '150.66'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '131.04'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '106.44'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '162.12'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '159.81'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '159.57'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '159.51'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '160.8'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '159.27'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '159.09'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '137.28'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '44521.4'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '157.35'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '9.75'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/NA12887_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12887/Data/Intensities/BaseCalls/Alignment/NA12887_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '28',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12887']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12887_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 189536,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 13972501,\n",
       "     'loadStats': {'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 4074436,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2179640478855,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 4141943,\n",
       "       'updatedVariants': 5696141,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 44556634,\n",
       "       'nonInsertedVariants': 4748,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 3943340199736,\n",
       "       'fillGapsNanoTime': 5922786464928,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr7SQFlAAMAAQ49KjtaVA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143453',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143453', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 910838017,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12888',\n",
       "     'uuid': '7Rr7oQFlAAQAAYHBsSwk7g',\n",
       "     'source': 'platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143453',\n",
       "     'status': {'name': 'READY', 'date': '20180918143453', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 53310147,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5180247,\n",
       "      'INDEL': 679760,\n",
       "      'NO_VARIATION': 47450138,\n",
       "      'DELETION': 2},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 46296288,\n",
       "     'chromosomeStats': {'22': {'count': 774415, 'density': 0.015094465},\n",
       "      'X': {'count': 4762097, 'density': 0.03066967},\n",
       "      'Y': {'count': 672996, 'density': 0.011334943},\n",
       "      '10': {'count': 2368171, 'density': 0.017472796},\n",
       "      '11': {'count': 2277750, 'density': 0.016871408},\n",
       "      '12': {'count': 2254313, 'density': 0.016841846},\n",
       "      '13': {'count': 1622149, 'density': 0.014084837},\n",
       "      '14': {'count': 1559828, 'density': 0.014530365},\n",
       "      '15': {'count': 1538702, 'density': 0.015007131},\n",
       "      '16': {'count': 1650184, 'density': 0.018263388},\n",
       "      '17': {'count': 1551668, 'density': 0.019110339},\n",
       "      '18': {'count': 1281049, 'density': 0.016407456},\n",
       "      'MT': {'count': 225, 'density': 0.013577937},\n",
       "      '19': {'count': 1277756, 'density': 0.02160964},\n",
       "      '1': {'count': 4033494, 'density': 0.016182482},\n",
       "      '2': {'count': 4085099, 'density': 0.016797325},\n",
       "      '3': {'count': 3196673, 'density': 0.016142985},\n",
       "      '4': {'count': 3207081, 'density': 0.016777448},\n",
       "      '5': {'count': 2929425, 'density': 0.016192248},\n",
       "      '6': {'count': 2823523, 'density': 0.016500726},\n",
       "      '7': {'count': 2878462, 'density': 0.018087761},\n",
       "      '8': {'count': 2437541, 'density': 0.016653964},\n",
       "      '9': {'count': 2324580, 'density': 0.016461466},\n",
       "      '20': {'count': 1092715, 'density': 0.01733766},\n",
       "      '21': {'count': 710251, 'density': 0.014756961}},\n",
       "     'tiTvRatio': 1.5865527,\n",
       "     'meanQuality': 796.5245,\n",
       "     'stdDevQuality': 823.37274,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.8132 <= x < -0.5681',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -81.2459',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -81.2459 <= x < -1.8132',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '169.68'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '158.52'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '159.39'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '134.52'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '136.26'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '130.71'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '153.54'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '157.74'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '159.9'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '152.88'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '155.97'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '153'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '133.53'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '108.66'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '165'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '162.66'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '162.15'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '161.85'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '163.65'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '161.73'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '161.79'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '139.8'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '49803.6'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '80.73'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '58.47'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/NA12888_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12888/Data/Intensities/BaseCalls/Alignment/NA12888_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '30',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12888']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12888_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 202413,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 12622196,\n",
       "     'loadStats': {'merge': False,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5750372,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 47450138,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 1542885258245,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr8qwFlAAMAAb-kqFW0Vg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143453',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143453', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 877382987,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12889',\n",
       "     'uuid': '7Rr9AAFlAAQAATSTE8l9og',\n",
       "     'source': 'platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143453',\n",
       "     'status': {'name': 'READY', 'date': '20180918143453', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 49728394,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5149272,\n",
       "      'INDEL': 667818,\n",
       "      'NO_VARIATION': 43911304},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 42861353,\n",
       "     'chromosomeStats': {'22': {'count': 718369, 'density': 0.0140020475},\n",
       "      'X': {'count': 4462198, 'density': 0.02873821},\n",
       "      'Y': {'count': 645403, 'density': 0.010870207},\n",
       "      '10': {'count': 2241221, 'density': 0.016536135},\n",
       "      '11': {'count': 2110728, 'density': 0.015634269},\n",
       "      '12': {'count': 2085746, 'density': 0.015582492},\n",
       "      '13': {'count': 1528175, 'density': 0.013268877},\n",
       "      '14': {'count': 1461866, 'density': 0.013617814},\n",
       "      '15': {'count': 1445774, 'density': 0.014100794},\n",
       "      '16': {'count': 1539927, 'density': 0.017043121},\n",
       "      '17': {'count': 1440054, 'density': 0.017735701},\n",
       "      '18': {'count': 1193015, 'density': 0.015279932},\n",
       "      'MT': {'count': 197, 'density': 0.011888239},\n",
       "      '19': {'count': 1153651, 'density': 0.019510753},\n",
       "      '1': {'count': 3741340, 'density': 0.015010353},\n",
       "      '2': {'count': 3802853, 'density': 0.015636772},\n",
       "      '3': {'count': 2977073, 'density': 0.015034019},\n",
       "      '4': {'count': 2994262, 'density': 0.015664112},\n",
       "      '5': {'count': 2760481, 'density': 0.015258419},\n",
       "      '6': {'count': 2596125, 'density': 0.015171808},\n",
       "      '7': {'count': 2675321, 'density': 0.016811257},\n",
       "      '8': {'count': 2277399, 'density': 0.015559829},\n",
       "      '9': {'count': 2194364, 'density': 0.015539344},\n",
       "      '20': {'count': 1005888, 'density': 0.015960012},\n",
       "      '21': {'count': 676964, 'density': 0.014065353}},\n",
       "     'tiTvRatio': 1.6038222,\n",
       "     'meanQuality': 847.96674,\n",
       "     'stdDevQuality': 869.436,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.9394 <= x < -0.5435',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -52.1421',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -52.1421 <= x < -1.9394',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '179.1'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '168.99'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '170.25'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '144.15'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '146.88'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '140.94'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '169.53'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '169.17'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '170.25'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '164.28'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '166.08'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '163.14'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '141.36'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '117.33'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '175.68'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '173.67'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '173.91'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '173.4'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '174.48'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '172.2'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '172.98'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '149.67'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '58097.8'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '86.43'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '60.69'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/NA12889_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12889/Data/Intensities/BaseCalls/Alignment/NA12889_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '32',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12889']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12889_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 183335,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 12309177,\n",
       "     'loadStats': {'merge': False,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5160000,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 43911304,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 1224578941079,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "   'uuid': '7Rr--AFlAAMAAUPpArVMCA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143454',\n",
       "   'modificationDate': '20180920010649',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143454', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 932054336,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12890',\n",
       "     'uuid': '7Rr_bAFlAAQAATmXckLQ0Q',\n",
       "     'source': 'platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143454',\n",
       "     'status': {'name': 'READY', 'date': '20180918143454', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010649',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 55569550,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5138020,\n",
       "      'INDEL': 693497,\n",
       "      'NO_VARIATION': 49738033},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 48850454,\n",
       "     'chromosomeStats': {'22': {'count': 832903, 'density': 0.01623448},\n",
       "      'X': {'count': 2655805, 'density': 0.017104369},\n",
       "      'Y': {'count': 196678, 'density': 0.0033125514},\n",
       "      '10': {'count': 2608646, 'density': 0.019247064},\n",
       "      '11': {'count': 2517275, 'density': 0.018645583},\n",
       "      '12': {'count': 2512380, 'density': 0.01876985},\n",
       "      '13': {'count': 1796261, 'density': 0.015596622},\n",
       "      '14': {'count': 1716579, 'density': 0.015990559},\n",
       "      '15': {'count': 1681962, 'density': 0.01640436},\n",
       "      '16': {'count': 1777017, 'density': 0.019667111},\n",
       "      '17': {'count': 1699518, 'density': 0.02093126},\n",
       "      '18': {'count': 1407915, 'density': 0.018032335},\n",
       "      'MT': {'count': 211, 'density': 0.012733088},\n",
       "      '19': {'count': 1377273, 'density': 0.023292689},\n",
       "      '1': {'count': 4436137, 'density': 0.017797897},\n",
       "      '2': {'count': 4504081, 'density': 0.018520117},\n",
       "      '3': {'count': 3561325, 'density': 0.017984452},\n",
       "      '4': {'count': 3568809, 'density': 0.018669784},\n",
       "      '5': {'count': 3267108, 'density': 0.018058775},\n",
       "      '6': {'count': 3146229, 'density': 0.018386627},\n",
       "      '7': {'count': 3158135, 'density': 0.019845178},\n",
       "      '8': {'count': 2684762, 'density': 0.018343046},\n",
       "      '9': {'count': 2503010, 'density': 0.017725015},\n",
       "      '20': {'count': 1195790, 'density': 0.018973108},\n",
       "      '21': {'count': 763741, 'density': 0.015868329}},\n",
       "     'tiTvRatio': 1.6143893,\n",
       "     'meanQuality': 709.5657,\n",
       "     'stdDevQuality': 726.36945,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.3404 <= x < -0.7425',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -162.7559',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -162.7559 <= x < -2.3404',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '148.92'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '137.49'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '136.53'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '115.44'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '115.44'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '114.09'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '131.58'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '137.1'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '137.88'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '135.99'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '135.9'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '133.05'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '115.35'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '94.89'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '141.48'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '138.93'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '139.02'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '138.18'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '139.95'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '139.53'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '139.41'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '118.86'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '48957.4'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '134.73'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '8.97'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/NA12890_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12890/Data/Intensities/BaseCalls/Alignment/NA12890_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '34',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12890']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12890_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 194379,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 12608715,\n",
       "     'loadStats': {'merge': False,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5275292,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 49738033,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 1530212264974,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "   'uuid': '7RsAoQFlAAMAAcipSn1zag',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143454',\n",
       "   'modificationDate': '20180920010650',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143454', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 912085106,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12891',\n",
       "     'uuid': '7RsBBAFlAAQAAf7ayUWWNA',\n",
       "     'source': 'platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143454',\n",
       "     'status': {'name': 'READY', 'date': '20180918143454', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010650',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 53711209,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5156806,\n",
       "      'INDEL': 671320,\n",
       "      'NO_VARIATION': 47883082,\n",
       "      'DELETION': 1},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 46735244,\n",
       "     'chromosomeStats': {'22': {'count': 771692, 'density': 0.0150413895},\n",
       "      'X': {'count': 4615810, 'density': 0.029727528},\n",
       "      'Y': {'count': 652853, 'density': 0.010995684},\n",
       "      '10': {'count': 2394342, 'density': 0.01766589},\n",
       "      '11': {'count': 2301050, 'density': 0.017043993},\n",
       "      '12': {'count': 2296689, 'density': 0.017158434},\n",
       "      '13': {'count': 1640799, 'density': 0.014246772},\n",
       "      '14': {'count': 1560727, 'density': 0.01453874},\n",
       "      '15': {'count': 1543314, 'density': 0.015052112},\n",
       "      '16': {'count': 1658778, 'density': 0.018358503},\n",
       "      '17': {'count': 1564968, 'density': 0.019274142},\n",
       "      '18': {'count': 1293856, 'density': 0.016571486},\n",
       "      'MT': {'count': 234, 'density': 0.014121055},\n",
       "      '19': {'count': 1303248, 'density': 0.022040764},\n",
       "      '1': {'count': 4064238, 'density': 0.016305828},\n",
       "      '2': {'count': 4150803, 'density': 0.01706749},\n",
       "      '3': {'count': 3256727, 'density': 0.016446253},\n",
       "      '4': {'count': 3264106, 'density': 0.017075768},\n",
       "      '5': {'count': 2984083, 'density': 0.01649437},\n",
       "      '6': {'count': 2854909, 'density': 0.016684147},\n",
       "      '7': {'count': 2901164, 'density': 0.018230416},\n",
       "      '8': {'count': 2463617, 'density': 0.016832123},\n",
       "      '9': {'count': 2347811, 'density': 0.016625976},\n",
       "      '20': {'count': 1100429, 'density': 0.017460054},\n",
       "      '21': {'count': 724962, 'density': 0.015062613}},\n",
       "     'tiTvRatio': 1.5854445,\n",
       "     'meanQuality': 763.1102,\n",
       "     'stdDevQuality': 795.5039,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -1.9206 <= x < -0.5705',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -121.4007',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -121.4007 <= x < -1.9206',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '166.59'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '152.43'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '152.67'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '129.48'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '131.19'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '125.46'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '158.4'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '151.65'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '154.38'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '148.26'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '151.2'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '147.66'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '128.46'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '106.08'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '159'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '155.88'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '155.58'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '154.92'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '156.99'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '156.18'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '155.46'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '134.52'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '89313.5'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '78.03'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '61.2'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/NA12891_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12891/Data/Intensities/BaseCalls/Alignment/NA12891_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '36',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12891']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12891_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 185891,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 12906597,\n",
       "     'loadStats': {'merge': False,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5367567,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 47883082,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 1824458393361,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "   'uuid': '7RsCRwFlAAMAAYq7sPoJsA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143454',\n",
       "   'modificationDate': '20180920010650',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143454', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 857494765,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12892',\n",
       "     'uuid': '7RsC2gFlAAQAAY_i37kSHA',\n",
       "     'source': 'platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143455',\n",
       "     'status': {'name': 'READY', 'date': '20180918143455', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010650',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 48377167,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5204012,\n",
       "      'INDEL': 695713,\n",
       "      'NO_VARIATION': 42477442},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 41857151,\n",
       "     'chromosomeStats': {'22': {'count': 740168, 'density': 0.014426941},\n",
       "      'X': {'count': 2282531, 'density': 0.014700347},\n",
       "      'Y': {'count': 204891, 'density': 0.0034508791},\n",
       "      '10': {'count': 2278165, 'density': 0.016808715},\n",
       "      '11': {'count': 2189051, 'density': 0.01621441},\n",
       "      '12': {'count': 2170811, 'density': 0.016218007},\n",
       "      '13': {'count': 1562623, 'density': 0.013567983},\n",
       "      '14': {'count': 1488893, 'density': 0.01386958},\n",
       "      '15': {'count': 1472234, 'density': 0.014358861},\n",
       "      '16': {'count': 1569383, 'density': 0.017369125},\n",
       "      '17': {'count': 1473709, 'density': 0.018150197},\n",
       "      '18': {'count': 1233799, 'density': 0.015802287},\n",
       "      'MT': {'count': 193, 'density': 0.011646853},\n",
       "      '19': {'count': 1188629, 'density': 0.020102307},\n",
       "      '1': {'count': 3857357, 'density': 0.015475817},\n",
       "      '2': {'count': 3922617, 'density': 0.016129224},\n",
       "      '3': {'count': 3072013, 'density': 0.015513459},\n",
       "      '4': {'count': 3087622, 'density': 0.016152514},\n",
       "      '5': {'count': 2835195, 'density': 0.015671397},\n",
       "      '6': {'count': 2712278, 'density': 0.01585061},\n",
       "      '7': {'count': 2742882, 'density': 0.0172358},\n",
       "      '8': {'count': 2333288, 'density': 0.015941678},\n",
       "      '9': {'count': 2242022, 'density': 0.015876833},\n",
       "      '20': {'count': 1034648, 'density': 0.016416334},\n",
       "      '21': {'count': 682165, 'density': 0.0141734155}},\n",
       "     'tiTvRatio': 1.609316,\n",
       "     'meanQuality': 827.8475,\n",
       "     'stdDevQuality': 847.29846,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.0091 <= x < -0.648',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -63.2944',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -63.2944 <= x < -2.0091',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '176.4'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '162.06'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '162.48'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '137.52'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '137.43'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '132.15'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '155.43'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '161.58'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '163.32'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '158.82'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '160.62'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '156.99'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '136.65'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '111'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '168.45'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '165.72'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '165.81'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '165.48'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '167.91'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '166.02'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '165.33'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '145.17'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '64022.4'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '159.12'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '10.2'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/NA12892_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12892/Data/Intensities/BaseCalls/Alignment/NA12892_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '38',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12892']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12892_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 168838,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 13136261,\n",
       "     'loadStats': {'merge': False,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5468601,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 42477442,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2076435504869,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "   'annotationSets': [],\n",
       "   'name': 'platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "   'uuid': '7RsEUQFlAAMAAeL-lYaTng',\n",
       "   'type': 'FILE',\n",
       "   'format': 'VCF',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "   'path': 'data/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "   'release': 1,\n",
       "   'creationDate': '20180918143455',\n",
       "   'modificationDate': '20180920010650',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918143455', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 901763957,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12893',\n",
       "     'uuid': '7RsE2AFlAAQAAUy3qk3cAg',\n",
       "     'source': 'platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143455',\n",
       "     'status': {'name': 'READY', 'date': '20180918143455', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'status': {'name': 'READY',\n",
       "     'date': '20180920010650',\n",
       "     'message': 'Job finished. File index ready'},\n",
       "    'jobId': 0,\n",
       "    'release': 1},\n",
       "   'stats': {'variantFileStats': {'numVariants': 52580350,\n",
       "     'variantTypeCounts': {'SV': 0,\n",
       "      'SNP': 0,\n",
       "      'SNV': 5179891,\n",
       "      'INDEL': 687774,\n",
       "      'NO_VARIATION': 46712684,\n",
       "      'DELETION': 1},\n",
       "     'numSamples': 1,\n",
       "     'numPass': 45643882,\n",
       "     'chromosomeStats': {'22': {'count': 768812, 'density': 0.014985254},\n",
       "      'X': {'count': 4616991, 'density': 0.029735135},\n",
       "      'Y': {'count': 662538, 'density': 0.0111588035},\n",
       "      '10': {'count': 2357340, 'density': 0.017392883},\n",
       "      '11': {'count': 2261241, 'density': 0.016749125},\n",
       "      '12': {'count': 2223750, 'density': 0.016613511},\n",
       "      '13': {'count': 1602702, 'density': 0.013915982},\n",
       "      '14': {'count': 1528104, 'density': 0.014234845},\n",
       "      '15': {'count': 1520987, 'density': 0.014834355},\n",
       "      '16': {'count': 1616252, 'density': 0.017887847},\n",
       "      '17': {'count': 1526081, 'density': 0.01879521},\n",
       "      '18': {'count': 1264851, 'density': 0.016199995},\n",
       "      'MT': {'count': 224, 'density': 0.013517591},\n",
       "      '19': {'count': 1236599, 'density': 0.020913584},\n",
       "      '1': {'count': 3980210, 'density': 0.015968706},\n",
       "      '2': {'count': 4040973, 'density': 0.016615886},\n",
       "      '3': {'count': 3151891, 'density': 0.015916837},\n",
       "      '4': {'count': 3186143, 'density': 0.016667914},\n",
       "      '5': {'count': 2914456, 'density': 0.01610951},\n",
       "      '6': {'count': 2801763, 'density': 0.01637356},\n",
       "      '7': {'count': 2841850, 'density': 0.017857697},\n",
       "      '8': {'count': 2410341, 'density': 0.016468126},\n",
       "      '9': {'count': 2298513, 'density': 0.016276874},\n",
       "      '20': {'count': 1067215, 'density': 0.016933061},\n",
       "      '21': {'count': 700523, 'density': 0.014554841}},\n",
       "     'tiTvRatio': 1.5910804,\n",
       "     'meanQuality': 802.50867,\n",
       "     'stdDevQuality': 823.6104,\n",
       "     'numRareVariants': [],\n",
       "     'consequenceTypesCounts': {},\n",
       "     'variantBiotypeCounts': {}}},\n",
       "   'attributes': {'variantFileMetadata': {'header': {'version': 'VCFv4.2',\n",
       "      'complexLines': [{'key': 'FORMAT',\n",
       "        'id': 'GQX',\n",
       "        'description': 'Minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'AD',\n",
       "        'description': 'Allelic depths for the ref and alt alleles in the order listed',\n",
       "        'number': 'R',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth (reads with MQ=255 or with bad mates are filtered)',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GQ',\n",
       "        'description': 'Genotype Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'GT',\n",
       "        'description': 'Genotype',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'PL',\n",
       "        'description': 'Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification',\n",
       "        'number': 'G',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FORMAT',\n",
       "        'id': 'VF',\n",
       "        'description': 'Variant Frequency, the ratio of the sum of the called variant depth to the total depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AC',\n",
       "        'description': 'Allele count in genotypes, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AF',\n",
       "        'description': 'Allele Frequency, for each ALT allele, in the same order as listed',\n",
       "        'number': 'A',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'AN',\n",
       "        'description': 'Total number of alleles in called genotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DP',\n",
       "        'description': 'Approximate read depth; some reads may have been filtered',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'QD',\n",
       "        'description': 'Variant Confidence/Quality by Depth',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BLOCKAVG_min30p3a',\n",
       "        'description': 'Non-variant site block. All sites in a block are constrained to be non-variant, have the same filter value, and have all sample values in range [x,y], y <= max(x+3,(x*1.3)). All printed site block sample values are the minimum observed in the region spanned by the block',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'BaseQRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'DS',\n",
       "        'description': 'Were any of the samples downsampled?',\n",
       "        'number': '0',\n",
       "        'type': 'Flag',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'Dels',\n",
       "        'description': 'Fraction of Reads Containing Spanning Deletions',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'END',\n",
       "        'description': 'End position of the region described in this record',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'FS',\n",
       "        'description': \"Phred-scaled p-value using Fisher's exact test to detect strand bias\",\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HRun',\n",
       "        'description': 'Largest Contiguous Homopolymer Run of Variant Allele In Either Direction',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'HaplotypeScore',\n",
       "        'description': 'Consistency of the site with at most two segregating haplotypes',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'InbreedingCoeff',\n",
       "        'description': 'Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ',\n",
       "        'description': 'RMS Mapping Quality',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQ0',\n",
       "        'description': 'Total Mapping Quality Zero Reads',\n",
       "        'number': '1',\n",
       "        'type': 'Integer',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'MQRankSum',\n",
       "        'description': 'Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'ReadPosRankSum',\n",
       "        'description': 'Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'SB',\n",
       "        'description': 'Strand Bias',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'VQSLOD',\n",
       "        'description': 'Log odds ratio of being a true variant versus being false under the trained gaussian mixture model',\n",
       "        'number': '1',\n",
       "        'type': 'Float',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'culprit',\n",
       "        'description': 'The annotation which was the worst performing in the Gaussian mixture model, likely the reason why the variant was filtered out',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'INFO',\n",
       "        'id': 'set',\n",
       "        'description': 'Source VCF for the merged record in CombineVariants',\n",
       "        'number': '1',\n",
       "        'type': 'String',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowGQX',\n",
       "        'description': 'Locus GQX is less than 30.0000 or not present',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowQD',\n",
       "        'description': 'Locus QD is less than 2.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'LowMQ',\n",
       "        'description': 'Site MQ is less than 20.0000',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'IndelConflict',\n",
       "        'description': 'Locus is in region with conflicting indel calls.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'MaxDepth',\n",
       "        'description': 'Site depth is greater than 3.0x the mean chromosome depth',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'SiteConflict',\n",
       "        'description': 'Site genotype conflicts with proximal indel call. This is typically a heterozygous SNV call made inside of a heterozygous deletion.',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.00to99.90',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -2.1401 <= x < -0.6433',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00+',\n",
       "        'description': 'Truth sensitivity tranche level at VQS Lod < -62.7734',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'FILTER',\n",
       "        'id': 'TruthSensitivityTranche99.90to100.00',\n",
       "        'description': 'Truth sensitivity tranche level at VSQ Lod: -62.7734 <= x < -2.1401',\n",
       "        'genericFields': {}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr1',\n",
       "        'genericFields': {'length': '249250621'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr10',\n",
       "        'genericFields': {'length': '135534747'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr11',\n",
       "        'genericFields': {'length': '135006516'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr12',\n",
       "        'genericFields': {'length': '133851895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr13',\n",
       "        'genericFields': {'length': '115169878'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr14',\n",
       "        'genericFields': {'length': '107349540'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr15',\n",
       "        'genericFields': {'length': '102531392'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr16',\n",
       "        'genericFields': {'length': '90354753'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr17',\n",
       "        'genericFields': {'length': '81195210'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr18',\n",
       "        'genericFields': {'length': '78077248'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr19',\n",
       "        'genericFields': {'length': '59128983'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr2',\n",
       "        'genericFields': {'length': '243199373'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr20',\n",
       "        'genericFields': {'length': '63025520'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr21',\n",
       "        'genericFields': {'length': '48129895'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr22',\n",
       "        'genericFields': {'length': '51304566'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr3',\n",
       "        'genericFields': {'length': '198022430'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr4',\n",
       "        'genericFields': {'length': '191154276'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr5',\n",
       "        'genericFields': {'length': '180915260'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr6',\n",
       "        'genericFields': {'length': '171115067'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr7',\n",
       "        'genericFields': {'length': '159138663'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr8',\n",
       "        'genericFields': {'length': '146364022'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chr9',\n",
       "        'genericFields': {'length': '141213431'}},\n",
       "       {'key': 'contig', 'id': 'chrM', 'genericFields': {'length': '16571'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrX',\n",
       "        'genericFields': {'length': '155270560'}},\n",
       "       {'key': 'contig',\n",
       "        'id': 'chrY',\n",
       "        'genericFields': {'length': '59373566'}}],\n",
       "      'simpleLines': [{'key': 'ApplyRecalibration',\n",
       "        'value': '\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false input=[(RodBinding name=input source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.Indel.txt)] recal_file=(RodBinding name=recal_file source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.recal) tranches_file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.tranches out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub ts_filter_level=99.0 ignore_filter=null mode=INDEL filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'CombineVariants',\n",
       "        'value': '\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=INFO log_to_file=null help=false variant=[(RodBinding name=variant source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.0.output), (RodBinding name=variant2 source=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/tempvariants_G1.1.vcf.1.output)] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub genotypemergeoption=PRIORITIZE filteredrecordsmergetype=KEEP_IF_ANY_UNFILTERED multipleallelesmergetype=BY_TYPE rod_priority_list=variant2,variant printComplexMerges=false filteredAreUncalled=false minimalVCF=false setKey=set assumeIdenticalSamples=false minimumN=1 suppressCommandLineHeader=false mergeInfoWithMaxAC=false filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'MaxDepth_chr10', 'value': '170.58'},\n",
       "       {'key': 'MaxDepth_chr11', 'value': '158.97'},\n",
       "       {'key': 'MaxDepth_chr12', 'value': '159.3'},\n",
       "       {'key': 'MaxDepth_chr13', 'value': '135.45'},\n",
       "       {'key': 'MaxDepth_chr14', 'value': '136.44'},\n",
       "       {'key': 'MaxDepth_chr15', 'value': '131.55'},\n",
       "       {'key': 'MaxDepth_chr16', 'value': '155.4'},\n",
       "       {'key': 'MaxDepth_chr17', 'value': '158.13'},\n",
       "       {'key': 'MaxDepth_chr18', 'value': '160.02'},\n",
       "       {'key': 'MaxDepth_chr19', 'value': '155.04'},\n",
       "       {'key': 'MaxDepth_chr1', 'value': '156.39'},\n",
       "       {'key': 'MaxDepth_chr20', 'value': '153.51'},\n",
       "       {'key': 'MaxDepth_chr21', 'value': '133.32'},\n",
       "       {'key': 'MaxDepth_chr22', 'value': '110.1'},\n",
       "       {'key': 'MaxDepth_chr2', 'value': '165.03'},\n",
       "       {'key': 'MaxDepth_chr3', 'value': '162.39'},\n",
       "       {'key': 'MaxDepth_chr4', 'value': '161.64'},\n",
       "       {'key': 'MaxDepth_chr5', 'value': '162.42'},\n",
       "       {'key': 'MaxDepth_chr6', 'value': '163.41'},\n",
       "       {'key': 'MaxDepth_chr7', 'value': '162.12'},\n",
       "       {'key': 'MaxDepth_chr8', 'value': '161.79'},\n",
       "       {'key': 'MaxDepth_chr9', 'value': '143.28'},\n",
       "       {'key': 'MaxDepth_chrM', 'value': '53755.3'},\n",
       "       {'key': 'MaxDepth_chrX', 'value': '81.03'},\n",
       "       {'key': 'MaxDepth_chrY', 'value': '59.25'},\n",
       "       {'key': 'UnifiedGenotyper',\n",
       "        'value': '\"analysis_type=UnifiedGenotyper input_file=[/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/NA12893_S1.bam] read_buffer_size=null phone_home=NO_ET gatk_key=null read_filter=[] intervals=[chrM:1-16571] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL reference_sequence=/illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa nonDeterministicRandomSeed=false downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=-1 defaultBaseQualities=-1 validation_strictness=SILENT unsafe=null num_threads=1 num_cpu_threads=null num_io_threads=null num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false logging_level=OFF log_to_file=null help=false genotype_likelihoods_model=BOTH p_nonref_model=EXACT heterozygosity=0.0010 pcr_error_rate=1.0E-4 genotyping_mode=DISCOVERY output_mode=EMIT_ALL_SITES standard_min_confidence_threshold_for_calling=1.0 standard_min_confidence_threshold_for_emitting=30.0 noSLOD=false annotateNDA=false alleles=(RodBinding name= source=UNBOUND) min_base_quality_score=17 max_deletion_fraction=0.05 max_alternate_alleles=3 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indel_heterozygosity=1.25E-4 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 noBandedIndel=false indelDebug=false ignoreSNPAlleles=false includeGVCFData=true dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub NO_HEADER=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VCFWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false\"'},\n",
       "       {'key': 'gvcftools_cmdline',\n",
       "        'value': '\"/illumina/development/Isis/2.2.13/gatk_to_gvcf --chrom-depth-file=/illumina/builds/platinumgenomes/Platinum_Genomes_Short_Insert/bwa_0.6.1/NA12893/Data/Intensities/BaseCalls/Alignment/NA12893_S1.bam.depth --no-default-filters --min-qd 2.0000 --min-gqx 30.0000 --min-mq 20.0000\"'},\n",
       "       {'key': 'gvcftools_version', 'value': '\"v0.8\"'},\n",
       "       {'key': 'reference',\n",
       "        'value': 'file:///illumina/development/Isis/Genomes/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFASTA/genome.fa'},\n",
       "       {'key': 'source', 'value': 'GATK 1.6'}]},\n",
       "     'id': '40',\n",
       "     'attributes': {},\n",
       "     'path': 'platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "     'sampleIds': ['NA12893']},\n",
       "    'storagePipelineResult': {'input': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "     'preTransformResult': 'file:///mnt/data/downloads/datasets/vcf/platinum_genomes/gz/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz',\n",
       "     'transformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postTransformResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'preLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'loadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'postLoadResult': 'file:/mnt/data/opencga-test/sessions/jobs/j_tmp/platinum-genomes-vcf-NA12893_S1.genome.vcf.gz.variants.avro.gz',\n",
       "     'transformExecuted': True,\n",
       "     'transformTimeMillis': 189123,\n",
       "     'transformStats': {},\n",
       "     'loadExecuted': True,\n",
       "     'loadTimeMillis': 13555150,\n",
       "     'loadStats': {'merge': True,\n",
       "      'stageWriteResult': {'newVariants': 0,\n",
       "       'updatedVariants': 5788849,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 0,\n",
       "       'newVariantsNanoTime': 2482757632798,\n",
       "       'existingVariantsNanoTime': 0,\n",
       "       'fillGapsNanoTime': 0,\n",
       "       'genotypes': []},\n",
       "      'stage': True,\n",
       "      'mergeWriteResult': {'newVariants': 2291997,\n",
       "       'updatedVariants': 8628342,\n",
       "       'updatedMissingVariants': 0,\n",
       "       'overlappedVariants': 0,\n",
       "       'skippedVariants': 46712684,\n",
       "       'nonInsertedVariants': 2770,\n",
       "       'newVariantsNanoTime': 0,\n",
       "       'existingVariantsNanoTime': 1474606930134,\n",
       "       'fillGapsNanoTime': 7757474696817,\n",
       "       'genotypes': ['1',\n",
       "        '0/1',\n",
       "        '1/1',\n",
       "        '1/2',\n",
       "        '1/3',\n",
       "        '1/4',\n",
       "        '1/5',\n",
       "        '1/6',\n",
       "        '-1/-1']}}}}},\n",
       "  {'id': 'data:NA12877_S1.bam',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12877_S1.bam',\n",
       "   'uuid': 'rOjidwFnAAMAAZbltGDiUQ',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAM',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12877_S1.bam',\n",
       "   'path': 'data/NA12877_S1.bam',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133002',\n",
       "   'modificationDate': '20190118021738',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133002', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 130862123347,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12877',\n",
       "     'uuid': '7RrnvgFlAAQAAfde27jGkg',\n",
       "     'source': 'platinum-genomes-vcf-NA12877_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143448',\n",
       "     'status': {'name': 'READY', 'date': '20180918143448', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {'globalStats': '{\"numMapped\":44353,\"numUnmapped\":5647,\"numPaired\":29971,\"numMappedFirst\":23082,\"numMappedSecond\":21271,\"NM\":86967,\"numHardC\":0,\"numSoftC\":8342,\"numIn\":5539,\"numDel\":332,\"numPad\":0,\"numSkip\":0,\"accMappingQuality\":2241515,\"mappingQualityMap\":{\"0\":7,\"37\":14219,\"7\":4,\"11\":1,\"13\":1,\"15\":32,\"48\":9,\"17\":104,\"50\":11,\"55\":2,\"23\":6,\"25\":162,\"57\":4,\"60\":27215,\"29\":2576},\"accInsert\":9077449,\"insertMap\":{\"73\":2,\"100\":2,\"102\":10,\"103\":6,\"104\":2,\"105\":4,\"106\":8,\"107\":4,\"108\":6,\"109\":6,\"111\":2,\"112\":10,\"113\":6,\"115\":4,\"116\":8,\"117\":8,\"118\":12,\"119\":8,\"120\":11,\"121\":2,\"122\":10,\"123\":4,\"124\":2,\"125\":6,\"126\":6,\"127\":4,\"128\":6,\"129\":6,\"130\":8,\"131\":8,\"132\":4,\"133\":12,\"134\":18,\"135\":10,\"136\":4,\"137\":10,\"138\":5,\"139\":24,\"140\":12,\"141\":8,\"142\":13,\"143\":12,\"144\":8,\"145\":11,\"146\":10,\"147\":13,\"148\":14,\"149\":15,\"150\":17,\"151\":12,\"152\":13,\"153\":8,\"154\":13,\"155\":10,\"156\":18,\"157\":30,\"158\":20,\"159\":22,\"160\":8,\"161\":16,\"162\":13,\"163\":22,\"164\":11,\"165\":21,\"166\":14,\"167\":30,\"168\":21,\"169\":17,\"170\":21,\"171\":24,\"172\":28,\"173\":12,\"174\":26,\"175\":26,\"176\":25,\"177\":28,\"178\":20,\"179\":33,\"180\":20,\"181\":39,\"182\":22,\"183\":32,\"184\":35,\"185\":46,\"186\":34,\"187\":43,\"188\":37,\"189\":45,\"190\":39,\"191\":64,\"192\":43,\"193\":29,\"194\":40,\"195\":52,\"196\":32,\"197\":45,\"198\":45,\"199\":42,\"200\":51,\"201\":34,\"202\":44,\"203\":49,\"204\":60,\"205\":44,\"206\":68,\"207\":54,\"208\":49,\"209\":56,\"210\":70,\"211\":58,\"212\":52,\"213\":70,\"214\":78,\"215\":71,\"216\":80,\"217\":56,\"218\":56,\"219\":80,\"220\":81,\"221\":97,\"222\":72,\"223\":73,\"224\":64,\"225\":70,\"226\":79,\"227\":101,\"228\":81,\"229\":90,\"230\":63,\"231\":115,\"232\":95,\"233\":99,\"234\":106,\"235\":121,\"236\":99,\"237\":95,\"238\":117,\"239\":96,\"240\":99,\"241\":111,\"242\":133,\"243\":130,\"244\":121,\"245\":115,\"246\":117,\"247\":139,\"248\":137,\"249\":137,\"250\":144,\"251\":133,\"252\":154,\"253\":147,\"254\":148,\"255\":116,\"256\":164,\"257\":145,\"258\":152,\"259\":160,\"260\":155,\"261\":144,\"262\":150,\"263\":195,\"264\":168,\"265\":175,\"266\":165,\"267\":152,\"268\":166,\"269\":171,\"270\":175,\"271\":161,\"272\":178,\"273\":193,\"274\":162,\"275\":203,\"276\":190,\"277\":176,\"278\":164,\"279\":210,\"280\":140,\"281\":177,\"282\":186,\"283\":213,\"284\":226,\"285\":173,\"286\":197,\"287\":196,\"288\":184,\"289\":235,\"290\":191,\"291\":188,\"292\":201,\"293\":151,\"294\":193,\"295\":201,\"296\":197,\"297\":182,\"298\":206,\"299\":208,\"300\":203,\"301\":167,\"302\":166,\"303\":206,\"304\":200,\"305\":154,\"306\":180,\"307\":200,\"308\":208,\"309\":207,\"310\":172,\"311\":165,\"312\":180,\"313\":174,\"314\":229,\"315\":165,\"316\":159,\"317\":166,\"318\":169,\"319\":160,\"320\":173,\"321\":166,\"322\":174,\"323\":161,\"324\":175,\"325\":163,\"326\":168,\"327\":179,\"328\":160,\"329\":184,\"330\":132,\"331\":167,\"332\":165,\"333\":180,\"334\":156,\"335\":176,\"336\":138,\"337\":163,\"338\":180,\"339\":133,\"340\":138,\"341\":147,\"342\":163,\"343\":150,\"344\":134,\"345\":178,\"346\":136,\"347\":120,\"348\":150,\"349\":122,\"350\":129,\"351\":145,\"352\":126,\"353\":83,\"354\":133,\"355\":127,\"356\":133,\"357\":105,\"358\":121,\"359\":137,\"360\":115,\"361\":120,\"362\":123,\"363\":121,\"364\":121,\"365\":125,\"366\":116,\"367\":110,\"368\":118,\"369\":119,\"370\":125,\"371\":93,\"372\":129,\"373\":82,\"374\":75,\"375\":108,\"376\":73,\"377\":76,\"378\":79,\"379\":89,\"380\":89,\"381\":85,\"382\":61,\"383\":79,\"384\":71,\"385\":73,\"386\":70,\"387\":76,\"388\":78,\"389\":52,\"390\":76,\"391\":55,\"392\":76,\"393\":64,\"394\":44,\"395\":70,\"396\":69,\"397\":64,\"398\":71,\"399\":41,\"400\":51,\"401\":61,\"402\":49,\"403\":50,\"404\":34,\"405\":50,\"406\":44,\"407\":63,\"408\":47,\"409\":32,\"410\":44,\"411\":48,\"412\":49,\"413\":37,\"414\":37,\"415\":37,\"416\":30,\"417\":44,\"418\":29,\"419\":31,\"420\":27,\"421\":32,\"422\":38,\"423\":25,\"424\":37,\"425\":26,\"426\":18,\"427\":33,\"428\":37,\"429\":33,\"430\":27,\"431\":30,\"432\":18,\"433\":24,\"434\":18,\"435\":20,\"436\":21,\"437\":19,\"438\":22,\"439\":15,\"440\":17,\"441\":17,\"442\":15,\"443\":16,\"444\":15,\"445\":17,\"446\":19,\"447\":14,\"448\":21,\"449\":20,\"450\":16,\"451\":10,\"452\":21,\"453\":13,\"454\":12,\"455\":13,\"456\":13,\"457\":11,\"458\":14,\"459\":11,\"460\":18,\"461\":10,\"462\":11,\"463\":13,\"464\":13,\"465\":7,\"466\":9,\"467\":9,\"468\":18,\"469\":5,\"470\":9,\"471\":4,\"472\":9,\"473\":7,\"474\":8,\"475\":7,\"476\":7,\"477\":8,\"478\":6,\"479\":4,\"480\":8,\"481\":7,\"482\":6,\"483\":4,\"484\":6,\"485\":7,\"486\":5,\"487\":8,\"488\":3,\"489\":1,\"490\":7,\"491\":4,\"492\":5,\"493\":4,\"494\":7,\"495\":7,\"496\":3,\"497\":2,\"498\":4,\"499\":1,\"500\":3,\"501\":5,\"502\":6,\"503\":7,\"504\":3,\"505\":2,\"506\":2,\"507\":3,\"508\":1,\"509\":1,\"510\":7,\"511\":1,\"512\":5,\"513\":3,\"514\":4,\"515\":1,\"516\":2,\"517\":3,\"518\":3,\"519\":3,\"520\":1,\"521\":4,\"522\":2,\"523\":1,\"524\":4,\"525\":3,\"526\":2,\"527\":4,\"528\":2,\"529\":1,\"530\":2,\"531\":5,\"532\":3,\"533\":3,\"534\":2,\"535\":1,\"536\":1,\"537\":1,\"538\":1,\"539\":1,\"540\":1,\"541\":3,\"543\":1,\"544\":1,\"545\":3,\"546\":2,\"548\":1,\"549\":3,\"550\":1,\"552\":2,\"556\":2,\"557\":1,\"558\":1,\"559\":3,\"560\":1,\"561\":1,\"562\":2,\"563\":1,\"564\":2,\"565\":1,\"567\":1,\"568\":2,\"569\":2,\"570\":1,\"571\":2,\"574\":1,\"575\":1,\"577\":2,\"579\":2,\"580\":1,\"582\":1,\"584\":1,\"589\":2,\"595\":1,\"598\":1,\"602\":1,\"616\":1,\"626\":1,\"634\":1,\"644\":2,\"648\":2,\"686\":1,\"704\":1,\"706\":1,\"720\":1,\"736\":1},\"seqStats\":{\"numSeqs\":50000,\"numA\":1442938,\"numT\":1334642,\"numG\":864882,\"numC\":1406849,\"numN\":689,\"minSeqLength\":101,\"maxSeqLength\":101,\"accSeqQual\":171761432,\"lengthMap\":{\"101\":50000},\"infoMap\":{\"0\":{\"numA\":13114,\"numT\":12834,\"numG\":9235,\"numC\":14816,\"numN\":1,\"numQual\":50000,\"accQual\":1417823},\"1\":{\"numA\":13914,\"numT\":12955,\"numG\":8942,\"numC\":14187,\"numN\":2,\"numQual\":50000,\"accQual\":1524634},\"2\":{\"numA\":13801,\"numT\":13302,\"numG\":8737,\"numC\":14158,\"numN\":2,\"numQual\":50000,\"accQual\":1531573},\"3\":{\"numA\":13982,\"numT\":12650,\"numG\":9093,\"numC\":14273,\"numN\":2,\"numQual\":50000,\"accQual\":1603701},\"4\":{\"numA\":13769,\"numT\":12897,\"numG\":9216,\"numC\":14113,\"numN\":5,\"numQual\":50000,\"accQual\":1604275},\"5\":{\"numA\":14595,\"numT\":12156,\"numG\":8749,\"numC\":14444,\"numN\":56,\"numQual\":50000,\"accQual\":1604012},\"6\":{\"numA\":13731,\"numT\":12856,\"numG\":8911,\"numC\":14494,\"numN\":8,\"numQual\":50000,\"accQual\":1605734},\"7\":{\"numA\":13726,\"numT\":12114,\"numG\":9326,\"numC\":14830,\"numN\":4,\"numQual\":50000,\"accQual\":1610791},\"8\":{\"numA\":13531,\"numT\":13391,\"numG\":8520,\"numC\":14541,\"numN\":17,\"numQual\":50000,\"accQual\":1644723},\"9\":{\"numA\":14005,\"numT\":13017,\"numG\":8575,\"numC\":14374,\"numN\":29,\"numQual\":50000,\"accQual\":1646394},\"10\":{\"numA\":14131,\"numT\":12892,\"numG\":8840,\"numC\":14117,\"numN\":20,\"numQual\":50000,\"accQual\":1648546},\"11\":{\"numA\":14298,\"numT\":12503,\"numG\":8837,\"numC\":14352,\"numN\":10,\"numQual\":50000,\"accQual\":1651699},\"12\":{\"numA\":14065,\"numT\":12962,\"numG\":8432,\"numC\":14538,\"numN\":3,\"numQual\":50000,\"accQual\":1655589},\"13\":{\"numA\":14047,\"numT\":12949,\"numG\":8969,\"numC\":14033,\"numN\":2,\"numQual\":50000,\"accQual\":1681019},\"14\":{\"numA\":14231,\"numT\":12344,\"numG\":8764,\"numC\":14660,\"numN\":1,\"numQual\":50000,\"accQual\":1680059},\"15\":{\"numA\":14267,\"numT\":12612,\"numG\":8602,\"numC\":14518,\"numN\":1,\"numQual\":50000,\"accQual\":1683982},\"16\":{\"numA\":14190,\"numT\":12779,\"numG\":8852,\"numC\":14179,\"numN\":0,\"numQual\":50000,\"accQual\":1686304},\"17\":{\"numA\":14215,\"numT\":12944,\"numG\":8818,\"numC\":14023,\"numN\":0,\"numQual\":50000,\"accQual\":1686963},\"18\":{\"numA\":14803,\"numT\":12941,\"numG\":8494,\"numC\":13762,\"numN\":0,\"numQual\":50000,\"accQual\":1691422},\"19\":{\"numA\":15000,\"numT\":12394,\"numG\":9008,\"numC\":13597,\"numN\":1,\"numQual\":50000,\"accQual\":1694725},\"20\":{\"numA\":14221,\"numT\":13044,\"numG\":8846,\"numC\":13889,\"numN\":0,\"numQual\":50000,\"accQual\":1697963},\"21\":{\"numA\":14154,\"numT\":13018,\"numG\":8722,\"numC\":14106,\"numN\":0,\"numQual\":50000,\"accQual\":1700263},\"22\":{\"numA\":14569,\"numT\":12482,\"numG\":8780,\"numC\":14168,\"numN\":1,\"numQual\":50000,\"accQual\":1703608},\"23\":{\"numA\":14523,\"numT\":12329,\"numG\":8774,\"numC\":14373,\"numN\":1,\"numQual\":50000,\"accQual\":1704441},\"24\":{\"numA\":14518,\"numT\":12300,\"numG\":8825,\"numC\":14356,\"numN\":1,\"numQual\":50000,\"accQual\":1700684},\"25\":{\"numA\":14646,\"numT\":12670,\"numG\":8697,\"numC\":13986,\"numN\":1,\"numQual\":50000,\"accQual\":1693975},\"26\":{\"numA\":14439,\"numT\":12310,\"numG\":9041,\"numC\":14209,\"numN\":1,\"numQual\":50000,\"accQual\":1712067},\"27\":{\"numA\":14831,\"numT\":12666,\"numG\":8444,\"numC\":14059,\"numN\":0,\"numQual\":50000,\"accQual\":1716168},\"28\":{\"numA\":13821,\"numT\":13426,\"numG\":8884,\"numC\":13869,\"numN\":0,\"numQual\":50000,\"accQual\":1717933},\"29\":{\"numA\":14309,\"numT\":13057,\"numG\":8846,\"numC\":13787,\"numN\":1,\"numQual\":50000,\"accQual\":1723272},\"30\":{\"numA\":14322,\"numT\":13069,\"numG\":8915,\"numC\":13694,\"numN\":0,\"numQual\":50000,\"accQual\":1728078},\"31\":{\"numA\":14348,\"numT\":12686,\"numG\":9116,\"numC\":13849,\"numN\":1,\"numQual\":50000,\"accQual\":1733083},\"32\":{\"numA\":14552,\"numT\":13014,\"numG\":9189,\"numC\":13245,\"numN\":0,\"numQual\":50000,\"accQual\":1740551},\"33\":{\"numA\":14145,\"numT\":13226,\"numG\":9403,\"numC\":13225,\"numN\":1,\"numQual\":50000,\"accQual\":1745647},\"34\":{\"numA\":14350,\"numT\":13330,\"numG\":8963,\"numC\":13357,\"numN\":0,\"numQual\":50000,\"accQual\":1748720},\"35\":{\"numA\":14220,\"numT\":13573,\"numG\":8961,\"numC\":13245,\"numN\":1,\"numQual\":50000,\"accQual\":1753407},\"36\":{\"numA\":13939,\"numT\":13582,\"numG\":8713,\"numC\":13765,\"numN\":1,\"numQual\":50000,\"accQual\":1757274},\"37\":{\"numA\":14389,\"numT\":13658,\"numG\":8783,\"numC\":13170,\"numN\":0,\"numQual\":50000,\"accQual\":1759446},\"38\":{\"numA\":14540,\"numT\":13526,\"numG\":8797,\"numC\":13135,\"numN\":2,\"numQual\":50000,\"accQual\":1762186},\"39\":{\"numA\":14448,\"numT\":13168,\"numG\":8976,\"numC\":13407,\"numN\":1,\"numQual\":50000,\"accQual\":1764038},\"40\":{\"numA\":14902,\"numT\":13406,\"numG\":8570,\"numC\":13121,\"numN\":1,\"numQual\":50000,\"accQual\":1763453},\"41\":{\"numA\":14285,\"numT\":13717,\"numG\":9173,\"numC\":12824,\"numN\":1,\"numQual\":50000,\"accQual\":1763440},\"42\":{\"numA\":13873,\"numT\":13877,\"numG\":8768,\"numC\":13481,\"numN\":1,\"numQual\":50000,\"accQual\":1763896},\"43\":{\"numA\":14807,\"numT\":13462,\"numG\":9084,\"numC\":12644,\"numN\":3,\"numQual\":50000,\"accQual\":1762705},\"44\":{\"numA\":14119,\"numT\":13513,\"numG\":8939,\"numC\":13424,\"numN\":5,\"numQual\":50000,\"accQual\":1764024},\"45\":{\"numA\":14253,\"numT\":13603,\"numG\":9193,\"numC\":12938,\"numN\":13,\"numQual\":50000,\"accQual\":1762316},\"46\":{\"numA\":14329,\"numT\":12852,\"numG\":9551,\"numC\":13259,\"numN\":9,\"numQual\":50000,\"accQual\":1761931},\"47\":{\"numA\":14318,\"numT\":13038,\"numG\":9567,\"numC\":13068,\"numN\":9,\"numQual\":50000,\"accQual\":1758277},\"48\":{\"numA\":14170,\"numT\":13199,\"numG\":9092,\"numC\":13527,\"numN\":12,\"numQual\":50000,\"accQual\":1757486},\"49\":{\"numA\":13725,\"numT\":13953,\"numG\":9286,\"numC\":12961,\"numN\":75,\"numQual\":50000,\"accQual\":1748473},\"50\":{\"numA\":13505,\"numT\":13850,\"numG\":9724,\"numC\":12854,\"numN\":67,\"numQual\":50000,\"accQual\":1730024},\"51\":{\"numA\":14367,\"numT\":13308,\"numG\":9378,\"numC\":12860,\"numN\":87,\"numQual\":50000,\"accQual\":1750670},\"52\":{\"numA\":13854,\"numT\":14007,\"numG\":9242,\"numC\":12887,\"numN\":10,\"numQual\":50000,\"accQual\":1758983},\"53\":{\"numA\":13765,\"numT\":14070,\"numG\":9180,\"numC\":12981,\"numN\":4,\"numQual\":50000,\"accQual\":1764350},\"54\":{\"numA\":13994,\"numT\":13599,\"numG\":9352,\"numC\":13048,\"numN\":7,\"numQual\":50000,\"accQual\":1766364},\"55\":{\"numA\":14281,\"numT\":13602,\"numG\":9053,\"numC\":13055,\"numN\":9,\"numQual\":50000,\"accQual\":1766747},\"56\":{\"numA\":13775,\"numT\":13304,\"numG\":9623,\"numC\":13296,\"numN\":2,\"numQual\":50000,\"accQual\":1766495},\"57\":{\"numA\":13384,\"numT\":13575,\"numG\":9397,\"numC\":13643,\"numN\":1,\"numQual\":50000,\"accQual\":1766439},\"58\":{\"numA\":13610,\"numT\":13674,\"numG\":9206,\"numC\":13509,\"numN\":1,\"numQual\":50000,\"accQual\":1766755},\"59\":{\"numA\":13658,\"numT\":13686,\"numG\":9177,\"numC\":13479,\"numN\":0,\"numQual\":50000,\"accQual\":1765225},\"60\":{\"numA\":13742,\"numT\":13172,\"numG\":9135,\"numC\":13951,\"numN\":0,\"numQual\":50000,\"accQual\":1760282},\"61\":{\"numA\":13911,\"numT\":13328,\"numG\":9064,\"numC\":13697,\"numN\":0,\"numQual\":50000,\"accQual\":1760856},\"62\":{\"numA\":13715,\"numT\":13529,\"numG\":9161,\"numC\":13594,\"numN\":1,\"numQual\":50000,\"accQual\":1756498},\"63\":{\"numA\":14220,\"numT\":12921,\"numG\":9360,\"numC\":13499,\"numN\":0,\"numQual\":50000,\"accQual\":1756737},\"64\":{\"numA\":14172,\"numT\":12654,\"numG\":9091,\"numC\":14083,\"numN\":0,\"numQual\":50000,\"accQual\":1755807},\"65\":{\"numA\":13626,\"numT\":13157,\"numG\":9287,\"numC\":13929,\"numN\":1,\"numQual\":50000,\"accQual\":1751633},\"66\":{\"numA\":13718,\"numT\":13197,\"numG\":9124,\"numC\":13961,\"numN\":0,\"numQual\":50000,\"accQual\":1746861},\"67\":{\"numA\":14128,\"numT\":13422,\"numG\":8930,\"numC\":13520,\"numN\":0,\"numQual\":50000,\"accQual\":1745345},\"68\":{\"numA\":14100,\"numT\":13558,\"numG\":8596,\"numC\":13746,\"numN\":0,\"numQual\":50000,\"accQual\":1741946},\"69\":{\"numA\":14419,\"numT\":13722,\"numG\":8344,\"numC\":13515,\"numN\":0,\"numQual\":50000,\"accQual\":1738799},\"70\":{\"numA\":14577,\"numT\":13385,\"numG\":8181,\"numC\":13857,\"numN\":0,\"numQual\":50000,\"accQual\":1738299},\"71\":{\"numA\":14319,\"numT\":13641,\"numG\":8100,\"numC\":13940,\"numN\":0,\"numQual\":50000,\"accQual\":1735003},\"72\":{\"numA\":14981,\"numT\":13116,\"numG\":7940,\"numC\":13962,\"numN\":1,\"numQual\":50000,\"accQual\":1733700},\"73\":{\"numA\":14570,\"numT\":13068,\"numG\":8182,\"numC\":14179,\"numN\":1,\"numQual\":50000,\"accQual\":1733666},\"74\":{\"numA\":14820,\"numT\":13161,\"numG\":8065,\"numC\":13954,\"numN\":0,\"numQual\":50000,\"accQual\":1729328},\"75\":{\"numA\":14604,\"numT\":13657,\"numG\":7851,\"numC\":13887,\"numN\":1,\"numQual\":50000,\"accQual\":1717724},\"76\":{\"numA\":15435,\"numT\":12987,\"numG\":7598,\"numC\":13980,\"numN\":0,\"numQual\":50000,\"accQual\":1723610},\"77\":{\"numA\":15311,\"numT\":12926,\"numG\":7996,\"numC\":13767,\"numN\":0,\"numQual\":50000,\"accQual\":1724140},\"78\":{\"numA\":15274,\"numT\":12855,\"numG\":7827,\"numC\":14043,\"numN\":1,\"numQual\":50000,\"accQual\":1726561},\"79\":{\"numA\":15325,\"numT\":13140,\"numG\":8079,\"numC\":13455,\"numN\":1,\"numQual\":50000,\"accQual\":1724115},\"80\":{\"numA\":14795,\"numT\":13364,\"numG\":8117,\"numC\":13723,\"numN\":1,\"numQual\":50000,\"accQual\":1723129},\"81\":{\"numA\":15092,\"numT\":13651,\"numG\":7523,\"numC\":13733,\"numN\":1,\"numQual\":50000,\"accQual\":1720192},\"82\":{\"numA\":14855,\"numT\":13590,\"numG\":7709,\"numC\":13845,\"numN\":1,\"numQual\":50000,\"accQual\":1720265},\"83\":{\"numA\":14744,\"numT\":13754,\"numG\":7716,\"numC\":13785,\"numN\":1,\"numQual\":50000,\"accQual\":1711620},\"84\":{\"numA\":14774,\"numT\":13555,\"numG\":7717,\"numC\":13953,\"numN\":1,\"numQual\":50000,\"accQual\":1712764},\"85\":{\"numA\":14936,\"numT\":13569,\"numG\":7357,\"numC\":14137,\"numN\":1,\"numQual\":50000,\"accQual\":1709431},\"86\":{\"numA\":14880,\"numT\":13599,\"numG\":7422,\"numC\":14098,\"numN\":1,\"numQual\":50000,\"accQual\":1708934},\"87\":{\"numA\":14347,\"numT\":14097,\"numG\":7545,\"numC\":14009,\"numN\":2,\"numQual\":50000,\"accQual\":1708208},\"88\":{\"numA\":14439,\"numT\":13456,\"numG\":7753,\"numC\":14350,\"numN\":2,\"numQual\":50000,\"accQual\":1675008},\"89\":{\"numA\":14596,\"numT\":13554,\"numG\":7303,\"numC\":14539,\"numN\":8,\"numQual\":50000,\"accQual\":1672355},\"90\":{\"numA\":14315,\"numT\":13543,\"numG\":7303,\"numC\":14820,\"numN\":19,\"numQual\":50000,\"accQual\":1671957},\"91\":{\"numA\":14312,\"numT\":13495,\"numG\":7555,\"numC\":14596,\"numN\":42,\"numQual\":50000,\"accQual\":1668791},\"92\":{\"numA\":15109,\"numT\":12830,\"numG\":7585,\"numC\":14452,\"numN\":24,\"numQual\":50000,\"accQual\":1668957},\"93\":{\"numA\":14548,\"numT\":13338,\"numG\":7375,\"numC\":14736,\"numN\":3,\"numQual\":50000,\"accQual\":1618078},\"94\":{\"numA\":14482,\"numT\":13267,\"numG\":7134,\"numC\":15101,\"numN\":16,\"numQual\":50000,\"accQual\":1616648},\"95\":{\"numA\":14007,\"numT\":14009,\"numG\":7175,\"numC\":14749,\"numN\":60,\"numQual\":50000,\"accQual\":1618087},\"96\":{\"numA\":13741,\"numT\":13863,\"numG\":7463,\"numC\":14928,\"numN\":5,\"numQual\":50000,\"accQual\":1619403},\"97\":{\"numA\":14292,\"numT\":13586,\"numG\":6688,\"numC\":15434,\"numN\":0,\"numQual\":50000,\"accQual\":1620404},\"98\":{\"numA\":14128,\"numT\":13121,\"numG\":6878,\"numC\":15873,\"numN\":0,\"numQual\":50000,\"accQual\":1513899},\"99\":{\"numA\":14037,\"numT\":13294,\"numG\":7379,\"numC\":15290,\"numN\":0,\"numQual\":50000,\"accQual\":1505438},\"100\":{\"numA\":13869,\"numT\":12590,\"numG\":7124,\"numC\":16417,\"numN\":0,\"numQual\":50000,\"accQual\":1452129}},\"kmers\":{\"kvalue\":0,\"kmersMap\":{}}},\"nm\":86967}'},\n",
       "   'attributes': {'alignmentHeader': {'attributes': {},\n",
       "     'sequenceDiccionary': [{'sequenceName': 'chrM',\n",
       "       'sequenceLength': 16571,\n",
       "       'attributes': {}},\n",
       "      {'sequenceName': 'chr1', 'sequenceLength': 249250621, 'attributes': {}},\n",
       "      {'sequenceName': 'chr2', 'sequenceLength': 243199373, 'attributes': {}},\n",
       "      {'sequenceName': 'chr3', 'sequenceLength': 198022430, 'attributes': {}},\n",
       "      {'sequenceName': 'chr4', 'sequenceLength': 191154276, 'attributes': {}},\n",
       "      {'sequenceName': 'chr5', 'sequenceLength': 180915260, 'attributes': {}},\n",
       "      {'sequenceName': 'chr6', 'sequenceLength': 171115067, 'attributes': {}},\n",
       "      {'sequenceName': 'chr7', 'sequenceLength': 159138663, 'attributes': {}},\n",
       "      {'sequenceName': 'chr8', 'sequenceLength': 146364022, 'attributes': {}},\n",
       "      {'sequenceName': 'chr9', 'sequenceLength': 141213431, 'attributes': {}},\n",
       "      {'sequenceName': 'chr10', 'sequenceLength': 135534747, 'attributes': {}},\n",
       "      {'sequenceName': 'chr11', 'sequenceLength': 135006516, 'attributes': {}},\n",
       "      {'sequenceName': 'chr12', 'sequenceLength': 133851895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr13', 'sequenceLength': 115169878, 'attributes': {}},\n",
       "      {'sequenceName': 'chr14', 'sequenceLength': 107349540, 'attributes': {}},\n",
       "      {'sequenceName': 'chr15', 'sequenceLength': 102531392, 'attributes': {}},\n",
       "      {'sequenceName': 'chr16', 'sequenceLength': 90354753, 'attributes': {}},\n",
       "      {'sequenceName': 'chr17', 'sequenceLength': 81195210, 'attributes': {}},\n",
       "      {'sequenceName': 'chr18', 'sequenceLength': 78077248, 'attributes': {}},\n",
       "      {'sequenceName': 'chr19', 'sequenceLength': 59128983, 'attributes': {}},\n",
       "      {'sequenceName': 'chr20', 'sequenceLength': 63025520, 'attributes': {}},\n",
       "      {'sequenceName': 'chr21', 'sequenceLength': 48129895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr22', 'sequenceLength': 51304566, 'attributes': {}},\n",
       "      {'sequenceName': 'chrX', 'sequenceLength': 155270560, 'attributes': {}},\n",
       "      {'sequenceName': 'chrY', 'sequenceLength': 59373566, 'attributes': {}}],\n",
       "     'readGroups': [{'id': 'NA12877', 'attributes': {'SM': 'NA12877'}}],\n",
       "     'programRecords': [{'id': 'bwa',\n",
       "       'attributes': {'VN': '0.6.1-r104-tpx', 'PN': 'bwa'}}],\n",
       "     'comments': [],\n",
       "     'studyName': 'Platinum',\n",
       "     'origin': 'SAM'}}},\n",
       "  {'id': 'data:NA12877_S1.bam.bai',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12877_S1.bam.bai',\n",
       "   'uuid': 'rOjmHQFnAAMAAU5P-tInaw',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAI',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12877_S1.bam.bai',\n",
       "   'path': 'data/NA12877_S1.bam.bai',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133003',\n",
       "   'modificationDate': '20181214133003',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133003', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 8758280,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'data:NA12877_S1.bam.coverage.bw',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12877_S1.bam.coverage.bw',\n",
       "   'uuid': 'rOjnJQFnAAMAASOr7KbnxA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BIGWIG',\n",
       "   'bioformat': 'COVERAGE',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12877_S1.bam.coverage.bw',\n",
       "   'path': 'data/NA12877_S1.bam.coverage.bw',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133003',\n",
       "   'modificationDate': '20181214133003',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133003', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 309711777,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'data:NA12878_S1.bam',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12878_S1.bam',\n",
       "   'uuid': 'rOjnxQFnAAMAAXp9I0kaCA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAM',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12878_S1.bam',\n",
       "   'path': 'data/NA12878_S1.bam',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20190118234722',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 121691186212,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12878',\n",
       "     'uuid': '7RrrAgFlAAQAAU3OyOO-qA',\n",
       "     'source': 'platinum-genomes-vcf-NA12878_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143448',\n",
       "     'status': {'name': 'READY', 'date': '20180918143448', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {'globalStats': '{\"numMapped\":43322,\"numUnmapped\":6678,\"numPaired\":22316,\"numMappedFirst\":23118,\"numMappedSecond\":20204,\"NM\":83301,\"numHardC\":0,\"numSoftC\":2682,\"numIn\":29,\"numDel\":48,\"numPad\":0,\"numSkip\":0,\"accMappingQuality\":2091750,\"mappingQualityMap\":{\"17\":50,\"50\":2,\"37\":20790,\"23\":1,\"25\":215,\"60\":21630,\"29\":633,\"15\":1},\"accInsert\":6717036,\"insertMap\":{\"101\":1,\"102\":4,\"103\":16,\"104\":12,\"105\":6,\"106\":8,\"107\":10,\"108\":18,\"109\":14,\"110\":17,\"112\":8,\"113\":16,\"114\":18,\"115\":12,\"116\":14,\"117\":14,\"118\":15,\"119\":20,\"120\":22,\"121\":16,\"122\":18,\"123\":9,\"124\":23,\"125\":6,\"126\":11,\"127\":17,\"128\":8,\"129\":33,\"130\":21,\"131\":10,\"132\":16,\"133\":5,\"134\":21,\"135\":21,\"136\":20,\"137\":28,\"138\":16,\"139\":24,\"140\":20,\"141\":24,\"142\":13,\"143\":27,\"144\":24,\"145\":16,\"146\":22,\"147\":15,\"148\":13,\"149\":25,\"150\":14,\"151\":25,\"152\":13,\"153\":28,\"154\":22,\"155\":21,\"156\":28,\"157\":16,\"158\":21,\"159\":30,\"160\":29,\"161\":23,\"162\":27,\"163\":25,\"164\":23,\"165\":29,\"166\":34,\"167\":24,\"168\":32,\"169\":30,\"170\":18,\"171\":43,\"172\":32,\"173\":33,\"174\":44,\"175\":35,\"176\":45,\"177\":29,\"178\":30,\"179\":28,\"180\":26,\"181\":34,\"182\":35,\"183\":26,\"184\":29,\"185\":39,\"186\":28,\"187\":45,\"188\":31,\"189\":42,\"190\":35,\"191\":45,\"192\":28,\"193\":36,\"194\":37,\"195\":52,\"196\":34,\"197\":27,\"198\":49,\"199\":33,\"200\":55,\"201\":22,\"202\":34,\"203\":37,\"204\":50,\"205\":44,\"206\":30,\"207\":51,\"208\":47,\"209\":49,\"210\":46,\"211\":44,\"212\":40,\"213\":35,\"214\":30,\"215\":53,\"216\":45,\"217\":44,\"218\":54,\"219\":38,\"220\":64,\"221\":36,\"222\":66,\"223\":53,\"224\":49,\"225\":54,\"226\":58,\"227\":58,\"228\":58,\"229\":65,\"230\":61,\"231\":58,\"232\":71,\"233\":73,\"234\":75,\"235\":71,\"236\":59,\"237\":77,\"238\":66,\"239\":84,\"240\":64,\"241\":93,\"242\":85,\"243\":112,\"244\":76,\"245\":75,\"246\":106,\"247\":82,\"248\":97,\"249\":82,\"250\":106,\"251\":88,\"252\":77,\"253\":100,\"254\":86,\"255\":82,\"256\":116,\"257\":97,\"258\":120,\"259\":105,\"260\":101,\"261\":100,\"262\":114,\"263\":87,\"264\":116,\"265\":94,\"266\":97,\"267\":124,\"268\":137,\"269\":131,\"270\":98,\"271\":110,\"272\":120,\"273\":125,\"274\":98,\"275\":110,\"276\":90,\"277\":100,\"278\":136,\"279\":111,\"280\":126,\"281\":132,\"282\":103,\"283\":111,\"284\":94,\"285\":93,\"286\":117,\"287\":124,\"288\":130,\"289\":134,\"290\":113,\"291\":123,\"292\":144,\"293\":124,\"294\":102,\"295\":127,\"296\":130,\"297\":128,\"298\":119,\"299\":125,\"300\":120,\"301\":128,\"302\":123,\"303\":131,\"304\":122,\"305\":118,\"306\":125,\"307\":125,\"308\":135,\"309\":117,\"310\":121,\"311\":114,\"312\":128,\"313\":113,\"314\":117,\"315\":109,\"316\":123,\"317\":126,\"318\":121,\"319\":129,\"320\":101,\"321\":112,\"322\":125,\"323\":110,\"324\":117,\"325\":109,\"326\":117,\"327\":112,\"328\":118,\"329\":133,\"330\":101,\"331\":124,\"332\":120,\"333\":110,\"334\":110,\"335\":99,\"336\":96,\"337\":109,\"338\":114,\"339\":93,\"340\":104,\"341\":112,\"342\":117,\"343\":116,\"344\":108,\"345\":101,\"346\":106,\"347\":98,\"348\":114,\"349\":85,\"350\":87,\"351\":95,\"352\":115,\"353\":102,\"354\":97,\"355\":92,\"356\":81,\"357\":94,\"358\":81,\"359\":91,\"360\":108,\"361\":80,\"362\":87,\"363\":95,\"364\":81,\"365\":80,\"366\":79,\"367\":79,\"368\":77,\"369\":64,\"370\":76,\"371\":73,\"372\":68,\"373\":84,\"374\":79,\"375\":86,\"376\":78,\"377\":67,\"378\":64,\"379\":56,\"380\":64,\"381\":67,\"382\":63,\"383\":52,\"384\":62,\"385\":62,\"386\":66,\"387\":61,\"388\":45,\"389\":48,\"390\":49,\"391\":49,\"392\":59,\"393\":46,\"394\":56,\"395\":65,\"396\":60,\"397\":32,\"398\":61,\"399\":57,\"400\":40,\"401\":51,\"402\":52,\"403\":51,\"404\":41,\"405\":54,\"406\":28,\"407\":36,\"408\":41,\"409\":44,\"410\":38,\"411\":51,\"412\":52,\"413\":37,\"414\":29,\"415\":24,\"416\":35,\"417\":38,\"418\":38,\"419\":35,\"420\":33,\"421\":25,\"422\":31,\"423\":30,\"424\":24,\"425\":25,\"426\":31,\"427\":33,\"428\":21,\"429\":30,\"430\":32,\"431\":27,\"432\":19,\"433\":18,\"434\":24,\"435\":18,\"436\":21,\"437\":24,\"438\":21,\"439\":15,\"440\":26,\"441\":11,\"442\":16,\"443\":11,\"444\":10,\"445\":12,\"446\":11,\"447\":14,\"448\":14,\"449\":9,\"450\":17,\"451\":15,\"452\":15,\"453\":16,\"454\":11,\"455\":15,\"456\":18,\"457\":11,\"458\":15,\"459\":14,\"460\":12,\"461\":12,\"462\":7,\"463\":14,\"464\":15,\"465\":8,\"466\":12,\"467\":18,\"468\":10,\"469\":8,\"470\":12,\"471\":7,\"472\":11,\"473\":7,\"474\":3,\"475\":10,\"476\":8,\"477\":7,\"478\":2,\"479\":12,\"480\":8,\"481\":9,\"482\":5,\"483\":3,\"484\":8,\"485\":5,\"486\":7,\"487\":7,\"488\":7,\"489\":5,\"490\":9,\"491\":3,\"492\":3,\"493\":9,\"494\":3,\"495\":3,\"496\":4,\"497\":4,\"498\":2,\"499\":5,\"500\":3,\"501\":6,\"502\":6,\"503\":4,\"504\":2,\"505\":3,\"506\":5,\"507\":2,\"508\":8,\"509\":1,\"510\":3,\"511\":5,\"512\":6,\"513\":5,\"514\":5,\"515\":5,\"516\":1,\"517\":3,\"518\":6,\"519\":5,\"520\":3,\"521\":2,\"522\":2,\"523\":3,\"524\":3,\"525\":4,\"526\":1,\"527\":1,\"528\":2,\"529\":2,\"530\":3,\"531\":3,\"532\":5,\"533\":3,\"534\":2,\"536\":2,\"537\":1,\"538\":2,\"539\":1,\"540\":1,\"542\":2,\"543\":1,\"547\":1,\"549\":3,\"551\":2,\"553\":2,\"554\":2,\"556\":1,\"557\":2,\"558\":1,\"560\":1,\"561\":1,\"562\":2,\"565\":1,\"567\":3,\"568\":1,\"571\":1,\"572\":1,\"573\":2,\"574\":2,\"577\":1,\"579\":1,\"580\":1,\"581\":2,\"585\":1,\"587\":1,\"595\":1,\"598\":1,\"600\":1,\"607\":1,\"611\":1,\"613\":1,\"614\":1,\"615\":1,\"617\":2,\"620\":1,\"622\":1,\"623\":2,\"645\":1,\"648\":1,\"669\":1,\"671\":1,\"672\":1,\"680\":1,\"681\":1,\"687\":1,\"705\":1,\"715\":1,\"717\":1,\"733\":1,\"754\":1,\"789\":1,\"797\":1},\"seqStats\":{\"numSeqs\":50000,\"numA\":1295017,\"numT\":1498230,\"numG\":1079283,\"numC\":1173199,\"numN\":4271,\"minSeqLength\":101,\"maxSeqLength\":101,\"accSeqQual\":180042817,\"lengthMap\":{\"101\":50000},\"infoMap\":{\"0\":{\"numA\":11349,\"numT\":14351,\"numG\":11217,\"numC\":13056,\"numN\":27,\"numQual\":50000,\"accQual\":1535441},\"1\":{\"numA\":11858,\"numT\":14700,\"numG\":11715,\"numC\":11692,\"numN\":35,\"numQual\":50000,\"accQual\":1603067},\"2\":{\"numA\":11831,\"numT\":15092,\"numG\":10572,\"numC\":12469,\"numN\":36,\"numQual\":50000,\"accQual\":1604625},\"3\":{\"numA\":11600,\"numT\":14835,\"numG\":11643,\"numC\":11879,\"numN\":43,\"numQual\":50000,\"accQual\":1673284},\"4\":{\"numA\":11769,\"numT\":15042,\"numG\":11355,\"numC\":11794,\"numN\":40,\"numQual\":50000,\"accQual\":1672329},\"5\":{\"numA\":11757,\"numT\":14802,\"numG\":11123,\"numC\":12275,\"numN\":43,\"numQual\":50000,\"accQual\":1671311},\"6\":{\"numA\":11354,\"numT\":15345,\"numG\":11096,\"numC\":12162,\"numN\":43,\"numQual\":50000,\"accQual\":1673744},\"7\":{\"numA\":11419,\"numT\":14483,\"numG\":11019,\"numC\":13039,\"numN\":40,\"numQual\":50000,\"accQual\":1671305},\"8\":{\"numA\":11297,\"numT\":15048,\"numG\":10946,\"numC\":12665,\"numN\":44,\"numQual\":50000,\"accQual\":1705364},\"9\":{\"numA\":11094,\"numT\":15644,\"numG\":10672,\"numC\":12548,\"numN\":42,\"numQual\":50000,\"accQual\":1707707},\"10\":{\"numA\":11785,\"numT\":14343,\"numG\":11170,\"numC\":12661,\"numN\":41,\"numQual\":50000,\"accQual\":1708505},\"11\":{\"numA\":11901,\"numT\":15117,\"numG\":10975,\"numC\":11967,\"numN\":40,\"numQual\":50000,\"accQual\":1709880},\"12\":{\"numA\":11358,\"numT\":15176,\"numG\":11075,\"numC\":12352,\"numN\":39,\"numQual\":50000,\"accQual\":1713322},\"13\":{\"numA\":11408,\"numT\":15208,\"numG\":10862,\"numC\":12482,\"numN\":40,\"numQual\":50000,\"accQual\":1741793},\"14\":{\"numA\":11414,\"numT\":14930,\"numG\":11273,\"numC\":12343,\"numN\":40,\"numQual\":50000,\"accQual\":1739161},\"15\":{\"numA\":11597,\"numT\":14454,\"numG\":11408,\"numC\":12499,\"numN\":42,\"numQual\":50000,\"accQual\":1743891},\"16\":{\"numA\":11643,\"numT\":14632,\"numG\":11941,\"numC\":11742,\"numN\":42,\"numQual\":50000,\"accQual\":1744951},\"17\":{\"numA\":11886,\"numT\":14560,\"numG\":11703,\"numC\":11817,\"numN\":34,\"numQual\":50000,\"accQual\":1741812},\"18\":{\"numA\":11209,\"numT\":14924,\"numG\":11694,\"numC\":12124,\"numN\":49,\"numQual\":50000,\"accQual\":1746814},\"19\":{\"numA\":11880,\"numT\":14670,\"numG\":11686,\"numC\":11733,\"numN\":31,\"numQual\":50000,\"accQual\":1745571},\"20\":{\"numA\":12055,\"numT\":14560,\"numG\":12032,\"numC\":11313,\"numN\":40,\"numQual\":50000,\"accQual\":1747914},\"21\":{\"numA\":11514,\"numT\":15213,\"numG\":11333,\"numC\":11902,\"numN\":38,\"numQual\":50000,\"accQual\":1754018},\"22\":{\"numA\":11887,\"numT\":14805,\"numG\":11397,\"numC\":11872,\"numN\":39,\"numQual\":50000,\"accQual\":1756314},\"23\":{\"numA\":12339,\"numT\":14639,\"numG\":11356,\"numC\":11628,\"numN\":38,\"numQual\":50000,\"accQual\":1754636},\"24\":{\"numA\":11770,\"numT\":14887,\"numG\":11780,\"numC\":11526,\"numN\":37,\"numQual\":50000,\"accQual\":1745509},\"25\":{\"numA\":11987,\"numT\":14953,\"numG\":11608,\"numC\":11413,\"numN\":39,\"numQual\":50000,\"accQual\":1735158},\"26\":{\"numA\":12125,\"numT\":15370,\"numG\":11665,\"numC\":10801,\"numN\":39,\"numQual\":50000,\"accQual\":1769404},\"27\":{\"numA\":12246,\"numT\":14016,\"numG\":12148,\"numC\":11551,\"numN\":39,\"numQual\":50000,\"accQual\":1774846},\"28\":{\"numA\":12380,\"numT\":14938,\"numG\":11557,\"numC\":11085,\"numN\":40,\"numQual\":50000,\"accQual\":1781322},\"29\":{\"numA\":12369,\"numT\":14239,\"numG\":11684,\"numC\":11668,\"numN\":40,\"numQual\":50000,\"accQual\":1789192},\"30\":{\"numA\":12291,\"numT\":14145,\"numG\":12378,\"numC\":11146,\"numN\":40,\"numQual\":50000,\"accQual\":1794160},\"31\":{\"numA\":12609,\"numT\":14547,\"numG\":11567,\"numC\":11236,\"numN\":41,\"numQual\":50000,\"accQual\":1797417},\"32\":{\"numA\":12375,\"numT\":14958,\"numG\":11640,\"numC\":10986,\"numN\":41,\"numQual\":50000,\"accQual\":1803934},\"33\":{\"numA\":12387,\"numT\":14657,\"numG\":12203,\"numC\":10708,\"numN\":45,\"numQual\":50000,\"accQual\":1807672},\"34\":{\"numA\":12593,\"numT\":15230,\"numG\":11330,\"numC\":10803,\"numN\":44,\"numQual\":50000,\"accQual\":1814240},\"35\":{\"numA\":12647,\"numT\":15008,\"numG\":11521,\"numC\":10777,\"numN\":47,\"numQual\":50000,\"accQual\":1817799},\"36\":{\"numA\":13041,\"numT\":14864,\"numG\":11504,\"numC\":10546,\"numN\":45,\"numQual\":50000,\"accQual\":1820887},\"37\":{\"numA\":12292,\"numT\":15466,\"numG\":11259,\"numC\":10940,\"numN\":43,\"numQual\":50000,\"accQual\":1821325},\"38\":{\"numA\":12441,\"numT\":15826,\"numG\":11605,\"numC\":10082,\"numN\":46,\"numQual\":50000,\"accQual\":1824184},\"39\":{\"numA\":12219,\"numT\":15195,\"numG\":12132,\"numC\":10409,\"numN\":45,\"numQual\":50000,\"accQual\":1825587},\"40\":{\"numA\":12882,\"numT\":15514,\"numG\":10847,\"numC\":10714,\"numN\":43,\"numQual\":50000,\"accQual\":1828022},\"41\":{\"numA\":12408,\"numT\":15291,\"numG\":11424,\"numC\":10837,\"numN\":40,\"numQual\":50000,\"accQual\":1828958},\"42\":{\"numA\":12593,\"numT\":15290,\"numG\":11635,\"numC\":10444,\"numN\":38,\"numQual\":50000,\"accQual\":1832497},\"43\":{\"numA\":12381,\"numT\":15448,\"numG\":11894,\"numC\":10237,\"numN\":40,\"numQual\":50000,\"accQual\":1832411},\"44\":{\"numA\":12610,\"numT\":15489,\"numG\":11683,\"numC\":10176,\"numN\":42,\"numQual\":50000,\"accQual\":1828984},\"45\":{\"numA\":12557,\"numT\":14870,\"numG\":11985,\"numC\":10556,\"numN\":32,\"numQual\":50000,\"accQual\":1834359},\"46\":{\"numA\":12687,\"numT\":15210,\"numG\":11946,\"numC\":10121,\"numN\":36,\"numQual\":50000,\"accQual\":1834272},\"47\":{\"numA\":12498,\"numT\":15602,\"numG\":11170,\"numC\":10695,\"numN\":35,\"numQual\":50000,\"accQual\":1835123},\"48\":{\"numA\":12264,\"numT\":15716,\"numG\":11637,\"numC\":10349,\"numN\":34,\"numQual\":50000,\"accQual\":1831688},\"49\":{\"numA\":11590,\"numT\":15478,\"numG\":11712,\"numC\":11180,\"numN\":40,\"numQual\":50000,\"accQual\":1824039},\"50\":{\"numA\":11812,\"numT\":15143,\"numG\":11958,\"numC\":11046,\"numN\":41,\"numQual\":50000,\"accQual\":1792819},\"51\":{\"numA\":12064,\"numT\":15024,\"numG\":12400,\"numC\":10473,\"numN\":39,\"numQual\":50000,\"accQual\":1837790},\"52\":{\"numA\":11586,\"numT\":14776,\"numG\":12645,\"numC\":10955,\"numN\":38,\"numQual\":50000,\"accQual\":1844463},\"53\":{\"numA\":12137,\"numT\":15174,\"numG\":11099,\"numC\":11550,\"numN\":40,\"numQual\":50000,\"accQual\":1846800},\"54\":{\"numA\":11968,\"numT\":15407,\"numG\":11449,\"numC\":11137,\"numN\":39,\"numQual\":50000,\"accQual\":1847654},\"55\":{\"numA\":11973,\"numT\":15484,\"numG\":11675,\"numC\":10835,\"numN\":33,\"numQual\":50000,\"accQual\":1848889},\"56\":{\"numA\":12203,\"numT\":14613,\"numG\":11559,\"numC\":11579,\"numN\":46,\"numQual\":50000,\"accQual\":1848552},\"57\":{\"numA\":12761,\"numT\":14605,\"numG\":11516,\"numC\":11075,\"numN\":43,\"numQual\":50000,\"accQual\":1850451},\"58\":{\"numA\":11886,\"numT\":15042,\"numG\":11790,\"numC\":11237,\"numN\":45,\"numQual\":50000,\"accQual\":1852237},\"59\":{\"numA\":11979,\"numT\":14377,\"numG\":11549,\"numC\":12050,\"numN\":45,\"numQual\":50000,\"accQual\":1851082},\"60\":{\"numA\":12747,\"numT\":13752,\"numG\":11827,\"numC\":11629,\"numN\":45,\"numQual\":50000,\"accQual\":1846205},\"61\":{\"numA\":12428,\"numT\":14174,\"numG\":11322,\"numC\":12026,\"numN\":50,\"numQual\":50000,\"accQual\":1850590},\"62\":{\"numA\":13199,\"numT\":14422,\"numG\":10954,\"numC\":11373,\"numN\":52,\"numQual\":50000,\"accQual\":1847478},\"63\":{\"numA\":12267,\"numT\":15139,\"numG\":11045,\"numC\":11498,\"numN\":51,\"numQual\":50000,\"accQual\":1849988},\"64\":{\"numA\":12691,\"numT\":14862,\"numG\":10910,\"numC\":11484,\"numN\":53,\"numQual\":50000,\"accQual\":1851899},\"65\":{\"numA\":13268,\"numT\":14311,\"numG\":10697,\"numC\":11669,\"numN\":55,\"numQual\":50000,\"accQual\":1848388},\"66\":{\"numA\":12861,\"numT\":15178,\"numG\":10761,\"numC\":11150,\"numN\":50,\"numQual\":50000,\"accQual\":1850017},\"67\":{\"numA\":13485,\"numT\":14256,\"numG\":10774,\"numC\":11432,\"numN\":53,\"numQual\":50000,\"accQual\":1849224},\"68\":{\"numA\":13370,\"numT\":15199,\"numG\":10245,\"numC\":11135,\"numN\":51,\"numQual\":50000,\"accQual\":1846746},\"69\":{\"numA\":13632,\"numT\":14664,\"numG\":9865,\"numC\":11790,\"numN\":49,\"numQual\":50000,\"accQual\":1845358},\"70\":{\"numA\":13280,\"numT\":14515,\"numG\":10674,\"numC\":11483,\"numN\":48,\"numQual\":50000,\"accQual\":1843626},\"71\":{\"numA\":12969,\"numT\":14906,\"numG\":9699,\"numC\":12377,\"numN\":49,\"numQual\":50000,\"accQual\":1840287},\"72\":{\"numA\":13775,\"numT\":14165,\"numG\":10745,\"numC\":11271,\"numN\":44,\"numQual\":50000,\"accQual\":1839618},\"73\":{\"numA\":14032,\"numT\":14319,\"numG\":10068,\"numC\":11539,\"numN\":42,\"numQual\":50000,\"accQual\":1836277},\"74\":{\"numA\":14010,\"numT\":14248,\"numG\":10175,\"numC\":11525,\"numN\":42,\"numQual\":50000,\"accQual\":1832447},\"75\":{\"numA\":14565,\"numT\":14268,\"numG\":9291,\"numC\":11839,\"numN\":37,\"numQual\":50000,\"accQual\":1812671},\"76\":{\"numA\":14712,\"numT\":14374,\"numG\":9178,\"numC\":11697,\"numN\":39,\"numQual\":50000,\"accQual\":1822503},\"77\":{\"numA\":14382,\"numT\":14838,\"numG\":9228,\"numC\":11512,\"numN\":40,\"numQual\":50000,\"accQual\":1827645},\"78\":{\"numA\":14578,\"numT\":14899,\"numG\":9304,\"numC\":11179,\"numN\":40,\"numQual\":50000,\"accQual\":1830471},\"79\":{\"numA\":14319,\"numT\":14720,\"numG\":9458,\"numC\":11464,\"numN\":39,\"numQual\":50000,\"accQual\":1832340},\"80\":{\"numA\":13700,\"numT\":15325,\"numG\":9645,\"numC\":11292,\"numN\":38,\"numQual\":50000,\"accQual\":1832088},\"81\":{\"numA\":13709,\"numT\":15383,\"numG\":9339,\"numC\":11533,\"numN\":36,\"numQual\":50000,\"accQual\":1829575},\"82\":{\"numA\":14430,\"numT\":15018,\"numG\":9181,\"numC\":11316,\"numN\":55,\"numQual\":50000,\"accQual\":1827427},\"83\":{\"numA\":14105,\"numT\":14940,\"numG\":8938,\"numC\":11983,\"numN\":34,\"numQual\":50000,\"accQual\":1823012},\"84\":{\"numA\":14599,\"numT\":14963,\"numG\":8973,\"numC\":11421,\"numN\":44,\"numQual\":50000,\"accQual\":1827829},\"85\":{\"numA\":14604,\"numT\":14443,\"numG\":9275,\"numC\":11628,\"numN\":50,\"numQual\":50000,\"accQual\":1826685},\"86\":{\"numA\":14953,\"numT\":14598,\"numG\":8673,\"numC\":11729,\"numN\":47,\"numQual\":50000,\"accQual\":1824029},\"87\":{\"numA\":14861,\"numT\":15048,\"numG\":8397,\"numC\":11647,\"numN\":47,\"numQual\":50000,\"accQual\":1824377},\"88\":{\"numA\":14143,\"numT\":15197,\"numG\":8961,\"numC\":11651,\"numN\":48,\"numQual\":50000,\"accQual\":1779810},\"89\":{\"numA\":14393,\"numT\":14253,\"numG\":9015,\"numC\":12291,\"numN\":48,\"numQual\":50000,\"accQual\":1777853},\"90\":{\"numA\":14591,\"numT\":14456,\"numG\":8722,\"numC\":12183,\"numN\":48,\"numQual\":50000,\"accQual\":1776432},\"91\":{\"numA\":14675,\"numT\":14941,\"numG\":8506,\"numC\":11829,\"numN\":49,\"numQual\":50000,\"accQual\":1774707},\"92\":{\"numA\":14734,\"numT\":14457,\"numG\":9447,\"numC\":11313,\"numN\":49,\"numQual\":50000,\"accQual\":1775927},\"93\":{\"numA\":13954,\"numT\":14949,\"numG\":8351,\"numC\":12698,\"numN\":48,\"numQual\":50000,\"accQual\":1722820},\"94\":{\"numA\":14619,\"numT\":13744,\"numG\":8392,\"numC\":13197,\"numN\":48,\"numQual\":50000,\"accQual\":1722860},\"95\":{\"numA\":13988,\"numT\":14116,\"numG\":8556,\"numC\":13288,\"numN\":52,\"numQual\":50000,\"accQual\":1722755},\"96\":{\"numA\":13889,\"numT\":14353,\"numG\":7999,\"numC\":13718,\"numN\":41,\"numQual\":50000,\"accQual\":1721547},\"97\":{\"numA\":15226,\"numT\":14367,\"numG\":7696,\"numC\":12668,\"numN\":43,\"numQual\":50000,\"accQual\":1722040},\"98\":{\"numA\":14145,\"numT\":14879,\"numG\":8161,\"numC\":12777,\"numN\":38,\"numQual\":50000,\"accQual\":1616542},\"99\":{\"numA\":14396,\"numT\":14567,\"numG\":6998,\"numC\":13999,\"numN\":40,\"numQual\":50000,\"accQual\":1615488},\"100\":{\"numA\":14918,\"numT\":14599,\"numG\":7421,\"numC\":13029,\"numN\":33,\"numQual\":50000,\"accQual\":1574751}},\"kmers\":{\"kvalue\":0,\"kmersMap\":{}}},\"nm\":83301}'},\n",
       "   'attributes': {'alignmentHeader': {'attributes': {},\n",
       "     'sequenceDiccionary': [{'sequenceName': 'chrM',\n",
       "       'sequenceLength': 16571,\n",
       "       'attributes': {}},\n",
       "      {'sequenceName': 'chr1', 'sequenceLength': 249250621, 'attributes': {}},\n",
       "      {'sequenceName': 'chr2', 'sequenceLength': 243199373, 'attributes': {}},\n",
       "      {'sequenceName': 'chr3', 'sequenceLength': 198022430, 'attributes': {}},\n",
       "      {'sequenceName': 'chr4', 'sequenceLength': 191154276, 'attributes': {}},\n",
       "      {'sequenceName': 'chr5', 'sequenceLength': 180915260, 'attributes': {}},\n",
       "      {'sequenceName': 'chr6', 'sequenceLength': 171115067, 'attributes': {}},\n",
       "      {'sequenceName': 'chr7', 'sequenceLength': 159138663, 'attributes': {}},\n",
       "      {'sequenceName': 'chr8', 'sequenceLength': 146364022, 'attributes': {}},\n",
       "      {'sequenceName': 'chr9', 'sequenceLength': 141213431, 'attributes': {}},\n",
       "      {'sequenceName': 'chr10', 'sequenceLength': 135534747, 'attributes': {}},\n",
       "      {'sequenceName': 'chr11', 'sequenceLength': 135006516, 'attributes': {}},\n",
       "      {'sequenceName': 'chr12', 'sequenceLength': 133851895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr13', 'sequenceLength': 115169878, 'attributes': {}},\n",
       "      {'sequenceName': 'chr14', 'sequenceLength': 107349540, 'attributes': {}},\n",
       "      {'sequenceName': 'chr15', 'sequenceLength': 102531392, 'attributes': {}},\n",
       "      {'sequenceName': 'chr16', 'sequenceLength': 90354753, 'attributes': {}},\n",
       "      {'sequenceName': 'chr17', 'sequenceLength': 81195210, 'attributes': {}},\n",
       "      {'sequenceName': 'chr18', 'sequenceLength': 78077248, 'attributes': {}},\n",
       "      {'sequenceName': 'chr19', 'sequenceLength': 59128983, 'attributes': {}},\n",
       "      {'sequenceName': 'chr20', 'sequenceLength': 63025520, 'attributes': {}},\n",
       "      {'sequenceName': 'chr21', 'sequenceLength': 48129895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr22', 'sequenceLength': 51304566, 'attributes': {}},\n",
       "      {'sequenceName': 'chrX', 'sequenceLength': 155270560, 'attributes': {}},\n",
       "      {'sequenceName': 'chrY', 'sequenceLength': 59373566, 'attributes': {}}],\n",
       "     'readGroups': [{'id': 'NA12878', 'attributes': {'SM': 'NA12878'}}],\n",
       "     'programRecords': [{'id': 'bwa',\n",
       "       'attributes': {'VN': '0.6.1-r104-tpx', 'PN': 'bwa'}}],\n",
       "     'comments': [],\n",
       "     'studyName': 'Platinum',\n",
       "     'origin': 'SAM'}}},\n",
       "  {'id': 'data:NA12878_S1.bam.bai',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12878_S1.bam.bai',\n",
       "   'uuid': 'rOjotgFnAAMAAariF0UlZg',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAI',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12878_S1.bam.bai',\n",
       "   'path': 'data/NA12878_S1.bam.bai',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20181214133004',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 8750352,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'data:NA12878_S1.bam.coverage.bw',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12878_S1.bam.coverage.bw',\n",
       "   'uuid': 'rOjpMgFnAAMAAcK2u51t4Q',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BIGWIG',\n",
       "   'bioformat': 'COVERAGE',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12878_S1.bam.coverage.bw',\n",
       "   'path': 'data/NA12878_S1.bam.coverage.bw',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20181214133004',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 307413923,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'data:NA12879_S1.bam',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12879_S1.bam',\n",
       "   'uuid': 'rOjpoQFnAAMAAY6izEvxGw',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAM',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12879_S1.bam',\n",
       "   'path': 'data/NA12879_S1.bam',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20190122052441',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 125570264101,\n",
       "   'experiment': {},\n",
       "   'samples': [{'id': 'NA12879',\n",
       "     'uuid': '7Rrs-wFlAAQAAQMADI03VQ',\n",
       "     'source': 'platinum-genomes-vcf-NA12879_S1.genome.vcf.gz',\n",
       "     'individual': {'release': 0,\n",
       "      'version': 0,\n",
       "      'parentalConsanguinity': False},\n",
       "     'release': 1,\n",
       "     'version': 1,\n",
       "     'creationDate': '20180918143449',\n",
       "     'status': {'name': 'READY', 'date': '20180918143449', 'message': ''},\n",
       "     'description': '',\n",
       "     'type': '',\n",
       "     'somatic': False,\n",
       "     'phenotypes': [],\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {'globalStats': '{\"numMapped\":42946,\"numUnmapped\":7054,\"numPaired\":25576,\"numMappedFirst\":22864,\"numMappedSecond\":20082,\"NM\":78353,\"numHardC\":0,\"numSoftC\":5790,\"numIn\":606,\"numDel\":187,\"numPad\":0,\"numSkip\":0,\"accMappingQuality\":2131165,\"mappingQualityMap\":{\"48\":2,\"17\":38,\"50\":4,\"37\":17225,\"25\":150,\"60\":24161,\"29\":1357,\"15\":9},\"accInsert\":7452528,\"insertMap\":{\"66\":1,\"90\":1,\"102\":8,\"103\":4,\"104\":2,\"105\":6,\"107\":6,\"108\":2,\"109\":2,\"110\":2,\"111\":2,\"112\":2,\"113\":2,\"114\":3,\"115\":6,\"116\":2,\"119\":4,\"120\":7,\"121\":6,\"122\":10,\"123\":10,\"124\":4,\"126\":6,\"127\":3,\"128\":6,\"129\":10,\"130\":6,\"131\":3,\"132\":14,\"133\":4,\"134\":10,\"135\":8,\"136\":10,\"137\":6,\"138\":5,\"139\":10,\"140\":12,\"141\":12,\"142\":6,\"143\":8,\"144\":8,\"145\":9,\"146\":4,\"147\":16,\"148\":15,\"149\":10,\"150\":11,\"151\":6,\"152\":4,\"153\":13,\"154\":12,\"155\":8,\"156\":17,\"157\":9,\"158\":13,\"159\":2,\"160\":14,\"161\":16,\"162\":18,\"163\":35,\"164\":17,\"165\":22,\"166\":27,\"167\":20,\"168\":27,\"169\":23,\"170\":25,\"171\":20,\"172\":9,\"173\":41,\"174\":25,\"175\":37,\"176\":20,\"177\":20,\"178\":32,\"179\":38,\"180\":26,\"181\":25,\"182\":23,\"183\":26,\"184\":26,\"185\":32,\"186\":36,\"187\":29,\"188\":52,\"189\":38,\"190\":27,\"191\":18,\"192\":40,\"193\":40,\"194\":45,\"195\":36,\"196\":43,\"197\":50,\"198\":41,\"199\":51,\"200\":50,\"201\":52,\"202\":46,\"203\":50,\"204\":57,\"205\":53,\"206\":79,\"207\":50,\"208\":65,\"209\":66,\"210\":57,\"211\":73,\"212\":57,\"213\":66,\"214\":68,\"215\":84,\"216\":63,\"217\":77,\"218\":71,\"219\":62,\"220\":83,\"221\":81,\"222\":81,\"223\":81,\"224\":90,\"225\":105,\"226\":93,\"227\":88,\"228\":91,\"229\":105,\"230\":102,\"231\":104,\"232\":113,\"233\":82,\"234\":129,\"235\":143,\"236\":131,\"237\":116,\"238\":133,\"239\":172,\"240\":104,\"241\":129,\"242\":147,\"243\":129,\"244\":124,\"245\":126,\"246\":127,\"247\":139,\"248\":141,\"249\":145,\"250\":178,\"251\":120,\"252\":147,\"253\":173,\"254\":125,\"255\":159,\"256\":161,\"257\":156,\"258\":185,\"259\":158,\"260\":155,\"261\":143,\"262\":184,\"263\":169,\"264\":170,\"265\":174,\"266\":153,\"267\":196,\"268\":165,\"269\":155,\"270\":212,\"271\":162,\"272\":149,\"273\":160,\"274\":194,\"275\":206,\"276\":185,\"277\":197,\"278\":162,\"279\":154,\"280\":162,\"281\":181,\"282\":185,\"283\":144,\"284\":186,\"285\":150,\"286\":182,\"287\":178,\"288\":160,\"289\":179,\"290\":196,\"291\":193,\"292\":143,\"293\":189,\"294\":187,\"295\":192,\"296\":157,\"297\":142,\"298\":187,\"299\":184,\"300\":160,\"301\":184,\"302\":156,\"303\":184,\"304\":191,\"305\":155,\"306\":185,\"307\":189,\"308\":178,\"309\":162,\"310\":154,\"311\":170,\"312\":172,\"313\":180,\"314\":157,\"315\":145,\"316\":122,\"317\":152,\"318\":156,\"319\":132,\"320\":142,\"321\":136,\"322\":136,\"323\":117,\"324\":121,\"325\":138,\"326\":106,\"327\":132,\"328\":111,\"329\":128,\"330\":130,\"331\":126,\"332\":130,\"333\":113,\"334\":118,\"335\":108,\"336\":102,\"337\":118,\"338\":107,\"339\":119,\"340\":105,\"341\":122,\"342\":78,\"343\":111,\"344\":99,\"345\":93,\"346\":87,\"347\":89,\"348\":82,\"349\":84,\"350\":75,\"351\":99,\"352\":89,\"353\":102,\"354\":70,\"355\":68,\"356\":91,\"357\":90,\"358\":69,\"359\":67,\"360\":83,\"361\":85,\"362\":72,\"363\":58,\"364\":77,\"365\":63,\"366\":59,\"367\":56,\"368\":67,\"369\":57,\"370\":64,\"371\":60,\"372\":67,\"373\":62,\"374\":70,\"375\":55,\"376\":48,\"377\":51,\"378\":53,\"379\":46,\"380\":40,\"381\":58,\"382\":41,\"383\":46,\"384\":49,\"385\":36,\"386\":40,\"387\":40,\"388\":41,\"389\":37,\"390\":47,\"391\":37,\"392\":34,\"393\":25,\"394\":37,\"395\":31,\"396\":37,\"397\":36,\"398\":16,\"399\":48,\"400\":38,\"401\":30,\"402\":28,\"403\":27,\"404\":28,\"405\":23,\"406\":27,\"407\":24,\"408\":30,\"409\":30,\"410\":34,\"411\":23,\"412\":26,\"413\":19,\"414\":30,\"415\":23,\"416\":19,\"417\":22,\"418\":20,\"419\":18,\"420\":11,\"421\":16,\"422\":25,\"423\":21,\"424\":20,\"425\":14,\"426\":20,\"427\":15,\"428\":13,\"429\":16,\"430\":14,\"431\":17,\"432\":17,\"433\":16,\"434\":10,\"435\":16,\"436\":12,\"437\":13,\"438\":12,\"439\":23,\"440\":9,\"441\":8,\"442\":8,\"443\":7,\"444\":8,\"445\":11,\"446\":9,\"447\":7,\"448\":12,\"449\":13,\"450\":10,\"451\":13,\"452\":10,\"453\":10,\"454\":2,\"455\":8,\"456\":9,\"457\":7,\"458\":10,\"459\":6,\"460\":10,\"461\":7,\"462\":12,\"463\":9,\"464\":12,\"465\":7,\"466\":7,\"467\":6,\"468\":6,\"469\":10,\"470\":3,\"472\":6,\"473\":6,\"474\":10,\"475\":4,\"476\":4,\"477\":5,\"478\":2,\"479\":3,\"480\":3,\"481\":2,\"482\":4,\"483\":2,\"484\":3,\"485\":5,\"486\":1,\"487\":4,\"488\":1,\"489\":2,\"490\":6,\"491\":4,\"492\":1,\"493\":5,\"494\":3,\"495\":4,\"496\":2,\"497\":3,\"498\":2,\"499\":2,\"500\":5,\"501\":2,\"502\":1,\"504\":4,\"505\":4,\"506\":3,\"507\":2,\"508\":1,\"509\":3,\"513\":3,\"514\":2,\"515\":2,\"516\":4,\"517\":1,\"519\":2,\"520\":3,\"521\":2,\"522\":1,\"523\":2,\"524\":1,\"525\":1,\"528\":1,\"529\":1,\"530\":2,\"531\":3,\"532\":1,\"533\":1,\"534\":1,\"535\":1,\"537\":2,\"538\":1,\"539\":2,\"541\":1,\"542\":1,\"543\":1,\"544\":2,\"545\":2,\"546\":1,\"548\":1,\"553\":1,\"556\":4,\"558\":2,\"561\":1,\"564\":1,\"579\":3,\"582\":1,\"584\":1,\"585\":1,\"594\":2,\"596\":1,\"597\":1,\"598\":1,\"618\":1,\"621\":1,\"635\":1,\"649\":1,\"655\":1,\"658\":1,\"666\":1,\"670\":1,\"671\":1,\"683\":1,\"691\":1,\"694\":1,\"700\":1,\"728\":1},\"seqStats\":{\"numSeqs\":50000,\"numA\":1377402,\"numT\":1394169,\"numG\":961175,\"numC\":1314108,\"numN\":3146,\"minSeqLength\":101,\"maxSeqLength\":101,\"accSeqQual\":175671680,\"lengthMap\":{\"101\":50000},\"infoMap\":{\"0\":{\"numA\":13129,\"numT\":13950,\"numG\":10266,\"numC\":12605,\"numN\":50,\"numQual\":50000,\"accQual\":1486540},\"1\":{\"numA\":13121,\"numT\":14287,\"numG\":10898,\"numC\":11644,\"numN\":50,\"numQual\":50000,\"accQual\":1566154},\"2\":{\"numA\":13388,\"numT\":14560,\"numG\":9846,\"numC\":12165,\"numN\":41,\"numQual\":50000,\"accQual\":1568269},\"3\":{\"numA\":13065,\"numT\":14461,\"numG\":10818,\"numC\":11620,\"numN\":36,\"numQual\":50000,\"accQual\":1638708},\"4\":{\"numA\":13045,\"numT\":14557,\"numG\":10721,\"numC\":11633,\"numN\":44,\"numQual\":50000,\"accQual\":1637961},\"5\":{\"numA\":13568,\"numT\":13879,\"numG\":10078,\"numC\":12425,\"numN\":50,\"numQual\":50000,\"accQual\":1637425},\"6\":{\"numA\":13037,\"numT\":14707,\"numG\":10392,\"numC\":11824,\"numN\":40,\"numQual\":50000,\"accQual\":1642538},\"7\":{\"numA\":13216,\"numT\":14023,\"numG\":10217,\"numC\":12505,\"numN\":39,\"numQual\":50000,\"accQual\":1643549},\"8\":{\"numA\":13077,\"numT\":14767,\"numG\":9939,\"numC\":12174,\"numN\":43,\"numQual\":50000,\"accQual\":1675488},\"9\":{\"numA\":13369,\"numT\":14917,\"numG\":9751,\"numC\":11922,\"numN\":41,\"numQual\":50000,\"accQual\":1676712},\"10\":{\"numA\":13503,\"numT\":13930,\"numG\":9913,\"numC\":12614,\"numN\":40,\"numQual\":50000,\"accQual\":1678887},\"11\":{\"numA\":13983,\"numT\":14262,\"numG\":10135,\"numC\":11584,\"numN\":36,\"numQual\":50000,\"accQual\":1680130},\"12\":{\"numA\":12806,\"numT\":14409,\"numG\":10200,\"numC\":12549,\"numN\":36,\"numQual\":50000,\"accQual\":1684264},\"13\":{\"numA\":13269,\"numT\":14097,\"numG\":9835,\"numC\":12764,\"numN\":35,\"numQual\":50000,\"accQual\":1712723},\"14\":{\"numA\":13338,\"numT\":14127,\"numG\":9789,\"numC\":12710,\"numN\":36,\"numQual\":50000,\"accQual\":1710785},\"15\":{\"numA\":13162,\"numT\":14100,\"numG\":10144,\"numC\":12556,\"numN\":38,\"numQual\":50000,\"accQual\":1713514},\"16\":{\"numA\":13504,\"numT\":13881,\"numG\":10593,\"numC\":11979,\"numN\":43,\"numQual\":50000,\"accQual\":1714438},\"17\":{\"numA\":13466,\"numT\":13906,\"numG\":10434,\"numC\":12156,\"numN\":38,\"numQual\":50000,\"accQual\":1716997},\"18\":{\"numA\":13165,\"numT\":14348,\"numG\":10033,\"numC\":12425,\"numN\":29,\"numQual\":50000,\"accQual\":1721020},\"19\":{\"numA\":13448,\"numT\":13837,\"numG\":10582,\"numC\":12100,\"numN\":33,\"numQual\":50000,\"accQual\":1718526},\"20\":{\"numA\":13532,\"numT\":13711,\"numG\":10453,\"numC\":12275,\"numN\":29,\"numQual\":50000,\"accQual\":1722784},\"21\":{\"numA\":13065,\"numT\":14043,\"numG\":10021,\"numC\":12836,\"numN\":35,\"numQual\":50000,\"accQual\":1727756},\"22\":{\"numA\":13258,\"numT\":13754,\"numG\":10110,\"numC\":12843,\"numN\":35,\"numQual\":50000,\"accQual\":1728934},\"23\":{\"numA\":13276,\"numT\":13697,\"numG\":10066,\"numC\":12930,\"numN\":31,\"numQual\":50000,\"accQual\":1725076},\"24\":{\"numA\":13218,\"numT\":13474,\"numG\":10524,\"numC\":12754,\"numN\":30,\"numQual\":50000,\"accQual\":1717793},\"25\":{\"numA\":13283,\"numT\":13638,\"numG\":10206,\"numC\":12841,\"numN\":32,\"numQual\":50000,\"accQual\":1702164},\"26\":{\"numA\":13644,\"numT\":13791,\"numG\":10139,\"numC\":12402,\"numN\":24,\"numQual\":50000,\"accQual\":1738241},\"27\":{\"numA\":13453,\"numT\":13134,\"numG\":10339,\"numC\":13044,\"numN\":30,\"numQual\":50000,\"accQual\":1743605},\"28\":{\"numA\":13040,\"numT\":14178,\"numG\":10211,\"numC\":12542,\"numN\":29,\"numQual\":50000,\"accQual\":1747787},\"29\":{\"numA\":13005,\"numT\":13675,\"numG\":10040,\"numC\":13255,\"numN\":25,\"numQual\":50000,\"accQual\":1753961},\"30\":{\"numA\":13020,\"numT\":13285,\"numG\":10703,\"numC\":12964,\"numN\":28,\"numQual\":50000,\"accQual\":1760745},\"31\":{\"numA\":13267,\"numT\":13689,\"numG\":10174,\"numC\":12842,\"numN\":28,\"numQual\":50000,\"accQual\":1764794},\"32\":{\"numA\":13214,\"numT\":13822,\"numG\":10310,\"numC\":12626,\"numN\":28,\"numQual\":50000,\"accQual\":1770901},\"33\":{\"numA\":12994,\"numT\":13504,\"numG\":10893,\"numC\":12580,\"numN\":29,\"numQual\":50000,\"accQual\":1776404},\"34\":{\"numA\":13157,\"numT\":13983,\"numG\":10221,\"numC\":12609,\"numN\":30,\"numQual\":50000,\"accQual\":1782745},\"35\":{\"numA\":13505,\"numT\":14061,\"numG\":9921,\"numC\":12485,\"numN\":28,\"numQual\":50000,\"accQual\":1787730},\"36\":{\"numA\":13342,\"numT\":13755,\"numG\":10088,\"numC\":12786,\"numN\":29,\"numQual\":50000,\"accQual\":1793768},\"37\":{\"numA\":13080,\"numT\":14308,\"numG\":9908,\"numC\":12672,\"numN\":32,\"numQual\":50000,\"accQual\":1794778},\"38\":{\"numA\":13417,\"numT\":14610,\"numG\":9898,\"numC\":12045,\"numN\":30,\"numQual\":50000,\"accQual\":1797767},\"39\":{\"numA\":13313,\"numT\":13942,\"numG\":10518,\"numC\":12199,\"numN\":28,\"numQual\":50000,\"accQual\":1798697},\"40\":{\"numA\":13631,\"numT\":14345,\"numG\":9432,\"numC\":12563,\"numN\":29,\"numQual\":50000,\"accQual\":1803282},\"41\":{\"numA\":13006,\"numT\":14111,\"numG\":10010,\"numC\":12843,\"numN\":30,\"numQual\":50000,\"accQual\":1806988},\"42\":{\"numA\":13058,\"numT\":14505,\"numG\":10089,\"numC\":12319,\"numN\":29,\"numQual\":50000,\"accQual\":1806358},\"43\":{\"numA\":12856,\"numT\":14498,\"numG\":10745,\"numC\":11874,\"numN\":27,\"numQual\":50000,\"accQual\":1807155},\"44\":{\"numA\":13014,\"numT\":14665,\"numG\":9984,\"numC\":12310,\"numN\":27,\"numQual\":50000,\"accQual\":1806662},\"45\":{\"numA\":12983,\"numT\":13941,\"numG\":10480,\"numC\":12571,\"numN\":25,\"numQual\":50000,\"accQual\":1807971},\"46\":{\"numA\":13262,\"numT\":14215,\"numG\":10298,\"numC\":12203,\"numN\":22,\"numQual\":50000,\"accQual\":1805752},\"47\":{\"numA\":12827,\"numT\":14586,\"numG\":9900,\"numC\":12666,\"numN\":21,\"numQual\":50000,\"accQual\":1803973},\"48\":{\"numA\":12885,\"numT\":14712,\"numG\":10158,\"numC\":12225,\"numN\":20,\"numQual\":50000,\"accQual\":1797343},\"49\":{\"numA\":12450,\"numT\":14697,\"numG\":10233,\"numC\":12600,\"numN\":20,\"numQual\":50000,\"accQual\":1788117},\"50\":{\"numA\":12189,\"numT\":14387,\"numG\":10502,\"numC\":12904,\"numN\":18,\"numQual\":50000,\"accQual\":1755096},\"51\":{\"numA\":13036,\"numT\":13926,\"numG\":10775,\"numC\":12240,\"numN\":23,\"numQual\":50000,\"accQual\":1798796},\"52\":{\"numA\":12900,\"numT\":13872,\"numG\":10756,\"numC\":12453,\"numN\":19,\"numQual\":50000,\"accQual\":1803729},\"53\":{\"numA\":13080,\"numT\":14232,\"numG\":9690,\"numC\":12977,\"numN\":21,\"numQual\":50000,\"accQual\":1808445},\"54\":{\"numA\":12987,\"numT\":14115,\"numG\":9991,\"numC\":12883,\"numN\":24,\"numQual\":50000,\"accQual\":1811022},\"55\":{\"numA\":12635,\"numT\":14542,\"numG\":10400,\"numC\":12395,\"numN\":28,\"numQual\":50000,\"accQual\":1812187},\"56\":{\"numA\":12916,\"numT\":13624,\"numG\":10277,\"numC\":13157,\"numN\":26,\"numQual\":50000,\"accQual\":1815139},\"57\":{\"numA\":13533,\"numT\":13760,\"numG\":10175,\"numC\":12507,\"numN\":25,\"numQual\":50000,\"accQual\":1815429},\"58\":{\"numA\":12701,\"numT\":14091,\"numG\":10341,\"numC\":12839,\"numN\":28,\"numQual\":50000,\"accQual\":1815316},\"59\":{\"numA\":12780,\"numT\":13380,\"numG\":10055,\"numC\":13758,\"numN\":27,\"numQual\":50000,\"accQual\":1811993},\"60\":{\"numA\":13237,\"numT\":13195,\"numG\":10077,\"numC\":13462,\"numN\":29,\"numQual\":50000,\"accQual\":1807755},\"61\":{\"numA\":13097,\"numT\":13392,\"numG\":9980,\"numC\":13503,\"numN\":28,\"numQual\":50000,\"accQual\":1805607},\"62\":{\"numA\":13363,\"numT\":13873,\"numG\":9461,\"numC\":13277,\"numN\":26,\"numQual\":50000,\"accQual\":1802531},\"63\":{\"numA\":12756,\"numT\":14147,\"numG\":9673,\"numC\":13398,\"numN\":26,\"numQual\":50000,\"accQual\":1801373},\"64\":{\"numA\":13080,\"numT\":13674,\"numG\":9826,\"numC\":13402,\"numN\":18,\"numQual\":50000,\"accQual\":1801162},\"65\":{\"numA\":13733,\"numT\":13405,\"numG\":9271,\"numC\":13569,\"numN\":22,\"numQual\":50000,\"accQual\":1796594},\"66\":{\"numA\":13325,\"numT\":13918,\"numG\":9493,\"numC\":13238,\"numN\":26,\"numQual\":50000,\"accQual\":1794020},\"67\":{\"numA\":14021,\"numT\":13309,\"numG\":9430,\"numC\":13213,\"numN\":27,\"numQual\":50000,\"accQual\":1790924},\"68\":{\"numA\":13400,\"numT\":14149,\"numG\":9197,\"numC\":13232,\"numN\":22,\"numQual\":50000,\"accQual\":1783835},\"69\":{\"numA\":13563,\"numT\":13900,\"numG\":8845,\"numC\":13665,\"numN\":27,\"numQual\":50000,\"accQual\":1780793},\"70\":{\"numA\":13735,\"numT\":13763,\"numG\":9152,\"numC\":13319,\"numN\":31,\"numQual\":50000,\"accQual\":1777483},\"71\":{\"numA\":13223,\"numT\":13894,\"numG\":8640,\"numC\":14219,\"numN\":24,\"numQual\":50000,\"accQual\":1774556},\"72\":{\"numA\":13800,\"numT\":13702,\"numG\":9079,\"numC\":13396,\"numN\":23,\"numQual\":50000,\"accQual\":1771964},\"73\":{\"numA\":14005,\"numT\":13419,\"numG\":8787,\"numC\":13757,\"numN\":32,\"numQual\":50000,\"accQual\":1773014},\"74\":{\"numA\":14453,\"numT\":13362,\"numG\":8799,\"numC\":13360,\"numN\":26,\"numQual\":50000,\"accQual\":1770697},\"75\":{\"numA\":14028,\"numT\":13381,\"numG\":8468,\"numC\":14096,\"numN\":27,\"numQual\":50000,\"accQual\":1750954},\"76\":{\"numA\":14660,\"numT\":13365,\"numG\":8149,\"numC\":13802,\"numN\":24,\"numQual\":50000,\"accQual\":1761352},\"77\":{\"numA\":14383,\"numT\":13605,\"numG\":8354,\"numC\":13632,\"numN\":26,\"numQual\":50000,\"accQual\":1765675},\"78\":{\"numA\":14826,\"numT\":13441,\"numG\":8086,\"numC\":13608,\"numN\":39,\"numQual\":50000,\"accQual\":1766909},\"79\":{\"numA\":14624,\"numT\":13024,\"numG\":8720,\"numC\":13601,\"numN\":31,\"numQual\":50000,\"accQual\":1765251},\"80\":{\"numA\":14319,\"numT\":13495,\"numG\":8647,\"numC\":13519,\"numN\":20,\"numQual\":50000,\"accQual\":1768478},\"81\":{\"numA\":14305,\"numT\":13523,\"numG\":8383,\"numC\":13760,\"numN\":29,\"numQual\":50000,\"accQual\":1769663},\"82\":{\"numA\":14596,\"numT\":13552,\"numG\":8313,\"numC\":13518,\"numN\":21,\"numQual\":50000,\"accQual\":1769456},\"83\":{\"numA\":14697,\"numT\":13239,\"numG\":8208,\"numC\":13824,\"numN\":32,\"numQual\":50000,\"accQual\":1764213},\"84\":{\"numA\":15376,\"numT\":13246,\"numG\":8191,\"numC\":13135,\"numN\":52,\"numQual\":50000,\"accQual\":1760280},\"85\":{\"numA\":15064,\"numT\":12985,\"numG\":8352,\"numC\":13564,\"numN\":35,\"numQual\":50000,\"accQual\":1759698},\"86\":{\"numA\":14953,\"numT\":13071,\"numG\":8023,\"numC\":13921,\"numN\":32,\"numQual\":50000,\"accQual\":1760657},\"87\":{\"numA\":15019,\"numT\":13628,\"numG\":7964,\"numC\":13358,\"numN\":31,\"numQual\":50000,\"accQual\":1758734},\"88\":{\"numA\":14468,\"numT\":13735,\"numG\":8117,\"numC\":13643,\"numN\":37,\"numQual\":50000,\"accQual\":1717494},\"89\":{\"numA\":14873,\"numT\":13191,\"numG\":8050,\"numC\":13849,\"numN\":37,\"numQual\":50000,\"accQual\":1715848},\"90\":{\"numA\":15209,\"numT\":12975,\"numG\":8026,\"numC\":13752,\"numN\":38,\"numQual\":50000,\"accQual\":1714857},\"91\":{\"numA\":15237,\"numT\":13151,\"numG\":7770,\"numC\":13807,\"numN\":35,\"numQual\":50000,\"accQual\":1712999},\"92\":{\"numA\":15447,\"numT\":12967,\"numG\":8360,\"numC\":13185,\"numN\":41,\"numQual\":50000,\"accQual\":1712313},\"93\":{\"numA\":14438,\"numT\":13357,\"numG\":7758,\"numC\":14411,\"numN\":36,\"numQual\":50000,\"accQual\":1660582},\"94\":{\"numA\":14929,\"numT\":12415,\"numG\":7779,\"numC\":14840,\"numN\":37,\"numQual\":50000,\"accQual\":1660442},\"95\":{\"numA\":14807,\"numT\":12672,\"numG\":7856,\"numC\":14615,\"numN\":50,\"numQual\":50000,\"accQual\":1659848},\"96\":{\"numA\":14660,\"numT\":13050,\"numG\":7379,\"numC\":14874,\"numN\":37,\"numQual\":50000,\"accQual\":1660484},\"97\":{\"numA\":15731,\"numT\":13027,\"numG\":7257,\"numC\":13958,\"numN\":27,\"numQual\":50000,\"accQual\":1660628},\"98\":{\"numA\":14741,\"numT\":13273,\"numG\":7728,\"numC\":14232,\"numN\":26,\"numQual\":50000,\"accQual\":1550553},\"99\":{\"numA\":14583,\"numT\":13224,\"numG\":6803,\"numC\":15349,\"numN\":41,\"numQual\":50000,\"accQual\":1546825},\"100\":{\"numA\":15167,\"numT\":13212,\"numG\":7106,\"numC\":14474,\"numN\":41,\"numQual\":50000,\"accQual\":1503368}},\"kmers\":{\"kvalue\":0,\"kmersMap\":{}}},\"nm\":78353}'},\n",
       "   'attributes': {'alignmentHeader': {'attributes': {},\n",
       "     'sequenceDiccionary': [{'sequenceName': 'chrM',\n",
       "       'sequenceLength': 16571,\n",
       "       'attributes': {}},\n",
       "      {'sequenceName': 'chr1', 'sequenceLength': 249250621, 'attributes': {}},\n",
       "      {'sequenceName': 'chr2', 'sequenceLength': 243199373, 'attributes': {}},\n",
       "      {'sequenceName': 'chr3', 'sequenceLength': 198022430, 'attributes': {}},\n",
       "      {'sequenceName': 'chr4', 'sequenceLength': 191154276, 'attributes': {}},\n",
       "      {'sequenceName': 'chr5', 'sequenceLength': 180915260, 'attributes': {}},\n",
       "      {'sequenceName': 'chr6', 'sequenceLength': 171115067, 'attributes': {}},\n",
       "      {'sequenceName': 'chr7', 'sequenceLength': 159138663, 'attributes': {}},\n",
       "      {'sequenceName': 'chr8', 'sequenceLength': 146364022, 'attributes': {}},\n",
       "      {'sequenceName': 'chr9', 'sequenceLength': 141213431, 'attributes': {}},\n",
       "      {'sequenceName': 'chr10', 'sequenceLength': 135534747, 'attributes': {}},\n",
       "      {'sequenceName': 'chr11', 'sequenceLength': 135006516, 'attributes': {}},\n",
       "      {'sequenceName': 'chr12', 'sequenceLength': 133851895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr13', 'sequenceLength': 115169878, 'attributes': {}},\n",
       "      {'sequenceName': 'chr14', 'sequenceLength': 107349540, 'attributes': {}},\n",
       "      {'sequenceName': 'chr15', 'sequenceLength': 102531392, 'attributes': {}},\n",
       "      {'sequenceName': 'chr16', 'sequenceLength': 90354753, 'attributes': {}},\n",
       "      {'sequenceName': 'chr17', 'sequenceLength': 81195210, 'attributes': {}},\n",
       "      {'sequenceName': 'chr18', 'sequenceLength': 78077248, 'attributes': {}},\n",
       "      {'sequenceName': 'chr19', 'sequenceLength': 59128983, 'attributes': {}},\n",
       "      {'sequenceName': 'chr20', 'sequenceLength': 63025520, 'attributes': {}},\n",
       "      {'sequenceName': 'chr21', 'sequenceLength': 48129895, 'attributes': {}},\n",
       "      {'sequenceName': 'chr22', 'sequenceLength': 51304566, 'attributes': {}},\n",
       "      {'sequenceName': 'chrX', 'sequenceLength': 155270560, 'attributes': {}},\n",
       "      {'sequenceName': 'chrY', 'sequenceLength': 59373566, 'attributes': {}}],\n",
       "     'readGroups': [{'id': 'NA12879', 'attributes': {'SM': 'NA12879'}}],\n",
       "     'programRecords': [{'id': 'bwa',\n",
       "       'attributes': {'VN': '0.6.1-r104-tpx', 'PN': 'bwa'}}],\n",
       "     'comments': [],\n",
       "     'studyName': 'Platinum',\n",
       "     'origin': 'SAM'}}},\n",
       "  {'id': 'data:NA12879_S1.bam.bai',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12879_S1.bam.bai',\n",
       "   'uuid': 'rOjqVQFnAAMAAZ7XsP30GA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BAI',\n",
       "   'bioformat': 'ALIGNMENT',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12879_S1.bam.bai',\n",
       "   'path': 'data/NA12879_S1.bam.bai',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20181214133004',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 8744304,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'data:NA12879_S1.bam.coverage.bw',\n",
       "   'annotationSets': [],\n",
       "   'name': 'NA12879_S1.bam.coverage.bw',\n",
       "   'uuid': 'rOjq5gFnAAMAAbwN0F3AnA',\n",
       "   'type': 'FILE',\n",
       "   'format': 'BIGWIG',\n",
       "   'bioformat': 'COVERAGE',\n",
       "   'checksum': '',\n",
       "   'uri': 'file:///mnt/data/hgva/datasets/bams/NA12879_S1.bam.coverage.bw',\n",
       "   'path': 'data/NA12879_S1.bam.coverage.bw',\n",
       "   'release': 1,\n",
       "   'creationDate': '20181214133004',\n",
       "   'modificationDate': '20181214133004',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20181214133004', 'message': ''},\n",
       "   'external': True,\n",
       "   'size': 308311099,\n",
       "   'experiment': {},\n",
       "   'samples': [],\n",
       "   'tags': [],\n",
       "   'job': {'startTime': 0,\n",
       "    'endTime': 0,\n",
       "    'visited': False,\n",
       "    'size': 0,\n",
       "    'release': 0},\n",
       "   'relatedFiles': [],\n",
       "   'index': {'jobId': 0, 'release': 0},\n",
       "   'stats': {},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## test to check if multiple extra parameters work\n",
    "fc = oc.files\n",
    "fc.search(study = \"cafetero@GRCH37:platinum\", type=\"FILE\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Working with Users"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_category', '_cfg', '_client_login_handler', '_delete', '_get', '_get_query_id_str', '_post', '_refresh_token_client', '_rest_retry', 'auto_refresh', 'configs', 'create', 'delete', 'filters', 'info', 'login', 'login_handler', 'logout', 'on_retry', 'projects', 'refresh_token', 'session_id', 'update', 'update_configs', 'update_filter', 'update_filters', 'update_password']\n"
     ]
    }
   ],
   "source": [
    "user_client = oc.users\n",
    "print(dir(user_client))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'cafetero',\n",
       "   'name': 'Juan Valds',\n",
       "   'email': 'cafetero@cafe.rico.es',\n",
       "   'organization': '',\n",
       "   'account': {'type': 'full',\n",
       "    'creationDate': '20180918141732',\n",
       "    'expirationDate': '20190918141732',\n",
       "    'authOrigin': 'internal'},\n",
       "   'status': {'name': 'READY', 'date': '20180918141732', 'message': ''},\n",
       "   'lastModified': '20190221155441344',\n",
       "   'size': -1,\n",
       "   'quota': 200000,\n",
       "   'projects': [{'id': 'GRCH37',\n",
       "     'name': 'GRCH37',\n",
       "     'uuid': '7Q5e4QFlAAEAAQ7Nii5FVA',\n",
       "     'fqn': 'cafetero@GRCH37',\n",
       "     'creationDate': '20180918142106',\n",
       "     'description': '',\n",
       "     'organization': '',\n",
       "     'organism': {'scientificName': 'Homo Sapiens',\n",
       "      'commonName': 'grch37',\n",
       "      'taxonomyCode': 0,\n",
       "      'assembly': 'grch37'},\n",
       "     'currentRelease': 1,\n",
       "     'status': {'name': 'READY', 'date': '20180918142106', 'message': ''},\n",
       "     'size': 0,\n",
       "     'studies': [{'id': 'platinum',\n",
       "       'name': 'Platinum',\n",
       "       'uuid': '7REdkAFlAAIAAVnBP_uPGg',\n",
       "       'type': 'CASE_CONTROL',\n",
       "       'creationDate': '20180918142406',\n",
       "       'modificationDate': '20180918142406',\n",
       "       'description': '',\n",
       "       'status': {'name': 'READY', 'date': '20180918142406', 'message': ''},\n",
       "       'lastModified': '20180918142406',\n",
       "       'size': 0,\n",
       "       'cipher': 'none',\n",
       "       'fqn': 'cafetero@GRCH37:platinum',\n",
       "       'groups': [{'id': '@members',\n",
       "         'name': 'members',\n",
       "         'userIds': ['test2', 'test3', 'test4', 'test1', 'cafetero']},\n",
       "        {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "       'experiments': [],\n",
       "       'files': [],\n",
       "       'jobs': [],\n",
       "       'individuals': [],\n",
       "       'samples': [],\n",
       "       'datasets': [],\n",
       "       'cohorts': [],\n",
       "       'panels': [],\n",
       "       'permissionRules': {},\n",
       "       'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/2/',\n",
       "       'release': 1,\n",
       "       'dataStores': {},\n",
       "       'stats': {},\n",
       "       'attributes': {}},\n",
       "      {'id': 'GNOMAD_EXOMES',\n",
       "       'name': 'GenomAD Exomes',\n",
       "       'uuid': '7RFu4AFlAAIAAZSS4XjzBA',\n",
       "       'type': 'CASE_CONTROL',\n",
       "       'creationDate': '20180918142427',\n",
       "       'modificationDate': '20180918142427',\n",
       "       'description': '',\n",
       "       'status': {'name': 'READY', 'date': '20180918142427', 'message': ''},\n",
       "       'lastModified': '20180918142427',\n",
       "       'size': 0,\n",
       "       'cipher': 'none',\n",
       "       'fqn': 'cafetero@GRCH37:GNOMAD_EXOMES',\n",
       "       'groups': [{'id': '@members',\n",
       "         'name': 'members',\n",
       "         'userIds': ['test2', 'test4']},\n",
       "        {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "       'experiments': [],\n",
       "       'files': [],\n",
       "       'jobs': [],\n",
       "       'individuals': [],\n",
       "       'samples': [],\n",
       "       'datasets': [],\n",
       "       'cohorts': [],\n",
       "       'panels': [],\n",
       "       'permissionRules': {},\n",
       "       'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/4/',\n",
       "       'release': 1,\n",
       "       'dataStores': {},\n",
       "       'stats': {},\n",
       "       'attributes': {}}],\n",
       "     'dataStores': {},\n",
       "     'attributes': {}}],\n",
       "   'tools': [],\n",
       "   'configs': {'filters': [{'name': 'Indels',\n",
       "      'description': 'Fetch all INDEL variants',\n",
       "      'bioformat': 'VARIANT',\n",
       "      'query': {'studies': 'cafetero@GRCH37:platinum', 'type': 'INDEL'},\n",
       "      'options': {}}]},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## getting user information\n",
    "## [NOTE] User needs the quey_id string directly --> (user)\n",
    "user_client.info(user)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'GRCH37',\n",
       "   'name': 'GRCH37',\n",
       "   'uuid': '7Q5e4QFlAAEAAQ7Nii5FVA',\n",
       "   'fqn': 'cafetero@GRCH37',\n",
       "   'creationDate': '20180918142106',\n",
       "   'description': '',\n",
       "   'organization': '',\n",
       "   'organism': {'scientificName': 'Homo Sapiens',\n",
       "    'commonName': 'grch37',\n",
       "    'taxonomyCode': 0,\n",
       "    'assembly': 'grch37'},\n",
       "   'currentRelease': 1,\n",
       "   'status': {'name': 'READY', 'date': '20180918142106', 'message': ''},\n",
       "   'size': 0,\n",
       "   'studies': [{'id': 'platinum',\n",
       "     'name': 'Platinum',\n",
       "     'uuid': '7REdkAFlAAIAAVnBP_uPGg',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142406',\n",
       "     'modificationDate': '20180918142406',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142406', 'message': ''},\n",
       "     'lastModified': '20180918142406',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:platinum',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test3', 'test4', 'test1', 'cafetero']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/2/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}},\n",
       "    {'id': 'GNOMAD_EXOMES',\n",
       "     'name': 'GenomAD Exomes',\n",
       "     'uuid': '7RFu4AFlAAIAAZSS4XjzBA',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142427',\n",
       "     'modificationDate': '20180918142427',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142427', 'message': ''},\n",
       "     'lastModified': '20180918142427',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:GNOMAD_EXOMES',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test4']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/4/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'dataStores': {},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Getting user projects\n",
    "## [NOTE] Client specific methods have the query_id as a key:value (i.e (user=user_id)) \n",
    "user_client.projects(user=user)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{}]]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Getting user configurations\n",
    "## [NOTE] Client specific methods have the query_id as a key:value (i.e (user=user_id)) \n",
    "user_client.configs(user=user)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Indels',\n",
       "   'description': 'Fetch all INDEL variants',\n",
       "   'bioformat': 'VARIANT',\n",
       "   'query': {'studies': 'cafetero@GRCH37:platinum', 'type': 'INDEL'},\n",
       "   'options': {}}]]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Getting user filter configurations\n",
    "## [NOTE] Client specific methods have the query_id as a key:value (i.e (user=user_id)) \n",
    "user_client.filters(user=user)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Working with Projects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_category', '_cfg', '_client_login_handler', '_delete', '_get', '_get_query_id_str', '_post', '_refresh_token_client', '_rest_retry', 'auto_refresh', 'create', 'delete', 'increment_release', 'info', 'login_handler', 'on_retry', 'search', 'session_id', 'stats', 'studies', 'update']\n"
     ]
    }
   ],
   "source": [
    "project_client = oc.projects\n",
    "print(dir(oc.projects))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'GRCH37',\n",
       "   'name': 'GRCH37',\n",
       "   'uuid': '7Q5e4QFlAAEAAQ7Nii5FVA',\n",
       "   'fqn': 'cafetero@GRCH37',\n",
       "   'creationDate': '20180918142106',\n",
       "   'description': '',\n",
       "   'organization': '',\n",
       "   'organism': {'scientificName': 'Homo Sapiens',\n",
       "    'commonName': 'grch37',\n",
       "    'taxonomyCode': 0,\n",
       "    'assembly': 'grch37'},\n",
       "   'currentRelease': 1,\n",
       "   'status': {'name': 'READY', 'date': '20180918142106', 'message': ''},\n",
       "   'size': 0,\n",
       "   'studies': [{'id': 'platinum',\n",
       "     'name': 'Platinum',\n",
       "     'uuid': '7REdkAFlAAIAAVnBP_uPGg',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142406',\n",
       "     'modificationDate': '20180918142406',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142406', 'message': ''},\n",
       "     'lastModified': '20180918142406',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:platinum',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test3', 'test4', 'test1', 'cafetero']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/2/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}},\n",
       "    {'id': 'GNOMAD_EXOMES',\n",
       "     'name': 'GenomAD Exomes',\n",
       "     'uuid': '7RFu4AFlAAIAAZSS4XjzBA',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142427',\n",
       "     'modificationDate': '20180918142427',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142427', 'message': ''},\n",
       "     'lastModified': '20180918142427',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:GNOMAD_EXOMES',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test4']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/4/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'dataStores': {},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Getting all projects from logged in user\n",
    "project_client.search()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'GRCH37',\n",
       "   'name': 'GRCH37',\n",
       "   'uuid': '7Q5e4QFlAAEAAQ7Nii5FVA',\n",
       "   'fqn': 'cafetero@GRCH37',\n",
       "   'creationDate': '20180918142106',\n",
       "   'description': '',\n",
       "   'organization': '',\n",
       "   'organism': {'scientificName': 'Homo Sapiens',\n",
       "    'commonName': 'grch37',\n",
       "    'taxonomyCode': 0,\n",
       "    'assembly': 'grch37'},\n",
       "   'currentRelease': 1,\n",
       "   'status': {'name': 'READY', 'date': '20180918142106', 'message': ''},\n",
       "   'size': 0,\n",
       "   'studies': [{'id': 'platinum',\n",
       "     'name': 'Platinum',\n",
       "     'uuid': '7REdkAFlAAIAAVnBP_uPGg',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142406',\n",
       "     'modificationDate': '20180918142406',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142406', 'message': ''},\n",
       "     'lastModified': '20180918142406',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:platinum',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test3', 'test4', 'test1', 'cafetero']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/2/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}},\n",
       "    {'id': 'GNOMAD_EXOMES',\n",
       "     'name': 'GenomAD Exomes',\n",
       "     'uuid': '7RFu4AFlAAIAAZSS4XjzBA',\n",
       "     'type': 'CASE_CONTROL',\n",
       "     'creationDate': '20180918142427',\n",
       "     'modificationDate': '20180918142427',\n",
       "     'description': '',\n",
       "     'status': {'name': 'READY', 'date': '20180918142427', 'message': ''},\n",
       "     'lastModified': '20180918142427',\n",
       "     'size': 0,\n",
       "     'cipher': 'none',\n",
       "     'fqn': 'cafetero@GRCH37:GNOMAD_EXOMES',\n",
       "     'groups': [{'id': '@members',\n",
       "       'name': 'members',\n",
       "       'userIds': ['test2', 'test4']},\n",
       "      {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "     'experiments': [],\n",
       "     'files': [],\n",
       "     'jobs': [],\n",
       "     'individuals': [],\n",
       "     'samples': [],\n",
       "     'datasets': [],\n",
       "     'cohorts': [],\n",
       "     'panels': [],\n",
       "     'variableSets': [],\n",
       "     'permissionRules': {},\n",
       "     'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/4/',\n",
       "     'release': 1,\n",
       "     'dataStores': {},\n",
       "     'stats': {},\n",
       "     'attributes': {}}],\n",
       "   'dataStores': {},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Getting information from a specific project\n",
    "project_client.info('GRCH37')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'id': 'platinum',\n",
       "   'name': 'Platinum',\n",
       "   'uuid': '7REdkAFlAAIAAVnBP_uPGg',\n",
       "   'type': 'CASE_CONTROL',\n",
       "   'creationDate': '20180918142406',\n",
       "   'modificationDate': '20180918142406',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918142406', 'message': ''},\n",
       "   'lastModified': '20180918142406',\n",
       "   'size': 0,\n",
       "   'cipher': 'none',\n",
       "   'fqn': 'cafetero@GRCH37:platinum',\n",
       "   'groups': [{'id': '@members',\n",
       "     'name': 'members',\n",
       "     'userIds': ['test2', 'test3', 'test4', 'test1', 'cafetero']},\n",
       "    {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "   'experiments': [],\n",
       "   'files': [],\n",
       "   'jobs': [],\n",
       "   'individuals': [],\n",
       "   'samples': [],\n",
       "   'datasets': [],\n",
       "   'cohorts': [],\n",
       "   'panels': [],\n",
       "   'variableSets': [],\n",
       "   'permissionRules': {},\n",
       "   'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/2/',\n",
       "   'release': 1,\n",
       "   'dataStores': {},\n",
       "   'stats': {},\n",
       "   'attributes': {}},\n",
       "  {'id': 'GNOMAD_EXOMES',\n",
       "   'name': 'GenomAD Exomes',\n",
       "   'uuid': '7RFu4AFlAAIAAZSS4XjzBA',\n",
       "   'type': 'CASE_CONTROL',\n",
       "   'creationDate': '20180918142427',\n",
       "   'modificationDate': '20180918142427',\n",
       "   'description': '',\n",
       "   'status': {'name': 'READY', 'date': '20180918142427', 'message': ''},\n",
       "   'lastModified': '20180918142427',\n",
       "   'size': 0,\n",
       "   'cipher': 'none',\n",
       "   'fqn': 'cafetero@GRCH37:GNOMAD_EXOMES',\n",
       "   'groups': [{'id': '@members',\n",
       "     'name': 'members',\n",
       "     'userIds': ['test2', 'test4']},\n",
       "    {'id': '@admins', 'name': 'admins', 'userIds': []}],\n",
       "   'experiments': [],\n",
       "   'files': [],\n",
       "   'jobs': [],\n",
       "   'individuals': [],\n",
       "   'samples': [],\n",
       "   'datasets': [],\n",
       "   'cohorts': [],\n",
       "   'panels': [],\n",
       "   'variableSets': [],\n",
       "   'permissionRules': {},\n",
       "   'uri': 'file:///mnt/data/opencga-test/sessions/users/cafetero/projects/1/4/',\n",
       "   'release': 1,\n",
       "   'dataStores': {},\n",
       "   'stats': {},\n",
       "   'attributes': {}}]]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Fetching the studies from a project\n",
    "project_client.studies('GRCH37')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "b'{\"apiVersion\":\"v1\",\"time\":136,\"warning\":\"\",\"error\":\"Solr server is not alive\",\"queryOptions\":{\"metadata\":true,\"skipCount\":true,\"skip\":0,\"limit\":1000},\"response\":[{\"id\":\"\",\"dbTime\":-1,\"numResults\":-1,\"numTotalResults\":-1,\"warningMsg\":\"Future errors will ONLY be shown in the QueryResponse body\",\"errorMsg\":\"DEPRECATED: org.apache.solr.common.SolrException: Solr server is not alive\",\"resultType\":\"\",\"result\":[]}]}'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-4c0a654dfb74>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m## Getting stats from a project --> It seems that the Solr service is not working!!!\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mproject_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'GRCH37'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/project_client.py\u001b[0m in \u001b[0;36mstats\u001b[0;34m(self, project, **options)\u001b[0m\n\u001b[1;32m     54\u001b[0m             \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstudies\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m.\u001b[0m \u001b[0mFor\u001b[0m \u001b[0mnested\u001b[0m \u001b[0mfields\u001b[0m \u001b[0muse\u001b[0m \u001b[0;34m>>\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstudies\u001b[0m\u001b[0;34m>>\u001b[0m\u001b[0mbiotype\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m         \"\"\"\n\u001b[0;32m---> 56\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'stats'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquery_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mproject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mstudies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/_parent_rest_clients.py\u001b[0m in \u001b[0;36m_get\u001b[0;34m(self, resource, query_id, subcategory, second_query_id, **options)\u001b[0m\n\u001b[1;32m     74\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquery_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubcategory\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msecond_query_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m         \u001b[0;34m\"\"\"Queries the REST service and returns the result\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 76\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_rest_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'get'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquery_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubcategory\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msecond_query_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     77\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_post\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquery_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubcategory\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msecond_query_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/_parent_rest_clients.py\u001b[0m in \u001b[0;36m_rest_retry\u001b[0;34m(self, method, resource, query_id, subcategory, second_query_id, data, dont_retry, **options)\u001b[0m\n\u001b[1;32m     66\u001b[0m             \u001b[0mexec_retry\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cfg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_attempts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cfg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmin_retry_secs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cfg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_retry_secs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m             \u001b[0mlogin_handler\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_client_login_handler\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlogin_handler\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 68\u001b[0;31m             on_retry=notify_retry, dont_retry=dont_retry)\n\u001b[0m\u001b[1;32m     69\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mauto_refresh\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/retry.py\u001b[0m in \u001b[0;36mretry\u001b[0;34m(func, max_attempts, initial_retry_seconds, max_retry_seconds, login_handler, on_retry, dont_retry)\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOpenCgaInvalidToken\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/_parent_rest_clients.py\u001b[0m in \u001b[0;36mexec_retry\u001b[0;34m()\u001b[0m\n\u001b[1;32m     52\u001b[0m                            \u001b[0mresource\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mresource\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m                            \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 54\u001b[0;31m                            options=options)\n\u001b[0m\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mnotify_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_val\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_tb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/commons.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(host, version, sid, category, resource, method, subcategory, query_id, second_query_id, data, **options)\u001b[0m\n\u001b[1;32m    289\u001b[0m         response = _fetch(host=host, version=version, sid=sid, category=category, subcategory=subcategory,\n\u001b[1;32m    290\u001b[0m                           \u001b[0mresource\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mresource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 291\u001b[0;31m                           query_id=query_id, second_query_id=second_query_id, **options)\n\u001b[0m\u001b[1;32m    292\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    293\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/DATA/DATA_docker/projects_software/opencb/opencga/opencga-fork_pyclient-mgviz_branch/opencga-client/src/main/python/pyOpenCGA/pyopencga/commons.py\u001b[0m in \u001b[0;36m_fetch\u001b[0;34m(host, version, sid, category, resource, method, subcategory, query_id, second_query_id, data, **options)\u001b[0m\n\u001b[1;32m    209\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m200\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 211\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontent\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    212\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    213\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mException\u001b[0m: b'{\"apiVersion\":\"v1\",\"time\":136,\"warning\":\"\",\"error\":\"Solr server is not alive\",\"queryOptions\":{\"metadata\":true,\"skipCount\":true,\"skip\":0,\"limit\":1000},\"response\":[{\"id\":\"\",\"dbTime\":-1,\"numResults\":-1,\"numTotalResults\":-1,\"warningMsg\":\"Future errors will ONLY be shown in the QueryResponse body\",\"errorMsg\":\"DEPRECATED: org.apache.solr.common.SolrException: Solr server is not alive\",\"resultType\":\"\",\"result\":[]}]}'"
     ]
    }
   ],
   "source": [
    "## Getting stats from a project --> It seems that the Solr service is not working!!!\n",
    "project_client.stats('GRCH37')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
