% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/File-methods.R
\name{fileClient,OpencgaR-method}
\alias{fileClient,OpencgaR-method}
\title{FileClient methods}
\usage{
\S4method{fileClient}{OpencgaR}(
  OpencgaR,
  annotationSet,
  file,
  files,
  folder,
  members,
  endpointName,
  params = NULL,
  ...
)
}
\arguments{
\item{annotationSet}{AnnotationSet ID to be updated.}

\item{file}{File uuid, id, or name.}

\item{files}{Comma separated list of file ids, names or paths. Paths must be separated by : instead of /.}

\item{folder}{Folder id or name. Paths must be separated by : instead of /.}

\item{members}{Comma separated list of user or group ids.}

\item{variableSetId}{Variable set ID or name.}

\item{annotationSetId}{Annotation set id. If not provided, variableSetId will be used.}

\item{field}{Comma separated list of fields for which to obtain the distinct values.}

\item{jobDependsOn}{Comma separated list of existing job IDs the job will depend on.}

\item{jobDescription}{Job description.}

\item{jobTags}{Job tags.}

\item{jobScheduledStartTime}{Time when the job is scheduled to start.}

\item{jobPriority}{Priority of the job.}

\item{jobDryRun}{Flag indicating that the job will be executed in dry-run mode. In this mode, OpenCGA will validate that all parameters and prerequisites are correctly set for successful execution, but the job will not actually run.}

\item{id}{Comma separated list of file IDs up to a maximum of 100. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{uuid}{Comma separated list file UUIDs up to a maximum of 100.}

\item{path}{Comma separated list of paths. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{uri}{Comma separated list of uris. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{type}{File type, either FILE or DIRECTORY.}

\item{external}{Boolean field indicating whether to filter by external or non external files.}

\item{status}{Filter by status.}

\item{internalStatus}{Filter by internal status.}

\item{internalVariantIndexStatus}{Filter by internal variant index status.}

\item{softwareName}{Software name.}

\item{directory}{Directory under which we want to look for files or folders.}

\item{creationDate}{Creation date. Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805.}

\item{modificationDate}{Modification date. Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805.}

\item{tags}{Tags. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{size}{File size.}

\item{sampleIds}{Comma separated list sample IDs or UUIDs up to a maximum of 100.}

\item{jobId}{Job ID that created the file(s) or folder(s).}

\item{annotation}{Annotation filters. Example: age>30;gender=FEMALE. For more information, please visit http://docs.opencb.org/display/opencga/AnnotationSets+1.4.0.}

\item{acl}{Filter entries for which a user has the provided permissions. Format: acl={user}:{permissions}. Example: acl=john:WRITE,WRITE_ANNOTATIONS will return all entries for which user john has both WRITE and WRITE_ANNOTATIONS permissions. Only study owners or administrators can query by this field. .}

\item{release}{Release when it was created.}

\item{name}{File name to overwrite the input fileName.}

\item{fileName}{\link{DEPRECATED} File name to overwrite the input fileName.}

\item{format}{File format. Allowed values: \link{'VCF BCF GVCF TBI BIGWIG SAM BAM BAI CRAM CRAI FASTQ FASTA PED TAB_SEPARATED_VALUES COMMA_SEPARATED_VALUES XML PROTOCOL_BUFFER JSON AVRO PARQUET PDF IMAGE PLAIN BINARY NONE UNKNOWN'}}

\item{fileFormat}{\link{DEPRECATED} File format. Allowed values: \link{'VCF BCF GVCF TBI BIGWIG SAM BAM BAI CRAM CRAI FASTQ FASTA PED TAB_SEPARATED_VALUES COMMA_SEPARATED_VALUES XML PROTOCOL_BUFFER JSON AVRO PARQUET PDF IMAGE PLAIN BINARY NONE UNKNOWN'}}

\item{bioformat}{File bioformat. Allowed values: \link{'MICROARRAY_EXPRESSION_ONECHANNEL_AGILENT MICROARRAY_EXPRESSION_ONECHANNEL_AFFYMETRIX MICROARRAY_EXPRESSION_ONECHANNEL_GENEPIX MICROARRAY_EXPRESSION_TWOCHANNELS_AGILENT MICROARRAY_EXPRESSION_TWOCHANNELS_GENEPIX DATAMATRIX_EXPRESSION IDLIST IDLIST_RANKED ANNOTATION_GENEVSANNOTATION OTHER_NEWICK OTHER_BLAST OTHER_INTERACTION OTHER_GENOTYPE OTHER_PLINK OTHER_VCF OTHER_PED VCF4 VARIANT ALIGNMENT COVERAGE SEQUENCE PEDIGREE REFERENCE_GENOME NONE UNKNOWN'}}

\item{checksum}{Expected MD5 file checksum.}

\item{resource}{Boolean field indicating whether the file is a resource or not.}

\item{relativeFilePath}{Path within catalog (directory) where the file will be located (default: root folder).}

\item{description}{description.}

\item{parents}{Create the parent directories if they do not exist.}

\item{member}{User or group id.}

\item{silent}{Boolean to retrieve all possible entries that are queried for, false to raise an exception whenever one of the entries looked for cannot be shown for whichever reason.}

\item{skipTrash}{Skip trash and delete the files/folders from disk directly (CANNOT BE RECOVERED).}

\item{flattenAnnotations}{Flatten the annotations?.}

\item{deleted}{Boolean to retrieve deleted files.}

\item{sampleIdsAction}{Action to be performed if the array of samples is being updated. Allowed values: \link{'ADD SET REMOVE'}}

\item{annotationSetsAction}{Action to be performed if the array of annotationSets is being updated. Allowed values: \link{'ADD SET REMOVE'}}

\item{relatedFilesAction}{Action to be performed if the array of relatedFiles is being updated. Allowed values: \link{'ADD SET REMOVE'}}

\item{tagsAction}{Action to be performed if the array of tags is being updated. Allowed values: \link{'ADD SET REMOVE'}}

\item{action}{Action to be performed: ADD to add new annotations; REPLACE to replace the value of an already existing annotation; SET to set the new list of annotations removing any possible old annotations; REMOVE to remove some annotations; RESET to set some annotations to the default value configured in the corresponding variables of the VariableSet if any. Allowed values: \link{'ADD SET REMOVE RESET REPLACE'}}

\item{pattern}{String pattern.}

\item{ignoreCase}{Flag to perform a case insensitive search.}

\item{maxCount}{Stop reading a file after 'n' matching lines. 0 means no limit.}

\item{offset}{Starting byte from which the file will be read.}

\item{data}{Parameters to modify.}

\item{lines}{Maximum number of lines to be returned up to a maximum of 1000.}

\item{limit}{Number of results to be returned.}

\item{skip}{Number of results to skip.}

\item{count}{Get the total number of results matching the query. Deactivated by default.}

\item{include}{Fields included in the response, whole JSON path must be provided.}

\item{exclude}{Fields excluded in the response, whole JSON path must be provided.}

\item{study}{Study [\link{organization@}project:]study where study and project can be either the ID or UUID.}

\item{maxDepth}{Maximum depth to get files from.}
}
\description{
This function implements the OpenCGA calls for managing Files.
The following table summarises the available \emph{actions} for this client:\tabular{llr}{
   endpointName \tab Endpoint WS \tab parameters accepted \cr
   updateAcl \tab /{apiVersion}/files/acl/{members}/update \tab study, members\link{*}, action\link{*}, body\link{*} \cr
   aggregationStats \tab /{apiVersion}/files/aggregationStats \tab study, id, uuid, name, path, uri, type, bioformat, format, external, status, internalStatus, internalVariantIndexStatus, softwareName, directory, creationDate, modificationDate, description, tags, size, sampleIds, jobId, annotation, acl, deleted, release, field \cr
   loadAnnotationSets \tab /{apiVersion}/files/annotationSets/load \tab study, variableSetId\link{*}, path\link{*}, parents, annotationSetId, body \cr
   bioformats \tab /{apiVersion}/files/bioformats \tab  \cr
   create \tab /{apiVersion}/files/create \tab study, parents, body\link{*} \cr
   distinct \tab /{apiVersion}/files/distinct \tab study, id, uuid, name, path, uri, type, bioformat, format, external, resource, status, internalStatus, internalVariantIndexStatus, softwareName, directory, creationDate, modificationDate, description, tags, size, sampleIds, jobId, annotation, acl, deleted, release, field\link{*} \cr
   fetch \tab /{apiVersion}/files/fetch \tab jobId, jobDescription, jobDependsOn, jobTags, jobScheduledStartTime, jobPriority, jobDryRun, study, body\link{*} \cr
   formats \tab /{apiVersion}/files/formats \tab  \cr
   link \tab /{apiVersion}/files/link \tab study, parents, body\link{*} \cr
   runLink \tab /{apiVersion}/files/link/run \tab study, jobId, jobDependsOn, jobDescription, jobTags, jobScheduledStartTime, jobPriority, jobDryRun, body\link{*} \cr
   runPostlink \tab /{apiVersion}/files/postlink/run \tab study, jobId, jobDependsOn, jobDescription, jobTags, jobScheduledStartTime, jobPriority, jobDryRun, body\link{*} \cr
   search \tab /{apiVersion}/files/search \tab include, exclude, limit, skip, count, flattenAnnotations, study, id, uuid, name, path, uri, type, bioformat, format, external, resource, status, internalStatus, internalVariantIndexStatus, softwareName, directory, creationDate, modificationDate, description, tags, size, sampleIds, jobId, annotation, acl, deleted, release \cr
   upload \tab /{apiVersion}/files/upload \tab file, name, fileName, format, fileFormat, bioformat, checksum, resource, study, relativeFilePath, description, parents \cr
   acl \tab /{apiVersion}/files/{files}/acl \tab files\link{*}, study, member, silent \cr
   delete \tab /{apiVersion}/files/{files}/delete \tab study, files\link{*}, skipTrash \cr
   info \tab /{apiVersion}/files/{files}/info \tab include, exclude, flattenAnnotations, files\link{*}, study, deleted \cr
   unlink \tab /{apiVersion}/files/{files}/unlink \tab study, files\link{*} \cr
   update \tab /{apiVersion}/files/{files}/update \tab include, exclude, files\link{*}, study, sampleIdsAction, annotationSetsAction, relatedFilesAction, tagsAction, body\link{*} \cr
   updateAnnotationSetsAnnotations \tab /{apiVersion}/files/{file}/annotationSets/{annotationSet}/annotations/update \tab file\link{*}, study, annotationSet\link{*}, action, body \cr
   download \tab /{apiVersion}/files/{file}/download \tab file\link{*}, study \cr
   grep \tab /{apiVersion}/files/{file}/grep \tab file\link{*}, study, pattern, ignoreCase, maxCount \cr
   head \tab /{apiVersion}/files/{file}/head \tab file\link{*}, study, offset, lines \cr
   image \tab /{apiVersion}/files/{file}/image \tab file\link{*}, study \cr
   move \tab /{apiVersion}/files/{file}/move \tab include, exclude, file\link{*}, study, body\link{*} \cr
   refresh \tab /{apiVersion}/files/{file}/refresh \tab file\link{*}, study \cr
   tail \tab /{apiVersion}/files/{file}/tail \tab file\link{*}, study, lines \cr
   list \tab /{apiVersion}/files/{folder}/list \tab include, exclude, limit, skip, count, folder\link{*}, study \cr
   tree \tab /{apiVersion}/files/{folder}/tree \tab include, exclude, folder\link{*}, study, maxDepth \cr
}
}
\section{Endpoint /{apiVersion}/files/acl/{members}/update}{

Update the set of permissions granted for the member.
}

\section{Endpoint /{apiVersion}/files/aggregationStats}{

Fetch catalog file stats.
}

\section{Endpoint /{apiVersion}/files/annotationSets/load}{

Load annotation sets from a TSV file.
}

\section{Endpoint /{apiVersion}/files/bioformats}{

List of accepted file bioformats.
}

\section{Endpoint /{apiVersion}/files/create}{

Create file or folder.
}

\section{Endpoint /{apiVersion}/files/distinct}{

File distinct method.
}

\section{Endpoint /{apiVersion}/files/fetch}{

Download an external file to catalog and register it.
}

\section{Endpoint /{apiVersion}/files/formats}{

List of accepted file formats.
}

\section{Endpoint /{apiVersion}/files/link}{

Link an external file into catalog.
}

\section{Endpoint /{apiVersion}/files/link/run}{

Link an external file into catalog asynchronously.
}

\section{Endpoint /{apiVersion}/files/postlink/run}{

Associate non-registered samples for files with high volumes of samples.
}

\section{Endpoint /{apiVersion}/files/search}{

File search method.
}

\section{Endpoint /{apiVersion}/files/upload}{

Resource to upload a file by chunks.
}

\section{Endpoint /{apiVersion}/files/{files}/acl}{

Return the acl defined for the file or folder. If member is provided, it will only return the acl for the member.
}

\section{Endpoint /{apiVersion}/files/{files}/delete}{

Delete existing files and folders.
}

\section{Endpoint /{apiVersion}/files/{files}/info}{

File info.
}

\section{Endpoint /{apiVersion}/files/{files}/unlink}{

Unlink linked files and folders.
}

\section{Endpoint /{apiVersion}/files/{files}/update}{

Update some file attributes.
}

\section{Endpoint /{apiVersion}/files/{file}/annotationSets/{annotationSet}/annotations/update}{

Update annotations from an annotationSet.
}

\section{Endpoint /{apiVersion}/files/{file}/download}{

Download file.
}

\section{Endpoint /{apiVersion}/files/{file}/grep}{

Filter lines of the file containing the pattern.
}

\section{Endpoint /{apiVersion}/files/{file}/head}{

Show the first lines of a file (up to a limit).
}

\section{Endpoint /{apiVersion}/files/{file}/image}{

Obtain the base64 content of an image.
}

\section{Endpoint /{apiVersion}/files/{file}/move}{

Move file to a different path.
}

\section{Endpoint /{apiVersion}/files/{file}/refresh}{

Refresh metadata from the selected file or folder. Return updated files.
}

\section{Endpoint /{apiVersion}/files/{file}/tail}{

Show the last lines of a file (up to a limit).
}

\section{Endpoint /{apiVersion}/files/{folder}/list}{

List all the files inside the folder.
}

\section{Endpoint /{apiVersion}/files/{folder}/tree}{

Obtain a tree view of the files and folders within a folder.
}

\seealso{
\url{http://docs.opencb.org/display/opencga/Using+OpenCGA} and the RESTful API documentation
\url{http://bioinfo.hpc.cam.ac.uk/opencga-prod/webservices/}
\link{*}: Required parameter
}
