% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Sample-methods.R
\name{sampleClient,OpencgaR-method}
\alias{sampleClient,OpencgaR-method}
\title{SampleClient methods}
\usage{
\S4method{sampleClient}{OpencgaR}(
  OpencgaR,
  annotationSet,
  members,
  sample,
  samples,
  endpointName,
  params = NULL,
  ...
)
}
\arguments{
\item{annotationSet}{AnnotationSet ID to be updated.}

\item{members}{Comma separated list of user or group ids.}

\item{sample}{Sample ID.}

\item{samples}{Comma separated list sample IDs or UUIDs up to a maximum of 100.}

\item{variableSetId}{Variable set ID or name.}

\item{path}{Path where the TSV file is located in OpenCGA or where it should be located.}

\item{parents}{Flag indicating whether to create parent directories if they don't exist (only when TSV file was not previously associated).}

\item{annotationSetId}{Annotation set id. If not provided, variableSetId will be used.}

\item{field}{Comma separated list of fields for which to obtain the distinct values.}

\item{file}{file.}

\item{variableSet}{variableSet.}

\item{limit}{Number of results to be returned.}

\item{skip}{Number of results to skip.}

\item{count}{Get the total number of results matching the query. Deactivated by default.}

\item{id}{Comma separated list sample IDs up to a maximum of 100. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{uuid}{Comma separated list sample UUIDs up to a maximum of 100.}

\item{somatic}{Somatic sample.}

\item{individualId}{Individual ID or UUID.}

\item{fileIds}{Comma separated list of file IDs, paths or UUIDs.}

\item{cohortIds}{Comma separated list of cohort IDs.}

\item{creationDate}{Creation date. Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805.}

\item{modificationDate}{Modification date. Format: yyyyMMddHHmmss. Examples: >2018, 2017-2018, <201805.}

\item{internalStatus}{Filter by internal status.}

\item{status}{Filter by status.}

\item{processingProduct}{Processing product.}

\item{processingPreparationMethod}{Processing preparation method.}

\item{processingExtractionMethod}{Processing extraction method.}

\item{processingLabSampleId}{Processing lab sample id.}

\item{collectionFrom}{Collection from.}

\item{collectionType}{Collection type.}

\item{collectionMethod}{Collection method.}

\item{phenotypes}{Comma separated list of phenotype ids or names. Also admits basic regular expressions using the operator '~', i.e. '~{perl-regex}' e.g. '~value' for case sensitive, '~/value/i' for case insensitive search.}

\item{annotation}{Annotation filters. Example: age>30;gender=FEMALE. For more information, please visit http://docs.opencb.org/display/opencga/AnnotationSets+1.4.0.}

\item{acl}{Filter entries for which a user has the provided permissions. Format: acl={user}:{permissions}. Example: acl=john:WRITE,WRITE_ANNOTATIONS will return all entries for which user john has both WRITE and WRITE_ANNOTATIONS permissions. Only study owners or administrators can query by this field. .}

\item{internalRgaStatus}{Index status of the sample for the Recessive Gene Analysis. Allowed values: \link{'NOT_INDEXED INDEXED INVALID_PERMISSIONS INVALID_METADATA INVALID'}}

\item{release}{Release when it was created.}

\item{snapshot}{Snapshot value (Latest version of the entry in the specified release).}

\item{statsId}{Sample variant stats Id. If this field is not provided and the user filters by other stats fields, it will automatically be set to ALL.}

\item{statsVariantCount}{Sample variant stats VariantCount.}

\item{statsChromosomeCount}{Sample variant stats ChromosomeCount.}

\item{statsTypeCount}{Sample variant stats TypeCount.}

\item{statsGenotypeCount}{Sample variant stats GenotypeCount.}

\item{statsTiTvRatio}{Sample variant stats TiTvRatio.}

\item{statsQualityAvg}{Sample variant stats QualityAvg.}

\item{statsQualityStdDev}{Sample variant stats QualityStdDev.}

\item{statsHeterozygosityRate}{Sample variant stats HeterozygosityRate.}

\item{statsDepthCount}{Sample variant stats DepthCount.}

\item{statsBiotypeCount}{Sample variant stats BiotypeCount.}

\item{statsClinicalSignificanceCount}{Sample variant stats ClinicalSignificanceCount.}

\item{statsConsequenceTypeCount}{Sample variant stats ConsequenceTypeCount.}

\item{member}{User or group id.}

\item{silent}{Boolean to retrieve all possible entries that are queried for, false to raise an exception whenever one of the entries looked for cannot be shown for whichever reason.}

\item{force}{Force the deletion of samples even if they are associated to files, individuals or cohorts.}

\item{emptyFilesAction}{Action to be performed over files that were associated only to the sample to be deleted. Possible actions are NONE, TRASH, DELETE.}

\item{deleteEmptyCohorts}{Boolean indicating if the cohorts associated only to the sample to be deleted should be also deleted.}

\item{includeIndividual}{Include Individual object as an attribute.}

\item{flattenAnnotations}{Flatten the annotations?.}

\item{version}{Comma separated list of sample versions. 'all' to get all the sample versions. Not supported if multiple sample ids are provided.}

\item{deleted}{Boolean to retrieve deleted entries.}

\item{include}{Fields included in the response, whole JSON path must be provided.}

\item{exclude}{Fields excluded in the response, whole JSON path must be provided.}

\item{annotationSetsAction}{Action to be performed if the array of annotationSets is being updated. Allowed values: \link{'ADD SET REMOVE'}}

\item{phenotypesAction}{Action to be performed if the array of phenotypes is being updated \link{SET, ADD, REMOVE}. Allowed values: \link{'ADD SET REMOVE'}}

\item{includeResult}{Flag indicating to include the created or updated document result in the response.}

\item{study}{Study [\link{organization@}project:]study where study and project can be either the ID or UUID.}

\item{action}{Action to be performed: ADD to add new annotations; REPLACE to replace the value of an already existing annotation; SET to set the new list of annotations removing any possible old annotations; REMOVE to remove some annotations; RESET to set some annotations to the default value configured in the corresponding variables of the VariableSet if any. Allowed values: \link{'ADD SET REMOVE RESET REPLACE'}}

\item{data}{Json containing the map of annotations when the action is ADD, SET or REPLACE, a json with only the key 'remove' containing the comma separated variables to be removed as a value when the action is REMOVE or a json with only the key 'reset' containing the comma separated variables that will be set to the default value when the action is RESET.}
}
\description{
This function implements the OpenCGA calls for managing Samples.
The following table summarises the available \emph{actions} for this client:\tabular{llr}{
   endpointName \tab Endpoint WS \tab parameters accepted \cr
   updateAcl \tab /{apiVersion}/samples/acl/{members}/update \tab study, members\link{*}, action\link{*}, body\link{*} \cr
   aggregationStats \tab /{apiVersion}/samples/aggregationStats \tab study, id, uuid, somatic, individualId, fileIds, cohortIds, creationDate, modificationDate, internalStatus, status, processingProduct, processingPreparationMethod, processingExtractionMethod, processingLabSampleId, collectionFrom, collectionType, collectionMethod, phenotypes, annotation, acl, internalRgaStatus, release, snapshot, deleted, statsId, statsVariantCount, statsChromosomeCount, statsTypeCount, statsGenotypeCount, statsTiTvRatio, statsQualityAvg, statsQualityStdDev, statsHeterozygosityRate, statsDepthCount, statsBiotypeCount, statsClinicalSignificanceCount, statsConsequenceTypeCount, field \cr
   loadAnnotationSets \tab /{apiVersion}/samples/annotationSets/load \tab study, variableSetId\link{*}, path\link{*}, parents, annotationSetId, body \cr
   create \tab /{apiVersion}/samples/create \tab include, exclude, study, includeResult, body\link{*} \cr
   distinct \tab /{apiVersion}/samples/distinct \tab study, id, uuid, somatic, individualId, fileIds, cohortIds, creationDate, modificationDate, internalStatus, status, processingProduct, processingPreparationMethod, processingExtractionMethod, processingLabSampleId, collectionFrom, collectionType, collectionMethod, phenotypes, annotation, acl, internalRgaStatus, release, snapshot, deleted, statsId, statsVariantCount, statsChromosomeCount, statsTypeCount, statsGenotypeCount, statsTiTvRatio, statsQualityAvg, statsQualityStdDev, statsHeterozygosityRate, statsDepthCount, statsBiotypeCount, statsClinicalSignificanceCount, statsConsequenceTypeCount, field\link{*} \cr
   load \tab /{apiVersion}/samples/load \tab study, file\link{*}, variableSet \cr
   search \tab /{apiVersion}/samples/search \tab include, exclude, limit, skip, count, includeIndividual, flattenAnnotations, study, id, uuid, somatic, individualId, fileIds, cohortIds, creationDate, modificationDate, internalStatus, status, processingProduct, processingPreparationMethod, processingExtractionMethod, processingLabSampleId, collectionFrom, collectionType, collectionMethod, phenotypes, annotation, acl, internalRgaStatus, release, snapshot, deleted, statsId, statsVariantCount, statsChromosomeCount, statsTypeCount, statsGenotypeCount, statsTiTvRatio, statsQualityAvg, statsQualityStdDev, statsHeterozygosityRate, statsDepthCount, statsBiotypeCount, statsClinicalSignificanceCount, statsConsequenceTypeCount \cr
   acl \tab /{apiVersion}/samples/{samples}/acl \tab samples\link{*}, study, member, silent \cr
   delete \tab /{apiVersion}/samples/{samples}/delete \tab force, emptyFilesAction, deleteEmptyCohorts, study, samples\link{*} \cr
   info \tab /{apiVersion}/samples/{samples}/info \tab include, exclude, includeIndividual, flattenAnnotations, samples\link{*}, study, version, deleted \cr
   update \tab /{apiVersion}/samples/{samples}/update \tab include, exclude, samples\link{*}, study, annotationSetsAction, phenotypesAction, includeResult, body \cr
   updateAnnotationSetsAnnotations \tab /{apiVersion}/samples/{sample}/annotationSets/{annotationSet}/annotations/update \tab sample\link{*}, study, annotationSet\link{*}, action, body \cr
}
}
\section{Endpoint /{apiVersion}/samples/acl/{members}/update}{

Update the set of permissions granted for the member.
}

\section{Endpoint /{apiVersion}/samples/aggregationStats}{

Fetch catalog sample stats.
}

\section{Endpoint /{apiVersion}/samples/annotationSets/load}{

Load annotation sets from a TSV file.
}

\section{Endpoint /{apiVersion}/samples/create}{

Create sample.
}

\section{Endpoint /{apiVersion}/samples/distinct}{

Sample distinct method.
}

\section{Endpoint /{apiVersion}/samples/load}{

Load samples from a ped file \link{EXPERIMENTAL}.
}

\section{Endpoint /{apiVersion}/samples/search}{

Sample search method.
}

\section{Endpoint /{apiVersion}/samples/{samples}/acl}{

Returns the acl of the samples. If member is provided, it will only return the acl for the member.
}

\section{Endpoint /{apiVersion}/samples/{samples}/delete}{

Delete samples.
}

\section{Endpoint /{apiVersion}/samples/{samples}/info}{

Get sample information.
}

\section{Endpoint /{apiVersion}/samples/{samples}/update}{

Update some sample attributes.
}

\section{Endpoint /{apiVersion}/samples/{sample}/annotationSets/{annotationSet}/annotations/update}{

Update annotations from an annotationSet.
}

\seealso{
\url{http://docs.opencb.org/display/opencga/Using+OpenCGA} and the RESTful API documentation
\url{http://bioinfo.hpc.cam.ac.uk/opencga-prod/webservices/}
\link{*}: Required parameter
}
