
# WARNING: AUTOGENERATED CODE
#
#    This code was generated by a tool.
#    Autogenerated on: 2022-06-21 13:26:41
#    
#    Manual changes to this file may cause unexpected behavior in your application.
#    Manual changes to this file will be overwritten if the code is regenerated.


# ##############################################################################
#' GA4GHClient methods
#' @include AllClasses.R
#' @include AllGenerics.R
#' @include commons.R

#' @description This function implements the OpenCGA calls for managing GA4GH.

#' The following table summarises the available *actions* for this client:
#'
#' | endpointName | Endpoint WS | parameters accepted |
#' | -- | :-- | --: |
#' | searchReads | /{apiVersion}/ga4gh/reads/search |  |
#' | fetchReads | /{apiVersion}/ga4gh/reads/{study}/{file} | file[*], study[*], referenceName, start, end, referenceGenome |
#' | responses | /{apiVersion}/ga4gh/responses | chrom[*], pos[*], allele[*], ref, beacon[*] |
#' | searchVariants | /{apiVersion}/ga4gh/variants/search |  |
#'
#' @md
#' @seealso \url{http://docs.opencb.org/display/opencga/Using+OpenCGA} and the RESTful API documentation
#' \url{http://bioinfo.hpc.cam.ac.uk/opencga-prod/webservices/}
#' [*]: Required parameter
#' @export

setMethod("ga4ghClient", "OpencgaR", function(OpencgaR, study, file, endpointName, params=NULL, ...) {
    switch(endpointName,

        #' @section Endpoint /{apiVersion}/ga4gh/reads/search:
        #' Description.

        searchReads=fetchOpenCGA(object=OpencgaR, category="ga4gh", categoryId=NULL, subcategory="reads",
                subcategoryId=NULL, action="search", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/ga4gh/reads/{study}/{file}:
        #' Fetch alignment files using HTSget protocol.
        #' @param file File id, name or path.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param referenceName Reference sequence name (Example: 'chr1', '1' or 'chrX'.
        #' @param start The start position of the range on the reference, 0-based, inclusive.
        #' @param end The end position of the range on the reference, 0-based, exclusive.
        #' @param referenceGenome Reference genome.
        fetchReads=fetchOpenCGA(object=OpencgaR, category="ga4gh/reads", categoryId=study, subcategory=NULL,
                subcategoryId=file, action="NULL", params=params, httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/ga4gh/responses:
        #' Beacon webservice.
        #' @param chrom Chromosome ID. Accepted values: 1-22, X, Y, MT. Note: For compatibility with conventions set by some of the existing beacons, an arbitrary prefix is accepted as well (e.g. chr1 is equivalent to chrom1 and 1).
        #' @param pos Coordinate within a chromosome. Position is a number and is 0-based.
        #' @param allele Any string of nucleotides A,C,T,G or D, I for deletion and insertion, respectively. Note: For compatibility with conventions set by some of the existing beacons, DEL and INS identifiers are also accepted.
        #' @param ref Genome ID. If not specified, all the genomes supported by the given beacons are queried. Note: For compatibility with conventions set by some of the existing beacons, both GRC or HG notation are accepted, case insensitive.
        #' @param beacon Beacon IDs. If specified, only beacons with the given IDs are queried. Responses from all the supported beacons are obtained otherwise. Format: [id1,id2].
        responses=fetchOpenCGA(object=OpencgaR, category="ga4gh", categoryId=NULL, subcategory=NULL,
                subcategoryId=NULL, action="responses", params=params, httpMethod="GET",
                as.queryParam=c("chrom","pos","allele","beacon"), ...),

        #' @section Endpoint /{apiVersion}/ga4gh/variants/search:
        #' Description.

        searchVariants=fetchOpenCGA(object=OpencgaR, category="ga4gh", categoryId=NULL, subcategory="variants",
                subcategoryId=NULL, action="search", params=params, httpMethod="POST", as.queryParam=NULL, ...),
    )
})