
# WARNING: AUTOGENERATED CODE
#
#    This code was generated by a tool.
#    Autogenerated on: 2022-12-20 11:57:32
#    
#    Manual changes to this file may cause unexpected behavior in your application.
#    Manual changes to this file will be overwritten if the code is regenerated.


# ##############################################################################
#' VariantClient methods
#' @include AllClasses.R
#' @include AllGenerics.R
#' @include commons.R

#' @description This function implements the OpenCGA calls for managing Analysis - Variant.

#' The following table summarises the available *actions* for this client:
#'
#' | endpointName | Endpoint WS | parameters accepted |
#' | -- | :-- | --: |
#' | aggregationStats | /{apiVersion}/analysis/variant/aggregationStats | savedFilter, region, type, project, study, cohort, cohortStatsRef, cohortStatsAlt, cohortStatsMaf, cohortStatsMgf, cohortStatsPass, missingAlleles, missingGenotypes, score, annotationExists, gene, ct, xref, biotype, proteinSubstitution, conservation, populationFrequencyAlt, populationFrequencyRef, populationFrequencyMaf, transcriptFlag, geneTraitId, go, expression, proteinKeyword, drug, functionalScore, clinical, clinicalSignificance, clinicalConfirmedStatus, customAnnotation, trait, field |
#' | metadataAnnotation | /{apiVersion}/analysis/variant/annotation/metadata | annotationId, project |
#' | queryAnnotation | /{apiVersion}/analysis/variant/annotation/query | id, region, include, exclude, limit, skip, annotationId |
#' | runCircos | /{apiVersion}/analysis/variant/circos/run | study, body[*] |
#' | deleteCohortStats | /{apiVersion}/analysis/variant/cohort/stats/delete | study, cohort |
#' | infoCohortStats | /{apiVersion}/analysis/variant/cohort/stats/info | study, cohort[*] |
#' | runCohortStats | /{apiVersion}/analysis/variant/cohort/stats/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runExomiser | /{apiVersion}/analysis/variant/exomiser/run | study, jobId, jobDependsOn, jobDescription, jobTags, body[*] |
#' | runExport | /{apiVersion}/analysis/variant/export/run | include, exclude, project, study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | genotypesFamily | /{apiVersion}/analysis/variant/family/genotypes | study, family, clinicalAnalysis, modeOfInheritance[*], penetrance, disorder |
#' | runFamilyQc | /{apiVersion}/analysis/variant/family/qc/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | deleteFile | /{apiVersion}/analysis/variant/file/delete | jobId, jobDescription, jobDependsOn, jobTags, study, file, resume |
#' | runGatk | /{apiVersion}/analysis/variant/gatk/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runGenomePlot | /{apiVersion}/analysis/variant/genomePlot/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runGwas | /{apiVersion}/analysis/variant/gwas/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runIndex | /{apiVersion}/analysis/variant/index/run | study, jobId, jobDependsOn, jobDescription, jobTags, body[*] |
#' | runIndividualQc | /{apiVersion}/analysis/variant/individual/qc/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runInferredSex | /{apiVersion}/analysis/variant/inferredSex/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | queryKnockoutGene | /{apiVersion}/analysis/variant/knockout/gene/query | limit, skip, study, job |
#' | queryKnockoutIndividual | /{apiVersion}/analysis/variant/knockout/individual/query | limit, skip, study, job |
#' | runKnockout | /{apiVersion}/analysis/variant/knockout/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runMendelianError | /{apiVersion}/analysis/variant/mendelianError/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | metadata | /{apiVersion}/analysis/variant/metadata | project, study, file, sample, includeStudy, includeFile, includeSample, include, exclude |
#' | queryMutationalSignature | /{apiVersion}/analysis/variant/mutationalSignature/query | study, sample, ct, biotype, fileData, filter, qual, region, gene, panel, panelModeOfInheritance, panelConfidence, panelFeatureType, panelRoleInCancer, panelIntersection, catalogues, cataloguesContent, fitMethod, nBoot, sigVersion, organ, thresholdPerc, thresholdPval, maxRareSigs, signaturesFile, rareSignaturesFile |
#' | runMutationalSignature | /{apiVersion}/analysis/variant/mutationalSignature/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runPlink | /{apiVersion}/analysis/variant/plink/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | query | /{apiVersion}/analysis/variant/query | include, exclude, limit, skip, count, sort, summary, approximateCount, approximateCountSamplingSize, savedFilter, id, region, type, reference, alternate, project, study, file, filter, qual, fileData, sample, genotype, sampleData, sampleAnnotation, sampleMetadata, unknownGenotype, sampleLimit, sampleSkip, cohort, cohortStatsRef, cohortStatsAlt, cohortStatsMaf, cohortStatsMgf, cohortStatsPass, missingAlleles, missingGenotypes, score, family, familyDisorder, familySegregation, familyMembers, familyProband, includeStudy, includeFile, includeSample, includeSampleData, includeGenotype, includeSampleId, annotationExists, gene, ct, xref, biotype, proteinSubstitution, conservation, populationFrequencyAlt, populationFrequencyRef, populationFrequencyMaf, transcriptFlag, geneTraitId, go, expression, proteinKeyword, drug, functionalScore, clinical, clinicalSignificance, clinicalConfirmedStatus, customAnnotation, panel, panelModeOfInheritance, panelConfidence, panelRoleInCancer, panelFeatureType, panelIntersection, trait |
#' | runRelatedness | /{apiVersion}/analysis/variant/relatedness/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runRvtests | /{apiVersion}/analysis/variant/rvtests/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | aggregationStatsSample | /{apiVersion}/analysis/variant/sample/aggregationStats | savedFilter, region, type, project, study, file, filter, sample, genotype, sampleAnnotation, family, familyDisorder, familySegregation, familyMembers, familyProband, ct, biotype, populationFrequencyAlt, clinical, clinicalSignificance, clinicalConfirmedStatus, field |
#' | runSampleEligibility | /{apiVersion}/analysis/variant/sample/eligibility/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runSampleQc | /{apiVersion}/analysis/variant/sample/qc/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | querySample | /{apiVersion}/analysis/variant/sample/query | limit, skip, variant, study, genotype |
#' | runSample | /{apiVersion}/analysis/variant/sample/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | querySampleStats | /{apiVersion}/analysis/variant/sample/stats/query | region, type, study, file, filter, sampleData, ct, biotype, transcriptFlag, populationFrequencyAlt, clinical, clinicalSignificance, clinicalConfirmedStatus, study, filterTranscript, sample[*] |
#' | runSampleStats | /{apiVersion}/analysis/variant/sample/stats/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runStatsExport | /{apiVersion}/analysis/variant/stats/export/run | project, study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#' | runStats | /{apiVersion}/analysis/variant/stats/run | study, jobId, jobDescription, jobDependsOn, jobTags, body[*] |
#'
#' @md
#' @seealso \url{http://docs.opencb.org/display/opencga/Using+OpenCGA} and the RESTful API documentation
#' \url{http://bioinfo.hpc.cam.ac.uk/opencga-prod/webservices/}
#' [*]: Required parameter
#' @export

setMethod("variantClient", "OpencgaR", function(OpencgaR, endpointName, params=NULL, ...) {
    switch(endpointName,

        #' @section Endpoint /{apiVersion}/analysis/variant/aggregationStats:
        #' Calculate and fetch aggregation stats.
        #' @param savedFilter Use a saved filter at User level.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param type List of types, accepted values are SNV, MNV, INDEL, SV, COPY_NUMBER, COPY_NUMBER_LOSS, COPY_NUMBER_GAIN, INSERTION, DELETION, DUPLICATION, TANDEM_DUPLICATION, BREAKEND, e.g. SNV,INDEL.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Filter variants from the given studies, these can be either the numeric ID or the alias with the format user@project:study.
        #' @param cohort Select variants with calculated stats for the selected cohorts.
        #' @param cohortStatsRef Reference Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsAlt Alternate Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsMaf Minor Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsMgf Minor Genotype Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsPass Filter PASS frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL>0.8.
        #' @param missingAlleles Number of missing alleles: [{study:}]{cohort}[<|>|<=|>=]{number}.
        #' @param missingGenotypes Number of missing genotypes: [{study:}]{cohort}[<|>|<=|>=]{number}.
        #' @param score Filter by variant score: [{study:}]{score}[<|>|<=|>=]{number}.
        #' @param annotationExists Return only annotated variants.
        #' @param gene List of genes, most gene IDs are accepted (HGNC, Ensembl gene, ...). This is an alias to 'xref' parameter.
        #' @param ct List of SO consequence types, e.g. missense_variant,stop_lost or SO:0001583,SO:0001578. Accepts aliases 'loss_of_function' and 'protein_altering'.
        #' @param xref List of any external reference, these can be genes, proteins or variants. Accepted IDs include HGNC, Ensembl genes, dbSNP, ClinVar, HPO, Cosmic, ...
        #' @param biotype List of biotypes, e.g. protein_coding.
        #' @param proteinSubstitution Protein substitution scores include SIFT and PolyPhen. You can query using the score {protein_score}[<|>|<=|>=]{number} or the description {protein_score}[~=|=]{description} e.g. polyphen>0.1,sift=tolerant.
        #' @param conservation Filter by conservation score: {conservation_score}[<|>|<=|>=]{number} e.g. phastCons>0.5,phylop<0.1,gerp>0.1.
        #' @param populationFrequencyAlt Alternate Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param populationFrequencyRef Reference Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param populationFrequencyMaf Population minor allele frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param transcriptFlag List of transcript flags. e.g. canonical, CCDS, basic, LRG, MANE Select, MANE Plus Clinical, EGLH_HaemOnc, TSO500.
        #' @param geneTraitId List of gene trait association id. e.g. "umls:C0007222" , "OMIM:269600".
        #' @param go List of GO (Gene Ontology) terms. e.g. "GO:0002020".
        #' @param expression List of tissues of interest. e.g. "lung".
        #' @param proteinKeyword List of Uniprot protein variant annotation keywords.
        #' @param drug List of drug names.
        #' @param functionalScore Functional score: {functional_score}[<|>|<=|>=]{number} e.g. cadd_scaled>5.2 , cadd_raw<=0.3.
        #' @param clinical Clinical source: clinvar, cosmic.
        #' @param clinicalSignificance Clinical significance: benign, likely_benign, likely_pathogenic, pathogenic.
        #' @param clinicalConfirmedStatus Clinical confirmed status.
        #' @param customAnnotation Custom annotation: {key}[<|>|<=|>=]{number} or {key}[~=|=]{text}.
        #' @param trait List of traits, based on ClinVar, HPO, COSMIC, i.e.: IDs, histologies, descriptions,...
        #' @param field List of facet fields separated by semicolons, e.g.: studies;type. For nested faceted fields use >>, e.g.: chromosome>>type;percentile(gerp).
        aggregationStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant",
                subcategoryId=NULL, action="aggregationStats", params=params, httpMethod="GET", as.queryParam=NULL,
                ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/annotation/metadata:
        #' Read variant annotations metadata from any saved versions.
        #' @param annotationId Annotation identifier.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        metadataAnnotation=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/annotation", subcategoryId=NULL, action="metadata", params=params,
                httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/annotation/query:
        #' Query variant annotations from any saved versions.
        #' @param id List of IDs, these can be rs IDs (dbSNP) or variants in the format chrom:start:ref:alt, e.g. rs116600158,19:7177679:C:T.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param include Fields included in the response, whole JSON path must be provided.
        #' @param exclude Fields excluded in the response, whole JSON path must be provided.
        #' @param limit Number of results to be returned.
        #' @param skip Number of results to skip.
        #' @param annotationId Annotation identifier.
        queryAnnotation=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/annotation", subcategoryId=NULL, action="query", params=params, httpMethod="GET",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/circos/run:
        #' Generate a Circos plot for a given sample.
        #' @param study study.
        #' @param data Circos analysis params to customize the plot. These parameters include the title,  the plot density (i.e., the number of points to display), the general query and the list of tracks. Currently, the supported track types are: COPY-NUMBER, INDEL, REARRANGEMENT and SNV. In addition, each track can contain a specific query.
        runCircos=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/circos",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/cohort/stats/delete:
        #' Delete cohort variant stats from a cohort.
        #' @param study study.
        #' @param cohort Cohort ID or UUID.
        deleteCohortStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/cohort/stats", subcategoryId=NULL, action="delete", params=params,
                httpMethod="DELETE", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/cohort/stats/info:
        #' Read cohort variant stats from list of cohorts.
        #' @param study study.
        #' @param cohort Comma separated list of cohort IDs or UUIDs up to a maximum of 100.
        infoCohortStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/cohort/stats", subcategoryId=NULL, action="info", params=params, httpMethod="GET",
                as.queryParam=c("cohort"), ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/cohort/stats/run:
        #' Compute cohort variant stats for the selected list of samples.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Cohort variant stats params.
        runCohortStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/cohort/stats", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/exomiser/run:
        #' The Exomiser is a Java program that finds potential disease-causing variants from whole-exome or whole-genome sequencing data.
        #' @param study study.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobDescription Job description.
        #' @param jobTags Job tags.
        #' @param data Exomiser parameters.
        runExomiser=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/exomiser",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/export/run:
        #' Filter and export variants from the variant storage to a file.
        #' @param include Fields included in the response, whole JSON path must be provided.
        #' @param exclude Fields excluded in the response, whole JSON path must be provided.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Variant export params.
        runExport=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/export",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/family/genotypes:
        #' Calculate the possible genotypes for the members of a family.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param family Family id.
        #' @param clinicalAnalysis Clinical analysis id.
        #' @param modeOfInheritance Mode of inheritance. Allowed values: ['AUTOSOMAL_DOMINANT AUTOSOMAL_RECESSIVE X_LINKED_DOMINANT X_LINKED_RECESSIVE Y_LINKED MITOCHONDRIAL DE_NOVO MENDELIAN_ERROR COMPOUND_HETEROZYGOUS UNKNOWN']
        #' @param penetrance Penetrance. Allowed values: ['COMPLETE INCOMPLETE UNKNOWN']
        #' @param disorder Disorder id.
        genotypesFamily=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/family", subcategoryId=NULL, action="genotypes", params=params, httpMethod="GET",
                as.queryParam=c("modeOfInheritance"), ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/family/qc/run:
        #' Run quality control (QC) for a given family. It computes the relatedness scores among the family members.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Family QC analysis params. Family ID. Relatedness method, by default 'PLINK/IBD'. Minor  allele frequence (MAF) is used to filter variants before computing relatedness, e.g.: 1000G:CEU>0.35 or cohort:ALL>0.05.
        runFamilyQc=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/family/qc", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/file/delete:
        #'  [DEPRECATED] Use operation/variant/delete.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param file Files to remove.
        #' @param resume Resume a previously failed indexation.
        deleteFile=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/file",
                subcategoryId=NULL, action="delete", params=params, httpMethod="DELETE", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/gatk/run:
        #' GATK is a Genome Analysis Toolkit for variant discovery in high-throughput sequencing data. Supported Gatk commands: HaplotypeCaller.
        #' @param study study.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Gatk parameters. Supported Gatk commands: HaplotypeCaller.
        runGatk=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/gatk",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/genomePlot/run:
        #' Generate a genome plot for a given sample.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Genome plot analysis params to customize the plot. The configuration file includes the title,  the plot density (i.e., the number of points to display), the general query and the list of tracks. Currently, the supported track types are: COPY-NUMBER, INDEL, REARRANGEMENT and SNV. In addition, each track can contain a specific query.
        runGenomePlot=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/genomePlot", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/gwas/run:
        #' Run a Genome Wide Association Study between two cohorts.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Gwas analysis params.
        runGwas=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/gwas",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/index/run:
        #'  [DEPRECATED] Use operation/variant/index.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobDescription Job description.
        #' @param jobTags Job tags.
        #' @param data Variant index params.
        runIndex=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/index",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/individual/qc/run:
        #' Run quality control (QC) for a given individual. It includes inferred sex and  mendelian errors (UDP).
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Individual QC analysis params.
        runIndividualQc=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/individual/qc", subcategoryId=NULL, action="run", params=params,
                httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/inferredSex/run:
        #' Infer sex from chromosome mean coverages.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Inferred sex analysis params.
        runInferredSex=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/inferredSex", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/knockout/gene/query:
        #' Fetch values from KnockoutAnalysis result, by genes.
        #' @param limit Number of results to be returned.
        #' @param skip Number of results to skip.
        #' @param study study.
        #' @param job Job ID or UUID.
        queryKnockoutGene=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/knockout/gene", subcategoryId=NULL, action="query", params=params,
                httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/knockout/individual/query:
        #' Fetch values from KnockoutAnalysis result, by individuals.
        #' @param limit Number of results to be returned.
        #' @param skip Number of results to skip.
        #' @param study study.
        #' @param job Job ID or UUID.
        queryKnockoutIndividual=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/knockout/individual", subcategoryId=NULL, action="query", params=params,
                httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/knockout/run:
        #' Obtains the list of knocked out genes for each sample.
        #' @param study study.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Gene knockout analysis params.
        runKnockout=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/knockout",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/mendelianError/run:
        #' Run mendelian error analysis to infer uniparental disomy regions.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Mendelian error analysis params.
        runMendelianError=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/mendelianError", subcategoryId=NULL, action="run", params=params,
                httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/metadata:
        #' .
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Filter variants from the given studies, these can be either the numeric ID or the alias with the format user@project:study.
        #' @param file Filter variants from the files specified. This will set includeFile parameter when not provided.
        #' @param sample Filter variants by sample genotype. This will automatically set 'includeSample' parameter when not provided. This filter accepts multiple 3 forms: 1) List of samples: Samples that contain the main variant. Accepts AND (;) and OR (,) operators.  e.g. HG0097,HG0098 . 2) List of samples with genotypes: {sample}:{gt1},{gt2}. Accepts AND (;) and OR (,) operators.  e.g. HG0097:0/0;HG0098:0/1,1/1 . Unphased genotypes (e.g. 0/1, 1/1) will also include phased genotypes (e.g. 0|1, 1|0, 1|1), but not vice versa. When filtering by multi-allelic genotypes, any secondary allele will match, regardless of its position e.g. 1/2 will match with genotypes 1/2, 1/3, 1/4, .... Genotype aliases accepted: HOM_REF, HOM_ALT, HET, HET_REF, HET_ALT, HET_MISS and MISS  e.g. HG0097:HOM_REF;HG0098:HET_REF,HOM_ALT . 3) Sample with segregation mode: {sample}:{segregation}. Only one sample accepted.Accepted segregation modes: [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, deNovoStrict, mendelianError, compoundHeterozygous ]. Value is case insensitive. e.g. HG0097:DeNovo Sample must have parents defined and indexed. .
        #' @param includeStudy List of studies to include in the result. Accepts 'all' and 'none'.
        #' @param includeFile List of files to be returned. Accepts 'all' and 'none'. If undefined, automatically includes files used for filtering. If none, no file is included.
        #' @param includeSample List of samples to be included in the result. Accepts 'all' and 'none'. If undefined, automatically includes samples used for filtering. If none, no sample is included.
        #' @param include Fields included in the response, whole JSON path must be provided.
        #' @param exclude Fields excluded in the response, whole JSON path must be provided.
        metadata=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant",
                subcategoryId=NULL, action="metadata", params=params, httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/mutationalSignature/query:
        #' Run mutational signature analysis for a given sample. Use context index.
        #' @param study Filter variants from the given studies, these can be either the numeric ID or the alias with the format user@project:study.
        #' @param sample Sample name.
        #' @param ct List of SO consequence types, e.g. missense_variant,stop_lost or SO:0001583,SO:0001578. Accepts aliases 'loss_of_function' and 'protein_altering'.
        #' @param biotype List of biotypes, e.g. protein_coding.
        #' @param fileData Filter by file data (i.e. FILTER, QUAL and INFO columns from VCF file). [{file}:]{key}{op}{value}[,;]* . If no file is specified, will use all files from "file" filter. e.g. AN>200 or file_1.vcf:AN>200;file_2.vcf:AN<10 . Many fields can be combined. e.g. file_1.vcf:AN>200;DB=true;file_2.vcf:AN<10,FILTER=PASS,LowDP.
        #' @param filter Specify the FILTER for any of the files. If 'file' filter is provided, will match the file and the filter. e.g.: PASS,LowGQX.
        #' @param qual Specify the QUAL for any of the files. If 'file' filter is provided, will match the file and the qual. e.g.: >123.4.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param gene List of genes, most gene IDs are accepted (HGNC, Ensembl gene, ...). This is an alias to 'xref' parameter.
        #' @param panel Filter by genes from the given disease panel.
        #' @param panelModeOfInheritance Filter genes from specific panels that match certain mode of inheritance. Accepted values : [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, mendelianError, compoundHeterozygous ].
        #' @param panelConfidence Filter genes from specific panels that match certain confidence. Accepted values : [ high, medium, low, rejected ].
        #' @param panelFeatureType Filter elements from specific panels by type. Accepted values : [ gene, region, str, variant ].
        #' @param panelRoleInCancer Filter genes from specific panels that match certain role in cancer. Accepted values : [ both, oncogene, tumorSuppressorGene, fusion ].
        #' @param panelIntersection Intersect panel genes and regions with given genes and regions from que input query. This will prevent returning variants from regions out of the panel.
        #' @param catalogues File name containing mutational catalogues. Each sample catalogue is in a column, with sample names as column headers and channel.
        #' @param cataloguesContent Mutational catalogues. Each sample catalogue is in a column, with sample names as column headers and channel.
        #' @param fitMethod Either Fit or FitMS. If not specified then FitMS.
        #' @param nBoot Number of bootstrap to be used.
        #' @param sigVersion Either COSMICv2, COSMICv3.2, RefSigv1 or RefSigv2. If not specified RefSigv2.
        #' @param organ When using RefSigv1 or RefSigv2 as SIGVERSION, organ-specific signatures will be used. If SIGVERSION is COSMICv2 or COSMICv3.2, then a selection of signatures found in the given organ will be used. Available organs depend on the selected SIGVERSION. For RefSigv1 or RefSigv2: Biliary, Bladder, Bone_SoftTissue, Breast, Cervix (v1 only), CNS, Colorectal, Esophagus, Head_neck, Kidney, Liver, Lung, Lymphoid, NET (v2 only), Oral_Oropharyngeal (v2 only), Ovary, Pancreas, Prostate, Skin, Stomach, Uterus.
        #' @param thresholdPerc Threshold in percentage of total mutations in a sample, only exposures larger than THRPERC are considered. If not specified 5.
        #' @param thresholdPval P-value to determine the empirical probability that the exposure is lower than the threshold. If not specified then 0.05.
        #' @param maxRareSigs Maximum number of rare signatures that are allowed to be present in each sample. If not specified 1.
        #' @param signaturesFile The file name containing mutational signatures. Each signature is in a column, with signature names as column hearders and channel names as row names in the first column with no header. Each column must sum to 1. Use only to provide your own signatures. When fitmethod=FitMS, these signatures are considered common signatures.
        #' @param rareSignaturesFile The file name containing mutational signatures. Each signature is in a column, with signature names as column hearders and channel names as row names in the first column with no header. Each column must sum to 1. Use only to provide your own signatures. When fitmethod=FitMS, these signatures are considered rare signatures.
        queryMutationalSignature=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/mutationalSignature", subcategoryId=NULL, action="query", params=params,
                httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/mutationalSignature/run:
        #' Run mutational signature analysis for a given sample.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Mutational signature analysis params.
        runMutationalSignature=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/mutationalSignature", subcategoryId=NULL, action="run", params=params,
                httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/plink/run:
        #' Plink is a whole genome association analysis toolset, designed to perform a range of basic, large-scale analyses.
        #' @param study study.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Plink params.
        runPlink=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/plink",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/query:
        #' Filter and fetch variants from indexed VCF files in the variant storage.
        #' @param include Fields included in the response, whole JSON path must be provided.
        #' @param exclude Fields excluded in the response, whole JSON path must be provided.
        #' @param limit Number of results to be returned.
        #' @param skip Number of results to skip.
        #' @param count Get the total number of results matching the query. Deactivated by default.
        #' @param sort Sort the results.
        #' @param summary Fast fetch of main variant parameters.
        #' @param approximateCount Get an approximate count, instead of an exact total count. Reduces execution time.
        #' @param approximateCountSamplingSize Sampling size to get the approximate count. Larger values increase accuracy but also increase execution time.
        #' @param savedFilter Use a saved filter at User level.
        #' @param id List of IDs, these can be rs IDs (dbSNP) or variants in the format chrom:start:ref:alt, e.g. rs116600158,19:7177679:C:T.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param type List of types, accepted values are SNV, MNV, INDEL, SV, COPY_NUMBER, COPY_NUMBER_LOSS, COPY_NUMBER_GAIN, INSERTION, DELETION, DUPLICATION, TANDEM_DUPLICATION, BREAKEND, e.g. SNV,INDEL.
        #' @param reference Reference allele.
        #' @param alternate Main alternate allele.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Filter variants from the given studies, these can be either the numeric ID or the alias with the format user@project:study.
        #' @param file Filter variants from the files specified. This will set includeFile parameter when not provided.
        #' @param filter Specify the FILTER for any of the files. If 'file' filter is provided, will match the file and the filter. e.g.: PASS,LowGQX.
        #' @param qual Specify the QUAL for any of the files. If 'file' filter is provided, will match the file and the qual. e.g.: >123.4.
        #' @param fileData Filter by file data (i.e. FILTER, QUAL and INFO columns from VCF file). [{file}:]{key}{op}{value}[,;]* . If no file is specified, will use all files from "file" filter. e.g. AN>200 or file_1.vcf:AN>200;file_2.vcf:AN<10 . Many fields can be combined. e.g. file_1.vcf:AN>200;DB=true;file_2.vcf:AN<10,FILTER=PASS,LowDP.
        #' @param sample Filter variants by sample genotype. This will automatically set 'includeSample' parameter when not provided. This filter accepts multiple 3 forms: 1) List of samples: Samples that contain the main variant. Accepts AND (;) and OR (,) operators.  e.g. HG0097,HG0098 . 2) List of samples with genotypes: {sample}:{gt1},{gt2}. Accepts AND (;) and OR (,) operators.  e.g. HG0097:0/0;HG0098:0/1,1/1 . Unphased genotypes (e.g. 0/1, 1/1) will also include phased genotypes (e.g. 0|1, 1|0, 1|1), but not vice versa. When filtering by multi-allelic genotypes, any secondary allele will match, regardless of its position e.g. 1/2 will match with genotypes 1/2, 1/3, 1/4, .... Genotype aliases accepted: HOM_REF, HOM_ALT, HET, HET_REF, HET_ALT, HET_MISS and MISS  e.g. HG0097:HOM_REF;HG0098:HET_REF,HOM_ALT . 3) Sample with segregation mode: {sample}:{segregation}. Only one sample accepted.Accepted segregation modes: [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, deNovoStrict, mendelianError, compoundHeterozygous ]. Value is case insensitive. e.g. HG0097:DeNovo Sample must have parents defined and indexed. .
        #' @param genotype Samples with a specific genotype: {samp_1}:{gt_1}(,{gt_n})*(;{samp_n}:{gt_1}(,{gt_n})*)* e.g. HG0097:0/0;HG0098:0/1,1/1. Unphased genotypes (e.g. 0/1, 1/1) will also include phased genotypes (e.g. 0|1, 1|0, 1|1), but not vice versa. When filtering by multi-allelic genotypes, any secondary allele will match, regardless of its position e.g. 1/2 will match with genotypes 1/2, 1/3, 1/4, .... Genotype aliases accepted: HOM_REF, HOM_ALT, HET, HET_REF, HET_ALT, HET_MISS and MISS  e.g. HG0097:HOM_REF;HG0098:HET_REF,HOM_ALT. This will automatically set 'includeSample' parameter when not provided.
        #' @param sampleData Filter by any SampleData field from samples. [{sample}:]{key}{op}{value}[,;]* . If no sample is specified, will use all samples from "sample" or "genotype" filter. e.g. DP>200 or HG0097:DP>200,HG0098:DP<10 . Many FORMAT fields can be combined. e.g. HG0097:DP>200;GT=1/1,0/1,HG0098:DP<10.
        #' @param sampleAnnotation Selects some samples using metadata information from Catalog. e.g. age>20;phenotype=hpo:123,hpo:456;name=smith.
        #' @param sampleMetadata Return the samples metadata group by study. Sample names will appear in the same order as their corresponding genotypes.
        #' @param unknownGenotype Returned genotype for unknown genotypes. Common values: [0/0, 0|0, ./.].
        #' @param sampleLimit Limit the number of samples to be included in the result.
        #' @param sampleSkip Skip some samples from the result. Useful for sample pagination.
        #' @param cohort Select variants with calculated stats for the selected cohorts.
        #' @param cohortStatsRef Reference Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsAlt Alternate Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsMaf Minor Allele Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsMgf Minor Genotype Frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL<=0.4.
        #' @param cohortStatsPass Filter PASS frequency: [{study:}]{cohort}[<|>|<=|>=]{number}. e.g. ALL>0.8.
        #' @param missingAlleles Number of missing alleles: [{study:}]{cohort}[<|>|<=|>=]{number}.
        #' @param missingGenotypes Number of missing genotypes: [{study:}]{cohort}[<|>|<=|>=]{number}.
        #' @param score Filter by variant score: [{study:}]{score}[<|>|<=|>=]{number}.
        #' @param family Filter variants where any of the samples from the given family contains the variant (HET or HOM_ALT).
        #' @param familyDisorder Specify the disorder to use for the family segregation.
        #' @param familySegregation Filter by segregation mode from a given family. Accepted values: [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, deNovoStrict, mendelianError, compoundHeterozygous ].
        #' @param familyMembers Sub set of the members of a given family.
        #' @param familyProband Specify the proband child to use for the family segregation.
        #' @param includeStudy List of studies to include in the result. Accepts 'all' and 'none'.
        #' @param includeFile List of files to be returned. Accepts 'all' and 'none'. If undefined, automatically includes files used for filtering. If none, no file is included.
        #' @param includeSample List of samples to be included in the result. Accepts 'all' and 'none'. If undefined, automatically includes samples used for filtering. If none, no sample is included.
        #' @param includeSampleData List of Sample Data keys (i.e. FORMAT column from VCF file) from Sample Data to include in the output. e.g: DP,AD. Accepts 'all' and 'none'.
        #' @param includeGenotype Include genotypes, apart of other formats defined with includeFormat.
        #' @param includeSampleId Include sampleId on each result.
        #' @param annotationExists Return only annotated variants.
        #' @param gene List of genes, most gene IDs are accepted (HGNC, Ensembl gene, ...). This is an alias to 'xref' parameter.
        #' @param ct List of SO consequence types, e.g. missense_variant,stop_lost or SO:0001583,SO:0001578. Accepts aliases 'loss_of_function' and 'protein_altering'.
        #' @param xref List of any external reference, these can be genes, proteins or variants. Accepted IDs include HGNC, Ensembl genes, dbSNP, ClinVar, HPO, Cosmic, ...
        #' @param biotype List of biotypes, e.g. protein_coding.
        #' @param proteinSubstitution Protein substitution scores include SIFT and PolyPhen. You can query using the score {protein_score}[<|>|<=|>=]{number} or the description {protein_score}[~=|=]{description} e.g. polyphen>0.1,sift=tolerant.
        #' @param conservation Filter by conservation score: {conservation_score}[<|>|<=|>=]{number} e.g. phastCons>0.5,phylop<0.1,gerp>0.1.
        #' @param populationFrequencyAlt Alternate Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param populationFrequencyRef Reference Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param populationFrequencyMaf Population minor allele frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param transcriptFlag List of transcript flags. e.g. canonical, CCDS, basic, LRG, MANE Select, MANE Plus Clinical, EGLH_HaemOnc, TSO500.
        #' @param geneTraitId List of gene trait association id. e.g. "umls:C0007222" , "OMIM:269600".
        #' @param go List of GO (Gene Ontology) terms. e.g. "GO:0002020".
        #' @param expression List of tissues of interest. e.g. "lung".
        #' @param proteinKeyword List of Uniprot protein variant annotation keywords.
        #' @param drug List of drug names.
        #' @param functionalScore Functional score: {functional_score}[<|>|<=|>=]{number} e.g. cadd_scaled>5.2 , cadd_raw<=0.3.
        #' @param clinical Clinical source: clinvar, cosmic.
        #' @param clinicalSignificance Clinical significance: benign, likely_benign, likely_pathogenic, pathogenic.
        #' @param clinicalConfirmedStatus Clinical confirmed status.
        #' @param customAnnotation Custom annotation: {key}[<|>|<=|>=]{number} or {key}[~=|=]{text}.
        #' @param panel Filter by genes from the given disease panel.
        #' @param panelModeOfInheritance Filter genes from specific panels that match certain mode of inheritance. Accepted values : [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, mendelianError, compoundHeterozygous ].
        #' @param panelConfidence Filter genes from specific panels that match certain confidence. Accepted values : [ high, medium, low, rejected ].
        #' @param panelRoleInCancer Filter genes from specific panels that match certain role in cancer. Accepted values : [ both, oncogene, tumorSuppressorGene, fusion ].
        #' @param panelFeatureType Filter elements from specific panels by type. Accepted values : [ gene, region, str, variant ].
        #' @param panelIntersection Intersect panel genes and regions with given genes and regions from que input query. This will prevent returning variants from regions out of the panel.
        #' @param trait List of traits, based on ClinVar, HPO, COSMIC, i.e.: IDs, histologies, descriptions,...
        query=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant",
                subcategoryId=NULL, action="query", params=params, httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/relatedness/run:
        #' Compute a score to quantify relatedness between samples.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Relatedness analysis params.
        runRelatedness=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/relatedness", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/rvtests/run:
        #' Rvtests is a flexible software package for genetic association studies. Supported RvTests commands: rvtest, vcf2kinship.
        #' @param study study.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data RvTests parameters. Supported RvTests commands: rvtest, vcf2kinship.
        runRvtests=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/rvtests",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/aggregationStats:
        #' Calculate and fetch sample aggregation stats.
        #' @param savedFilter Use a saved filter at User level.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param type List of types, accepted values are SNV, MNV, INDEL, SV, COPY_NUMBER, COPY_NUMBER_LOSS, COPY_NUMBER_GAIN, INSERTION, DELETION, DUPLICATION, TANDEM_DUPLICATION, BREAKEND, e.g. SNV,INDEL.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Filter variants from the given studies, these can be either the numeric ID or the alias with the format user@project:study.
        #' @param file Filter variants from the files specified. This will set includeFile parameter when not provided.
        #' @param filter Specify the FILTER for any of the files. If 'file' filter is provided, will match the file and the filter. e.g.: PASS,LowGQX.
        #' @param sample Filter variants by sample genotype. This will automatically set 'includeSample' parameter when not provided. This filter accepts multiple 3 forms: 1) List of samples: Samples that contain the main variant. Accepts AND (;) and OR (,) operators.  e.g. HG0097,HG0098 . 2) List of samples with genotypes: {sample}:{gt1},{gt2}. Accepts AND (;) and OR (,) operators.  e.g. HG0097:0/0;HG0098:0/1,1/1 . Unphased genotypes (e.g. 0/1, 1/1) will also include phased genotypes (e.g. 0|1, 1|0, 1|1), but not vice versa. When filtering by multi-allelic genotypes, any secondary allele will match, regardless of its position e.g. 1/2 will match with genotypes 1/2, 1/3, 1/4, .... Genotype aliases accepted: HOM_REF, HOM_ALT, HET, HET_REF, HET_ALT, HET_MISS and MISS  e.g. HG0097:HOM_REF;HG0098:HET_REF,HOM_ALT . 3) Sample with segregation mode: {sample}:{segregation}. Only one sample accepted.Accepted segregation modes: [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, deNovoStrict, mendelianError, compoundHeterozygous ]. Value is case insensitive. e.g. HG0097:DeNovo Sample must have parents defined and indexed. .
        #' @param genotype Samples with a specific genotype: {samp_1}:{gt_1}(,{gt_n})*(;{samp_n}:{gt_1}(,{gt_n})*)* e.g. HG0097:0/0;HG0098:0/1,1/1. Unphased genotypes (e.g. 0/1, 1/1) will also include phased genotypes (e.g. 0|1, 1|0, 1|1), but not vice versa. When filtering by multi-allelic genotypes, any secondary allele will match, regardless of its position e.g. 1/2 will match with genotypes 1/2, 1/3, 1/4, .... Genotype aliases accepted: HOM_REF, HOM_ALT, HET, HET_REF, HET_ALT, HET_MISS and MISS  e.g. HG0097:HOM_REF;HG0098:HET_REF,HOM_ALT. This will automatically set 'includeSample' parameter when not provided.
        #' @param sampleAnnotation Selects some samples using metadata information from Catalog. e.g. age>20;phenotype=hpo:123,hpo:456;name=smith.
        #' @param family Filter variants where any of the samples from the given family contains the variant (HET or HOM_ALT).
        #' @param familyDisorder Specify the disorder to use for the family segregation.
        #' @param familySegregation Filter by segregation mode from a given family. Accepted values: [ autosomalDominant, autosomalRecessive, XLinkedDominant, XLinkedRecessive, YLinked, mitochondrial, deNovo, deNovoStrict, mendelianError, compoundHeterozygous ].
        #' @param familyMembers Sub set of the members of a given family.
        #' @param familyProband Specify the proband child to use for the family segregation.
        #' @param ct List of SO consequence types, e.g. missense_variant,stop_lost or SO:0001583,SO:0001578. Accepts aliases 'loss_of_function' and 'protein_altering'.
        #' @param biotype List of biotypes, e.g. protein_coding.
        #' @param populationFrequencyAlt Alternate Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param clinical Clinical source: clinvar, cosmic.
        #' @param clinicalSignificance Clinical significance: benign, likely_benign, likely_pathogenic, pathogenic.
        #' @param clinicalConfirmedStatus Clinical confirmed status.
        #' @param field List of facet fields separated by semicolons, e.g.: studies;type. For nested faceted fields use >>, e.g.: chromosome>>type . Accepted values: chromosome, type, genotype, consequenceType, biotype, clinicalSignificance, dp, qual, filter.
        aggregationStatsSample=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/sample", subcategoryId=NULL, action="aggregationStats", params=params,
                httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/eligibility/run:
        #' Filter samples by a complex query involving metadata and variants data.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data .
        runSampleEligibility=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/sample/eligibility", subcategoryId=NULL, action="run", params=params,
                httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/qc/run:
        #' Run quality control (QC) for a given sample. It includes variant stats, and if the sample is somatic, mutational signature and genome plot are calculated.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Sample QC analysis params. Mutational signature and genome plot are calculated for somatic samples only.
        runSampleQc=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/sample/qc", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/query:
        #' Get sample data of a given variant.
        #' @param limit Number of results to be returned.
        #' @param skip Number of results to skip.
        #' @param variant Variant.
        #' @param study Study where all the samples belong to.
        #' @param genotype Genotypes that the sample must have to be selected.
        querySample=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/sample",
                subcategoryId=NULL, action="query", params=params, httpMethod="GET", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/run:
        #' Get samples given a set of variants.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Sample variant filter params.
        runSample=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/sample",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/stats/query:
        #' Obtain sample variant stats from a sample.
        #' @param region List of regions, these can be just a single chromosome name or regions in the format chr:start-end, e.g.: 2,3:100000-200000.
        #' @param type List of types, accepted values are SNV, MNV, INDEL, SV, COPY_NUMBER, COPY_NUMBER_LOSS, COPY_NUMBER_GAIN, INSERTION, DELETION, DUPLICATION, TANDEM_DUPLICATION, BREAKEND, e.g. SNV,INDEL.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param file Filter variants from the files specified. This will set includeFile parameter when not provided.
        #' @param filter Specify the FILTER for any of the files. If 'file' filter is provided, will match the file and the filter. e.g.: PASS,LowGQX.
        #' @param sampleData Filter by any SampleData field from samples. [{sample}:]{key}{op}{value}[,;]* . If no sample is specified, will use all samples from "sample" or "genotype" filter. e.g. DP>200 or HG0097:DP>200,HG0098:DP<10 . Many FORMAT fields can be combined. e.g. HG0097:DP>200;GT=1/1,0/1,HG0098:DP<10.
        #' @param ct List of SO consequence types, e.g. missense_variant,stop_lost or SO:0001583,SO:0001578. Accepts aliases 'loss_of_function' and 'protein_altering'.
        #' @param biotype List of biotypes, e.g. protein_coding.
        #' @param transcriptFlag List of transcript flags. e.g. canonical, CCDS, basic, LRG, MANE Select, MANE Plus Clinical, EGLH_HaemOnc, TSO500.
        #' @param populationFrequencyAlt Alternate Population Frequency: {study}:{population}[<|>|<=|>=]{number}. e.g. 1000G:ALL<0.01.
        #' @param clinical Clinical source: clinvar, cosmic.
        #' @param clinicalSignificance Clinical significance: benign, likely_benign, likely_pathogenic, pathogenic.
        #' @param clinicalConfirmedStatus Clinical confirmed status.
        #' @param filterTranscript Do filter transcripts when obtaining transcript counts.
        #' @param sample Sample ID.
        querySampleStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/sample/stats", subcategoryId=NULL, action="query", params=params,
                httpMethod="GET", as.queryParam=c("sample"), ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/sample/stats/run:
        #' Compute sample variant stats for the selected list of samples.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Sample variant stats params. Use index=true and indexId='' to store the result in catalog sample QC. indexId=ALL requires an empty query. Use sample=all to compute sample stats of all samples in the variant storage.
        runSampleStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/sample/stats", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/stats/export/run:
        #' Export calculated variant stats and frequencies.
        #' @param project Project [user@]project where project can be either the ID or the alias.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Variant stats export params.
        runStatsExport=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL,
                subcategory="variant/stats/export", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),

        #' @section Endpoint /{apiVersion}/analysis/variant/stats/run:
        #' Compute variant stats for any cohort and any set of variants.
        #' @param study Study [[user@]project:]study where study and project can be either the ID or UUID.
        #' @param jobId Job ID. It must be a unique string within the study. An ID will be autogenerated automatically if not provided.
        #' @param jobDescription Job description.
        #' @param jobDependsOn Comma separated list of existing job IDs the job will depend on.
        #' @param jobTags Job tags.
        #' @param data Variant stats params.
        runStats=fetchOpenCGA(object=OpencgaR, category="analysis", categoryId=NULL, subcategory="variant/stats",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
    )
})